// @superoffice/tsclient.webapi/src/AppointmentRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          AppointmentRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Appointments appear in the diary, and have links to a Contact/Person and possibly a Project or Sale. They have start and end time+date.
 * Collection of all services that works with Appointment data.
 *
 * `api/v1/Appointment`
 */
export interface IAppointmentRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.AppointmentEntity >;
    Post(entity: carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity >;
    PutById(id:number, entity: carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity >;
    DeleteById(AppointmentEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(AppointmentEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.AppointmentEntity >;
} // interface IAppointmentRestApi




/**
 * Appointments appear in the diary, and have links to a Contact/Person and possibly a Project or Sale. They have start and end time+date.
 * Collection of all services that works with Appointment data.
 *
 * `api/v1/Appointment`
 */
export class AppointmentRestApi extends base.WebApiBase implements IAppointmentRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Appointment/";
        super(options);
    }


    /**
     * `GET /api/v1/Appointment/default`
     * Set default values into a new AppointmentEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.AppointmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AppointmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.AppointmentEntity = converters.Create_AppointmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Appointment/`
     * Creates a new AppointmentEntity 
     * @param newAppointment - The AppointmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New AppointmentEntity
     */
    public async Post(newAppointment: carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AppointmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newAppointment, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.AppointmentEntity = converters.Create_AppointmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Appointment/123`
     * Updates the existing AppointmentEntity
     * @param appointment - The AppointmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AppointmentEntity
     */
    public async PutById(id:number, appointment: carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AppointmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), appointment, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AppointmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AppointmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Appointment/123`
     * Updates an existing AppointmentEntity
     * @param appointment - The partial AppointmentEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AppointmentEntity
     */
    public async PatchById(id:number, appointment: base.WebApiPatchOperation[] | carriers.AppointmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AppointmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AppointmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), appointment, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AppointmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AppointmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Appointment/123`
     * Deletes the AppointmentEntity
     * @param appointmentId - The id of the AppointmentEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(appointmentId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( appointmentId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Appointment/123`
     * Get an existing AppointmentEntity
     * @param appointmentId - The AppointmentEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The AppointmentEntity
     */
    public async GetById( appointmentId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.AppointmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AppointmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( appointmentId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AppointmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AppointmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class AppointmentRestApi
