// @superoffice/tsclient.webapi/src/AttachmentRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          AttachmentRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The Attachment carrier represent meta data for a Service attachment
 * Agent used for Ticket functions
 *
 * `api/v1/Attachment`
 */
export interface IAttachmentRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.AttachmentEntity >;
    Post(entity: carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity >;
    PutById(id:number, entity: carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity >;
    GetById(AttachmentEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.AttachmentEntity >;
} // interface IAttachmentRestApi




/**
 * The Attachment carrier represent meta data for a Service attachment
 * Agent used for Ticket functions
 *
 * `api/v1/Attachment`
 */
export class AttachmentRestApi extends base.WebApiBase implements IAttachmentRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Attachment/";
        super(options);
    }


    /**
     * `GET /api/v1/Attachment/default`
     * Set default values into a new AttachmentEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.AttachmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AttachmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.AttachmentEntity = converters.Create_AttachmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Attachment/`
     * Creates a new AttachmentEntity 
     * @param newAttachment - The AttachmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New AttachmentEntity
     */
    public async Post(newAttachment: carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AttachmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newAttachment, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.AttachmentEntity = converters.Create_AttachmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Attachment/123`
     * Updates the existing AttachmentEntity
     * @param attachment - The AttachmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AttachmentEntity
     */
    public async PutById(id:number, attachment: carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AttachmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), attachment, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AttachmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AttachmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Attachment/123`
     * Updates an existing AttachmentEntity
     * @param attachment - The partial AttachmentEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AttachmentEntity
     */
    public async PatchById(id:number, attachment: base.WebApiPatchOperation[] | carriers.AttachmentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.AttachmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AttachmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), attachment, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AttachmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AttachmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Attachment/123`
     * Get an existing AttachmentEntity
     * @param attachmentId - The AttachmentEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The AttachmentEntity
     */
    public async GetById( attachmentId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.AttachmentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.AttachmentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( attachmentId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.AttachmentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_AttachmentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class AttachmentRestApi
