// @superoffice/tsclient.webapi/src/ContactRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          ContactRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The Contact Service. The service implements all services working with the Contact object
 * Contact (company) data services.
 *
 * `api/v1/Contact`
 */
export interface IContactRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ContactEntity >;
    Post(entity: carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity >;
    PutById(id:number, entity: carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity >;
    DeleteById(ContactEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(ContactEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ContactEntity >;
} // interface IContactRestApi




/**
 * The Contact Service. The service implements all services working with the Contact object
 * Contact (company) data services.
 *
 * `api/v1/Contact`
 */
export class ContactRestApi extends base.WebApiBase implements IContactRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Contact/";
        super(options);
    }


    /**
     * `GET /api/v1/Contact/default`
     * Set default values into a new ContactEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ContactEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ContactEntity = converters.Create_ContactEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Contact/`
     * Creates a new ContactEntity 
     * @param newContact - The ContactEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New ContactEntity
     */
    public async Post(newContact: carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newContact, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ContactEntity = converters.Create_ContactEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Contact/123`
     * Updates the existing ContactEntity
     * @param contact - The ContactEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ContactEntity
     */
    public async PutById(id:number, contact: carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), contact, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Contact/123`
     * Updates an existing ContactEntity
     * @param contact - The partial ContactEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ContactEntity
     */
    public async PatchById(id:number, contact: base.WebApiPatchOperation[] | carriers.ContactEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), contact, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Contact/123`
     * Deletes the ContactEntity
     * @param contactId - The id of the ContactEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(contactId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( contactId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Contact/123`
     * Get an existing ContactEntity
     * @param contactId - The ContactEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The ContactEntity
     */
    public async GetById( contactId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ContactEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( contactId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class ContactRestApi
