// @superoffice/tsclient.webapi/src/CRMScriptRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          CRMScriptRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * CRM Scripts - stored in the ejScript table, including source code.
 * Manage and execute CRMScript functions.
 *
 * `api/v1/CRMScript`
 * 
 * **Online Restricted:** This agent is not available in Online by default. Access must be requested specifically when app is registered.
 */
export interface ICRMScriptRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.CRMScriptEntity >;
    Post(entity: carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity >;
    PutById(id:number, entity: carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity >;
    DeleteById(CRMScriptEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(CRMScriptEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.CRMScriptEntity >;
} // interface ICRMScriptRestApi




/**
 * CRM Scripts - stored in the ejScript table, including source code.
 * Manage and execute CRMScript functions.
 *
 * `api/v1/CRMScript`
 */
export class CRMScriptRestApi extends base.WebApiBase implements ICRMScriptRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/CRMScript/";
        super(options);
    }


    /**
     * `GET /api/v1/CRMScript/default`
     * Set default values into a new CRMScriptEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.CRMScriptEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CRMScriptEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CRMScriptEntity = converters.Create_CRMScriptEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/CRMScript/`
     * Creates a new CRMScriptEntity 
     * @param newCRMScript - The CRMScriptEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New CRMScriptEntity
     */
    public async Post(newCRMScript: carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CRMScriptEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newCRMScript, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CRMScriptEntity = converters.Create_CRMScriptEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/CRMScript/123`
     * Updates the existing CRMScriptEntity
     * @param cRMScript - The CRMScriptEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated CRMScriptEntity
     */
    public async PutById(id:number, cRMScript: carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CRMScriptEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), cRMScript, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.CRMScriptEntity = null;
                if( response.status != 304 )
                    res = converters.Create_CRMScriptEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/CRMScript/123`
     * Updates an existing CRMScriptEntity
     * @param cRMScript - The partial CRMScriptEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated CRMScriptEntity
     */
    public async PatchById(id:number, cRMScript: base.WebApiPatchOperation[] | carriers.CRMScriptEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.CRMScriptEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CRMScriptEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), cRMScript, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.CRMScriptEntity = null;
                if( response.status != 304 )
                    res = converters.Create_CRMScriptEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/CRMScript/123`
     * Deletes the CRMScriptEntity
     * @param cRMScriptId - The id of the CRMScriptEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(cRMScriptId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( cRMScriptId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/CRMScript/123`
     * Get an existing CRMScriptEntity
     * @param cRMScriptId - The CRMScriptEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The CRMScriptEntity
     */
    public async GetById( cRMScriptId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.CRMScriptEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CRMScriptEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( cRMScriptId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.CRMScriptEntity = null;
                if( response.status != 304 )
                    res = converters.Create_CRMScriptEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class CRMScriptRestApi
