// @superoffice/tsclient.webapi/src/DashThemeRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          DashThemeRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Dashboard theme configuration
 * Agent that lets you configure dashboard tiles and retrieve dashboard data
 *
 * `api/v1/DashTheme`
 */
export interface IDashThemeRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.DashTheme >;
    Post(entity: carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme >;
    PutById(id:number, entity: carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme >;
    DeleteById(DashThemeId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(DashThemeId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.DashTheme >;
} // interface IDashThemeRestApi




/**
 * Dashboard theme configuration
 * Agent that lets you configure dashboard tiles and retrieve dashboard data
 *
 * `api/v1/DashTheme`
 */
export class DashThemeRestApi extends base.WebApiBase implements IDashThemeRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/DashTheme/";
        super(options);
    }


    /**
     * `GET /api/v1/DashTheme/default`
     * Set default values into a new DashTheme.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.DashTheme > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DashTheme >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.DashTheme = converters.Create_DashTheme_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/DashTheme/`
     * Creates a new DashTheme 
     * @param newDashTheme - The DashTheme to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New DashTheme
     */
    public async Post(newDashTheme: carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DashTheme >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newDashTheme, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.DashTheme = converters.Create_DashTheme_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/DashTheme/123`
     * Updates the existing DashTheme
     * @param dashTheme - The DashTheme to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated DashTheme
     */
    public async PutById(id:number, dashTheme: carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DashTheme >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), dashTheme, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DashTheme = null;
                if( response.status != 304 )
                    res = converters.Create_DashTheme_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/DashTheme/123`
     * Updates an existing DashTheme
     * @param dashTheme - The partial DashTheme containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated DashTheme
     */
    public async PatchById(id:number, dashTheme: base.WebApiPatchOperation[] | carriers.DashTheme, options?:base.WebApiPutRequestOptions) : Promise< carriers.DashTheme > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DashTheme >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), dashTheme, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DashTheme = null;
                if( response.status != 304 )
                    res = converters.Create_DashTheme_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/DashTheme/123`
     * Deletes the DashTheme
     * @param dashThemeId - The id of the DashTheme to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(dashThemeId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( dashThemeId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/DashTheme/123`
     * Get an existing DashTheme
     * @param dashThemeId - The DashTheme to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The DashTheme
     */
    public async GetById( dashThemeId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.DashTheme > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DashTheme >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( dashThemeId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DashTheme = null;
                if( response.status != 304 )
                    res = converters.Create_DashTheme_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class DashThemeRestApi
