// @superoffice/tsclient.webapi/src/DashboardRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          DashboardRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Dashboard configuration
 * Agent lets you configure dashboard tiles and retrieve dashboard data
 *
 * `api/v1/Dashboard`
 */
export interface IDashboardRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Dashboard >;
    Post(entity: carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard >;
    PutById(id:number, entity: carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard >;
    DeleteById(DashboardId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(DashboardId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.Dashboard >;
} // interface IDashboardRestApi




/**
 * Dashboard configuration
 * Agent lets you configure dashboard tiles and retrieve dashboard data
 *
 * `api/v1/Dashboard`
 */
export class DashboardRestApi extends base.WebApiBase implements IDashboardRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Dashboard/";
        super(options);
    }


    /**
     * `GET /api/v1/Dashboard/default`
     * Set default values into a new Dashboard.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Dashboard > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Dashboard >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Dashboard = converters.Create_Dashboard_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Dashboard/`
     * Creates a new Dashboard 
     * @param newDashboard - The Dashboard to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New Dashboard
     */
    public async Post(newDashboard: carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Dashboard >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newDashboard, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Dashboard = converters.Create_Dashboard_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Dashboard/123`
     * Updates the existing Dashboard
     * @param dashboard - The Dashboard to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated Dashboard
     */
    public async PutById(id:number, dashboard: carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Dashboard >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), dashboard, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Dashboard = null;
                if( response.status != 304 )
                    res = converters.Create_Dashboard_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Dashboard/123`
     * Updates an existing Dashboard
     * @param dashboard - The partial Dashboard containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated Dashboard
     */
    public async PatchById(id:number, dashboard: base.WebApiPatchOperation[] | carriers.Dashboard, options?:base.WebApiPutRequestOptions) : Promise< carriers.Dashboard > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Dashboard >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), dashboard, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Dashboard = null;
                if( response.status != 304 )
                    res = converters.Create_Dashboard_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Dashboard/123`
     * Deletes the Dashboard
     * @param dashboardId - The id of the Dashboard to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(dashboardId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( dashboardId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Dashboard/123`
     * Get an existing Dashboard
     * @param dashboardId - The Dashboard to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The Dashboard
     */
    public async GetById( dashboardId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.Dashboard > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Dashboard >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( dashboardId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Dashboard = null;
                if( response.status != 304 )
                    res = converters.Create_Dashboard_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class DashboardRestApi
