// @superoffice/tsclient.webapi/src/DocumentRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          DocumentRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * 
 * Collection of all services that works with Document data. This is services for the document information, not the physical document themselves. These are handled by the BLOB service methods.
 *
 * `api/v1/Document`
 */
export interface IDocumentRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.DocumentEntity >;
    Post(entity: carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity >;
    PutById(id:number, entity: carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity >;
    DeleteById(DocumentEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(DocumentEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.DocumentEntity >;
} // interface IDocumentRestApi




/**
 * 
 * Collection of all services that works with Document data. This is services for the document information, not the physical document themselves. These are handled by the BLOB service methods.
 *
 * `api/v1/Document`
 */
export class DocumentRestApi extends base.WebApiBase implements IDocumentRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Document/";
        super(options);
    }


    /**
     * `GET /api/v1/Document/default`
     * Set default values into a new DocumentEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.DocumentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DocumentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.DocumentEntity = converters.Create_DocumentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Document/`
     * Creates a new DocumentEntity 
     * @param newDocument - The DocumentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New DocumentEntity
     */
    public async Post(newDocument: carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DocumentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newDocument, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.DocumentEntity = converters.Create_DocumentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Document/123`
     * Updates the existing DocumentEntity
     * @param document - The DocumentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated DocumentEntity
     */
    public async PutById(id:number, document: carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DocumentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), document, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DocumentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_DocumentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Document/123`
     * Updates an existing DocumentEntity
     * @param document - The partial DocumentEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated DocumentEntity
     */
    public async PatchById(id:number, document: base.WebApiPatchOperation[] | carriers.DocumentEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.DocumentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DocumentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), document, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DocumentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_DocumentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Document/123`
     * Deletes the DocumentEntity
     * @param documentId - The id of the DocumentEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(documentId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( documentId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Document/123`
     * Get an existing DocumentEntity
     * @param documentId - The DocumentEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The DocumentEntity
     */
    public async GetById( documentId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.DocumentEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.DocumentEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( documentId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.DocumentEntity = null;
                if( response.status != 304 )
                    res = converters.Create_DocumentEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class DocumentRestApi
