// @superoffice/tsclient.webapi/src/Enums.ts
// <auto-generated>
//      SuperOffice Code Generator
//      modelling\templates\servicesnew\tsenums.tt
// </auto-generated>

//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
// Typescript Enums                               Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------



  /**
   * The type of Address
   * This Enum is used by:  address.atype_idx
   */
  export enum AddressType {
        /**
         * 0: Unknown - used when initializing 
         */
             Unknown = "Unknown",        
        /**
         * 1: The address is a contact's postal address
         */
             ContactPostalAddress = "ContactPostalAddress",        
        /**
         * 2: The address is a contact's street address
         */
             ContactStreetAddress = "ContactStreetAddress",        
        /**
         * 16387: The address is a person's private address
         */
             PersonPrivateAddress = "PersonPrivateAddress",        
        /**
         * 8196: The address is the shipping address for a quote
         */
             QuoteBillingAddress = "QuoteBillingAddress",        
        /**
         * 8197: The address is the shipping address for a quote
         */
             QuoteShippingAddress = "QuoteShippingAddress",        
  }


  /**
   * Value for field 'private' in table 'appointment'.
   * This Enum is used by:  appointment.private
   */
  export enum AppointmentPrivate {
        /**
         * 0: This appointment can be read by anyone
         */
             Public = "Public",        
        /**
         * 1: This appointment can only be read/seen by the owner
         */
             PrivateUser = "PrivateUser",        
        /**
         * 2: This appointment can only be read by members of the owners group
         */
             PrivateGroup = "PrivateGroup",        
  }


  /**
   * Value for field 'status' in table 'appointment'.
   * This Enum is used by:  appointment.status
   */
  export enum AppointmentStatus {
        /**
         * 0: Appointment is not initialized or it is a Post-IT for users diary, used together with ApointmentType.Note only
         */
             UnknownOrPostIt = "UnknownOrPostIt",        
        /**
         * 1: Appointment has not been started
         */
             NotStarted = "NotStarted",        
        /**
         * 2: Appointment has been started, but not completed
         */
             Started = "Started",        
        /**
         * 3: Appointment is completed
         */
             Completed = "Completed",        
        /**
         * 4: Appointment is hidden
         */
             Hidden = "Hidden",        
        /**
         * 5: Appointment is a booking invitation (initial status for a booking)
         */
             Booking = "Booking",        
        /**
         * 6: Appointment is a booking that has been moved since it was seen by the invitee
         */
             BookingMoved = "BookingMoved",        
        /**
         * 7: Appointment is a booking and has been seen, but neither accepted nor rejected
         */
             BookingSeen = "BookingSeen",        
        /**
         * 8: Appointment is a booking that has been moved. The movement has been seen but neither accepted nor rejected
         */
             BookingMovedSeen = "BookingMovedSeen",        
        /**
         * 9: Appointment is a booking that has been seen and declined
         */
             BookingDeclined = "BookingDeclined",        
        /**
         * 10: Status of "child" appointment whose "mother" has been deleted (the meeting is cancelled)
         */
             BookingDeleted = "BookingDeleted",        
        /**
         * 11: You are assigned this appointment (initial status, like Booking(5) )
         */
             Assignment = "Assignment",        
        /**
         * 12: You have seen the assignment, but not yet declined or accepted it
         */
             AssignmentSeen = "AssignmentSeen",        
        /**
         * 13: You have declined the appointment assigned to you
         */
             AssignmentDeclined = "AssignmentDeclined",        
  }


  /**
   * Value for field 'type' in table 'appointment'.
   * This Enum is used by:  appointment.type
   */
  export enum AppointmentType {
        /**
         * 0: Unknown - used when initializing 
         */
             Unknown = "Unknown",        
        /**
         * 1: Show in diary. If overdue or done today, show in checklist also
         */
             inDiary = "inDiary",        
        /**
         * 2: Appointment is a followup with no specific starttime
         */
             inChecklist = "inChecklist",        
        /**
         * 3: Paper clip on diary, expands to “post-it” note when clicked
         */
             Note = "Note",        
        /**
         * 4: Document of any kind, including outgoing, incoming and email
         */
             Document = "Document",        
        /**
         * 5: Saved Report (i.e., not a Report Definition but the result of an actual report run)
         */
             SavedReport = "SavedReport",        
        /**
         * 6: Booking, made for diary. When accepted this will change type to inDiary
         */
             BookingForDiary = "BookingForDiary",        
        /**
         * 7: Booking, made for check list. When accepted this will change type to inChecklist
         */
             BookingForChecklist = "BookingForChecklist",        
        /**
         * 8: Document is a template for mailmerge operations
         */
             MergeDraft = "MergeDraft",        
        /**
         * 9: Document is a result of a mailmerge operations (there are probably many documents like this!)
         */
             MergeFinal = "MergeFinal",        
  }


  /**
   * Access restrictions and mandatory status, if any.
   * This Enum is used by:  ErpField.Access
   */
  export enum FieldAccess {
        /**
         * 0: Normal field, no particular restrictions
         */
             Normal = "Normal",        
        /**
         * 1: This field is mandatory.
         */
             Mandatory = "Mandatory",        
        /**
         * 2: This field is read-only.
         */
             ReadOnly = "ReadOnly",        
  }


  /**
   * Describes the different types of controls that can appear in the Configure connection dialog
   * This Enum is used by:  ErpField.FieldType
   */
  export enum FieldMetadataType {
        /**
         * 0: Checkbox widget: Boolean value
         */
             Checkbox = "Checkbox",        
        /**
         * 1: Single line textbox.
         */
             Text = "Text",        
        /**
         * 2: Single line edit field with password characters - input is hidden.
         */
             Password = "Password",        
        /**
         * 3: Single line integer number field. 
         */
             Integer = "Integer",        
        /**
         * 4: Single line decimal number field. 
         */
             Double = "Double",        
        /**
         * 5: Dropdown listbox. The list name must be provided. The list items are provided by another interface. The default value should be the list key, not the text of the list item.
         */
             List = "List",        
        /**
         * 6: Date
         */
             Date = "Date",        
        /**
         * 99: Static text - has no input widget associated with it. Can be used for spacing/layout.
         */
             Label = "Label",        
  }


  /**
   * Value for field 'direction' in table 'doctmpl'.
   * This Enum is used by:  DocTmpl.direction
   */
  export enum DocTmplDirection {
        /**
         * 0: The direction of documents from this template is unknown/undecided
         */
             Unknown = "Unknown",        
        /**
         * 1: Document template for incoming documents
         */
             Incoming = "Incoming",        
        /**
         * 2: Document template is used for outgoing documents
         */
             Outgoing = "Outgoing",        
        /**
         * 3: All' choice for Saint. This is NOT an acceptable value for a task definition, but is used by the Saint system for indexing all directions
         */
             SaintAll = "SaintAll",        
  }


  /**
   * Is this document template some kind of quote document, and if so what
   * This Enum is used by:  DocTmpl.quoteDocType
   */
  export enum DocTmplQuoteType {
        /**
         * 0: This document template is in no way related to Quote Management
         */
             None = "None",        
        /**
         * 1: This is a template for the offer mail body
         */
             MailBody = "MailBody",        
        /**
         * 2: This is a template for the main offer document, document instances will be referred through Quote.DocumentId
         */
             MainDocument = "MainDocument",        
        /**
         * 3: This is a template for the AsposeWords merge process, generating the offer line details
         */
             QuoteLines = "QuoteLines",        
        /**
         * 4: This is a template for the mail body of the order confirmation
         */
             ConfirmationMailBody = "ConfirmationMailBody",        
        /**
         * 5: This is a template for the lines document that is an attachment to the confirmation mail
         */
             ConfirmationLines = "ConfirmationLines",        
  }


  /**
   * Value for field 'record_type' in table 'doctmpl'.
   * This Enum is used by:  DocTmpl.record_type
   */
  export enum DocTmplType {
        /**
         * 0: The document type is unknown/undecided
         */
             Unknown = "Unknown",        
        /**
         * 1: (not used for document templates)
         */
             Appointment = "Appointment",        
        /**
         * 2: Generic text document (for instance, a letter)
         */
             Document = "Document",        
        /**
         * 3: Email message
         */
             Email = "Email",        
        /**
         * 4: Facsimile
         */
             Fax = "Fax",        
        /**
         * 5: (not used for document templates)
         */
             Phone = "Phone",        
        /**
         * 6: (not used for document templates)
         */
             Todo = "Todo",        
        /**
         * 7: Template for mailmerge operations
         */
             MergeDraft = "MergeDraft",        
        /**
         * 8: Result of a mailmerge operation
         */
             MergeFinal = "MergeFinal",        
        /**
         * 9: Saved result of running a report (output from Reporter)
         */
             SavedReport = "SavedReport",        
  }


  /**
   * Value for the executeOnEvent field in table externalapplication
   * This Enum is used by:  ExtApp.executeOnEvent
   */
  export enum ExecuteOnEvent {
        /**
         * 0: This application is never executed automatically
         */
             Never = "Never",        
        /**
         * 1: Execute this application during logon
         */
             Logon = "Logon",        
        /**
         * 2: Execute this application before logoff
         */
             Logoff = "Logoff",        
        /**
         * 3: Execute this application before a local update (start travelling)
         */
             LocalUpdate = "LocalUpdate",        
        /**
         * 16384: When executing this application, wait for it to start
         */
             Wait = "Wait",        
  }


  /**
   * String, int, decimal, image, url, etc. How should the value be interpreted.
   */
  export enum ExtraDataFieldType {
        /**
         * 0: Anything to be shown as text, left aligned. If type is string, it can contain formatspecifiers (use this for displaying numbers correctly as the user's computer is set up).
         */
             String = "String",        
        /**
         * 1: A web address
         */
             Url = "Url",        
        /**
         * 2: Contains the url of the image, or a base64 encoded version of it.
         */
             Image = "Image",        
  }


  /**
   * Value for the navigation field in table externalapplication
   * This Enum is used by:  ExtApp.navigation
   */
  export enum Navigation {
        /**
         * 0: This application has no explicit navigation in GUI
         */
             Invisible = "Invisible",        
        /**
         * 1: This application appears in the list of the Toolbox icon in the CRM5 Navigator sidebar
         */
             ToolboxMenu = "ToolboxMenu",        
        /**
         * 2: This application has its own navigator button (remember to set icon resource)
         */
             NavigatorButton = "NavigatorButton",        
        /**
         * 3: This application appears in the View pulldown menu
         */
             ViewMenu = "ViewMenu",        
        /**
         * 4: (not yet implemented) This application appears as a Task in the Selection Task card
         */
             SelectionTaskCard = "SelectionTaskCard",        
        /**
         * 5: This application (type IntegratedHTML or IntegratedURL) appears as a tab on the Contact card
         */
             ContactCard = "ContactCard",        
        /**
         * 6: This application (type IntegratedHTML or IntegratedURL)  appears as a tab on the Contact Archive
         */
             ContactArchive = "ContactArchive",        
        /**
         * 7: This application (type IntegratedHTML or IntegratedURL)  appears as a tab on the Project card
         */
             ProjectCard = "ProjectCard",        
        /**
         * 8: This application (type IntegratedHTML or IntegratedURL)  appears as a tab on the Project Archive
         */
             ProjectArchive = "ProjectArchive",        
        /**
         * 9: This application (type IntegratedHTML or IntegratedURL)  appears as a tab in the Sale dialog
         */
             SaleCard = "SaleCard",        
        /**
         * 10: This application (type IntegratedHTML or IntegratedURL)  appears as a tab in the Person Card
         */
             PersonCard = "PersonCard",        
        /**
         * 11: This application (type IntegratedHTML or IntegratedURL)  appears as a tab in the Appointment dialog
         */
             ActivityDialog = "ActivityDialog",        
        /**
         * 12: This application (type IntegratedHTML or IntegratedURL)  appears as a tab in the Document dialog
         */
             DocumentDialog = "DocumentDialog",        
        /**
         * 13: In the Browser panel
         */
             BrowserPanel = "BrowserPanel",        
        /**
         * 14: Task button visible on the Contact Selection task panel
         */
             ContSelectionTask = "ContSelectionTask",        
        /**
         * 15: Task button visible on the Appointment Selection task panel
         */
             AppntSelectionTask = "AppntSelectionTask",        
        /**
         * 16: Task button visible on the Sale Selection task panel
         */
             SaleSelectionTask = "SaleSelectionTask",        
        /**
         * 17: Task button visible on the Document Selection task panel
         */
             DocSelectionTask = "DocSelectionTask",        
        /**
         * 18: Task button visible on the Project Selection task panel
         */
             ProjSelectionTask = "ProjSelectionTask",        
        /**
         * 19: In company minicard
         */
             CompanyMinicard = "CompanyMinicard",        
        /**
         * 20: In project minicard
         */
             ProjectMinicard = "ProjectMinicard",        
        /**
         * 21: In diary minicard
         */
             DiaryMinicard = "DiaryMinicard",        
        /**
         * 22: In selection minicard
         */
             SelectionMinicard = "SelectionMinicard",        
        /**
         * 23: In the main ButtonBar
         */
             ButtonPanelTask = "ButtonPanelTask",        
        /**
         * 24: In the appointment dialog
         */
             AppointmentDialogTask = "AppointmentDialogTask",        
        /**
         * 25: In the sale dialog
         */
             SaleDialogTask = "SaleDialogTask",        
        /**
         * 26: In the document dialog
         */
             DocumentDialogTask = "DocumentDialogTask",        
        /**
         * 27: In the person dialog
         */
             PersonDialogTask = "PersonDialogTask",        
        /**
         * 28: In the sale minicard
         */
             SaleMinicard = "SaleMinicard",        
        /**
         * 29: In the sale archive
         */
             SaleArchive = "SaleArchive",        
        /**
         * 30: Task tab for appointment selection, while showing shadow sel
         */
             AppntSelectionShadowTask = "AppntSelectionShadowTask",        
        /**
         * 31: Task tab for sale selection, while showing shadow sel
         */
             SaleSelectionShadowTask = "SaleSelectionShadowTask",        
        /**
         * 32: Task tab for document selection, while showing shadow sel
         */
             DocSelectionShadowTask = "DocSelectionShadowTask",        
        /**
         * 33: Task tab for project selection, while showing shadow sel
         */
             ProjSelectionShadowTask = "ProjSelectionShadowTask",        
        /**
         * 34: Context (popup) menu in Diary archive
         */
             DiaryArchive = "DiaryArchive",        
        /**
         * 35: Context (popup) menu in Contact selection
         */
             SelectionContactArchive = "SelectionContactArchive",        
        /**
         * 36: Context (popup) menu in Project archive
         */
             SelectionProjectArchive = "SelectionProjectArchive",        
        /**
         * 37: Context (popup) menu in Sale archive
         */
             SelectionSaleArchive = "SelectionSaleArchive",        
        /**
         * 38: Context (popup) menu in Appointment archive
         */
             SelectionAppointmentArchive = "SelectionAppointmentArchive",        
        /**
         * 39: Context (popup) menu in Document archive
         */
             SelectionDocumentArchive = "SelectionDocumentArchive",        
        /**
         * 40: Task card in Contact selection, when a custom archive is shown
         */
             ContSelectionCustomTask = "ContSelectionCustomTask",        
        /**
         * 41: Task card in Appointment selection, when a custom archive is shown
         */
             AppntSelectionCustomTask = "AppntSelectionCustomTask",        
        /**
         * 42: Task card in Sale selection, when a custom archive is shown
         */
             SaleSelectionCustomTask = "SaleSelectionCustomTask",        
        /**
         * 43: Task card in Document selection, when a custom archive is shown
         */
             DocSelectionCustomTask = "DocSelectionCustomTask",        
        /**
         * 44: Task card in Project selection, when a custom archive is shown
         */
             ProjSelectionCustomTask = "ProjSelectionCustomTask",        
        /**
         * 45: ?
         */
             CustomArchiveMiniCard = "CustomArchiveMiniCard",        
        /**
         * 46: ?
         */
             SelectionCard = "SelectionCard",        
        /**
         * 47: In the Reporter panel minicard, so far only in Web
         */
             ReportMinicard = "ReportMinicard",        
        /**
         * 48: 
         */
             QuoteDialog = "QuoteDialog",        
        /**
         * 49: 
         */
             QuoteDialogTask = "QuoteDialogTask",        
        /**
         * 50: 
         */
             QuoteDialogArchive = "QuoteDialogArchive",        
        /**
         * 51: 
         */
             QuoteLineDialogTask = "QuoteLineDialogTask",        
        /**
         * 52: 
         */
             QuoteLineDialog = "QuoteLineDialog",        
        /**
         * 53: 
         */
             QuoteLineSelectionMainTask = "QuoteLineSelectionMainTask",        
        /**
         * 54: 
         */
             QuoteLineSelectionShadowTask = "QuoteLineSelectionShadowTask",        
        /**
         * 55: 
         */
             SelectionQuoteLineArchive = "SelectionQuoteLineArchive",        
        /**
         * 56: 
         */
             QuoteLineSelectionCustomTask = "QuoteLineSelectionCustomTask",        
        /**
         * 57: 
         */
             FindSystem = "FindSystem",        
        /**
         * 58: Task button visible on the Mailing Selection task panel
         */
             MailingSelectionTask = "MailingSelectionTask",        
        /**
         * 59: Task button visible on the Mailing Selection task panel
         */
             ContactSelectionMailingsTask = "ContactSelectionMailingsTask",        
        /**
         * 60: Task button visible on the Mailing Selection task panel
         */
             AppointmentSelectionMailingsTask = "AppointmentSelectionMailingsTask",        
        /**
         * 61: Task button visible on the Mailing Selection task panel
         */
             SaleSelectionMailingsTask = "SaleSelectionMailingsTask",        
        /**
         * 62: Task button visible on the Mailing Selection task panel
         */
             DocumentSelectionMailingsTask = "DocumentSelectionMailingsTask",        
        /**
         * 63: Task button visible on the Mailing Selection task panel
         */
             ProjectSelectionMailingsTask = "ProjectSelectionMailingsTask",        
        /**
         * 64: Task button visible on the Mailing Selection task panel
         */
             QuoteLineSelectionMailingsTask = "QuoteLineSelectionMailingsTask",        
        /**
         * 65: Visible in top panels new menu
         */
             TopPanelNewMenu = "TopPanelNewMenu",        
        /**
         * 66: Dashboard panel
         */
             Dashboard = "Dashboard",        
        /**
         * 67: This application (type IntegratedHTML or IntegratedURL)  appears as a tab in the Person Archive
         */
             PersonArchive = "PersonArchive",        
        /**
         * 68: In the person minicard
         */
             PersonMinicard = "PersonMinicard",        
        /**
         * 69: In the company card, task button
         */
             CompanyCardTask = "CompanyCardTask",        
        /**
         * 70: In the project card, task button
         */
             ProjectCardTask = "ProjectCardTask",        
        /**
         * 71: This application (type IntegratedHTML or IntegratedURL) appears as a tab in the Request card
         */
             TicketCard = "TicketCard",        
        /**
         * 72: In the ticket minicard
         */
             TicketMinicard = "TicketMinicard",        
  }


  /**
   * Value for field 'accessflags' in table 'prefdesc'.
   * This Enum is used by:  PrefDesc.accessFlags
   * This Enum consists of flag values that can be combined.
   */
  export enum PrefDescAccessFlags {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = 0,        
        /**
         * 1: Wizard mode is password-protected and only accessible to SuperOffice personnel.
         */
             WizardMode = 1,        
        /**
         * 2: Preference settable only by CRM5 superuser
         */
             Level0 = 2,        
        /**
         * 4: Show preference in SOADMIN GUI
         */
             adminGUI = 4,        
        /**
         * 8: Show preference in CRM5 GUI
         */
             CRMGUI = 8,        
  }


  /**
   * Value for field 'valueType' in table 'prefdesc'.
   * This Enum is used by:  PrefDesc.valueType
   */
  export enum PrefDescValueType {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: Preference value is a number
         */
             Number = "Number",        
        /**
         * 2: Preference value is a text (max len 99 characters)
         */
             Text = "Text",        
        /**
         * 3: Preference value is Boolean
         */
             Bool = "Bool",        
        /**
         * 4: Preference value is a list of separate value
         */
             ListOfValues = "ListOfValues",        
        /**
         * 5: Preference value is a record ID in the given MDO list table; maxValue contains the tablenumberø
         */
             ListTableRef = "ListTableRef",        
        /**
         * 6: Preference value is a time, value is seconds since midnight (28800 = 08:00)
         */
             TimeList = "TimeList",        
        /**
         * 7: Preference value is a contact ID
         */
             ContactID = "ContactID",        
        /**
         * 8: Preference value is a person ID
         */
             PersonID = "PersonID",        
        /**
         * 9: Preference value is a project ID
         */
             ProjectID = "ProjectID",        
        /**
         * 10: Preference value is a selection ID
         */
             SelectionID = "SelectionID",        
        /**
         * 11: Obsolete: Windows position and size
         */
             PosSize = "PosSize",        
        /**
         * 12: TimeZone location
         */
             TimeZone = "TimeZone",        
        /**
         * 13: Time, preference value is 13:45:50
         */
             Time = "Time",        
        /**
         * 14: Password, stored as plain text but shown with dots
         */
             Password = "Password",        
        /**
         * 15: Multi-line text that preserves line breaks
         */
             MultiLineText = "MultiLineText",        
  }


  /**
   * Value for the 'status' field in the sale table
   * This Enum is used by:  sale.status; SaleHist.status; CounterValue.sale_status
   */
  export enum SaleStatus {
        /**
         * 0: Sale status is unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Sale is open
         */
             Open = "Open",        
        /**
         * 2: Sale has been sold (green $ in GUI)
         */
             Sold = "Sold",        
        /**
         * 3: Sale has been lost (red $ in GUI)
         */
             Lost = "Lost",        
        /**
         * 4: Sale has been stalled, or "parked", awaiting further developments
         */
             Stalled = "Stalled",        
        /**
         * 1000: All' choice for Saint. This is NOT an acceptable value for a sale, but is used by the Saint system for indexing all sales
         */
             SaintAll = "SaintAll",        
  }


  /**
   * Value for the showState field in table externalapplication
   * This Enum is used by:  ExtApp.showState
   * This Enum consists of flag values that can be combined.
   */
  export enum ShowState {
        /**
         * 0: Start application in minimized state
         */
             Minimized = 0,        
        /**
         * 1: Start application in maximized state
         */
             Maximized = 1,        
        /**
         * 2: Start application with default window size and position
         */
             Default = 2,        
        /**
         * 4: Undefined and generated value 4
         */
             UndefinedValue4 = 4,        
        /**
         * 8: Undefined and generated value 8
         */
             UndefinedValue8 = 8,        
        /**
         * 16: Undefined and generated value 16
         */
             UndefinedValue16 = 16,        
        /**
         * 32: Undefined and generated value 32
         */
             UndefinedValue32 = 32,        
        /**
         * 64: Undefined and generated value 64
         */
             UndefinedValue64 = 64,        
        /**
         * 128: Undefined and generated value 128
         */
             UndefinedValue128 = 128,        
        /**
         * 256: Show the Browser toolbar
         */
             ToolBar = 256,        
        /**
         * 512: Show the Browser addressbar (where the user can type URL's and stuff)
         */
             AddressBar = 512,        
        /**
         * 1024: Show the Browser statusbar
         */
             StatusBar = 1024,        
        /**
         * 2048: Show the Browser Menu bar (currently not implemented)
         */
             MenuBar = 2048,        
  }


  /**
   * Value for field 'direction' in table 'task'. Controls icons used in GUI
   * This Enum is used by:  Task.direction
   */
  export enum TaskDirection {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: The task is sent in to SuperOffice CRM 5
         */
             Incoming = "Incoming",        
        /**
         * 2: The task is sent out from SuperOffice CRM 5
         */
             Outgoing = "Outgoing",        
        /**
         * 3: All' choice for Saint. This is NOT an acceptable value for a task definition, but is used by the Saint system for indexing all directions
         */
             SaintAll = "SaintAll",        
  }


  /**
   * Value for field 'record_type' in table 'task'. Controls icons used in GUI
   * This Enum is used by:  Task.record_type
   */
  export enum TaskType {
        /**
         * 0: This task is of an unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: The task is an appointment
         */
             Appointment = "Appointment",        
        /**
         * 2: The task is a document
         */
             Document = "Document",        
        /**
         * 3: The task is an email
         */
             Email = "Email",        
        /**
         * 4: The task is a fax
         */
             Fax = "Fax",        
        /**
         * 5: The task is a phone call
         */
             Phone = "Phone",        
        /**
         * 6: The task is a To-do
         */
             ToDo = "ToDo",        
        /**
         * 7: Mail merge draft document
         */
             MailMergeDraft = "MailMergeDraft",        
        /**
         * 8: Merged result of mail merge
         */
             MailMergeFinal = "MailMergeFinal",        
        /**
         * 9: Saved report run
         */
             Report = "Report",        
        /**
         * 10: All' choice for Saint, every activity is recorded under this type as well as the specific type. This is NOT an acceptable value for an individual activity
         */
             SaintAll = "SaintAll",        
  }


  /**
   * Value for field 'deflevel' in table 'userpreference'.
   * This Enum is used by:  UserPreference.deflevel; UserPreference.maxlevel; PrefDesc.maxLevel; PrefDesc.sysMaxLevel
   */
  export enum PreferenceLevel {
        /**
         * 0: Undefined value (error)
         */
             Undefined = "Undefined",        
        /**
         * 1: Default value used if no other value has been specified
         */
             HardDefault = "HardDefault",        
        /**
         * 2: Value valid for all databases under this licence
         */
             SystemWide = "SystemWide",        
        /**
         * 3: Value valid for this database
         */
             Database = "Database",        
        /**
         * 4: Value valid for this group
         */
             Group = "Group",        
        /**
         * 5: Value valid for this user
         */
             Individual = "Individual",        
        /**
         * 6: Value valid for this computer (note: stored in the Registry, not in the database)
         */
             PC = "PC",        
  }


  /**
   * Units of duration - from seconds to centuries
   * This Enum is used by:  ProjType.durationUnit; SaleType.durationUnit
   */
  export enum DurationUnit {
        /**
         * 0: Duration is unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Duration is in Seconds
         */
             Second = "Second",        
        /**
         * 2: Duration is in Minutes
         */
             Minute = "Minute",        
        /**
         * 3: Duration is in Hours
         */
             Hour = "Hour",        
        /**
         * 4: Duration is in Days
         */
             Day = "Day",        
        /**
         * 5: Duration is in Weeks
         */
             Week = "Week",        
        /**
         * 6: Duration is in Months
         */
             Month = "Month",        
        /**
         * 7: Duration is in Quarters
         */
             Quarter = "Quarter",        
        /**
         * 8: Duration is in HalfYears
         */
             HalfYear = "HalfYear",        
        /**
         * 9: Duration is in Years
         */
             Year = "Year",        
        /**
         * 10: Duration is in Decades
         */
             Decade = "Decade",        
        /**
         * 11: Duration is in Centurys
         */
             Century = "Century",        
        /**
         * 12: Duration is in Milleniums
         */
             Millenium = "Millenium",        
  }


  /**
   * How the selections in a combined selections should be put together
   * This Enum is used by:  selection.combinationType
   */
  export enum SelectionUnionType {
        /**
         * 0: Combination type is unknown/indefined
         */
             Unknown = "Unknown",        
        /**
         * 1: All of left, except where it overlaps with right
         */
             SubtractRightFromLeft = "SubtractRightFromLeft",        
        /**
         * 2: All of right, except where it overlaps with left
         */
             SubtractLeftFromRight = "SubtractLeftFromRight",        
        /**
         * 3: Those present in left AND right only
         */
             Intersect = "Intersect",        
        /**
         * 4: Those present in left or right, but not both
         */
             XOR = "XOR",        
        /**
         * 5: All, but no duplicates for those present in left and right
         */
             Union = "Union",        
  }


  /**
   * Status for shipment (started, cancelled etc)
   * This Enum is used by:  s_shipment.status
   */
  export enum ShipmentStatus {
        /**
         * 0: Initial status, shipment is left alone, can be used for working with shipment
         */
             None = "None",        
        /**
         * 1: Finished, no errors.
         */
             Ok = "Ok",        
        /**
         * 2: Stopped by user.
         */
             Canceled = "Canceled",        
        /**
         * 3: Ready for execution, in the queue.
         */
             Waiting = "Waiting",        
        /**
         * 4: Executing (sending).
         */
             Started = "Started",        
        /**
         * 5: Finished, had some errors.
         */
             StatusFailed = "StatusFailed",        
        /**
         * 6: Retrying the addresses which did not work
         */
             Retry = "Retry",        
        /**
         * 7: Between waiting and starting. Used when we populate the recipient lists.
         */
             Populating = "Populating",        
        /**
         * 8: Deleted
         */
             Deleted = "Deleted",        
        /**
         * 9: Awaiting ejournalCron to start populate
         */
             AwaitPopulate = "AwaitPopulate",        
        /**
         * 10: eJournalCron has started populateOnly after someone saved choose recipients screen.
         */
             PopulatingOnly = "PopulatingOnly",        
        /**
         * 11: eJournalCron has started populating, but by user request it is canceled. Now we finish populating, and then set status back to none.
         */
             PopulateAborted = "PopulateAborted",        
        /**
         * 12: 
         */
             PopulateError = "PopulateError",        
        /**
         * 13: 
         */
             TooManyRecipients = "TooManyRecipients",        
  }


  /**
   * Status for sending to a particular recipient
   * This Enum is used by:  s_shipment_addr.status
   */
  export enum ShipmentAddrStatus {
        /**
         * 0: Default value for empty rows
         */
             Unknown = "Unknown",        
        /**
         * 1: Ready for shipment
         */
             Ready = "Ready",        
        /**
         * 2: Blocked because it duplicates another shipment addr in the same batch
         */
             Duplicate = "Duplicate",        
        /**
         * 3: Blocked because of nomailing flag
         */
             Blocked = "Blocked",        
        /**
         * 4: Sent
         */
             Sent = "Sent",        
        /**
         * 5: Bounce received from transport channel
         */
             Bounced = "Bounced",        
        /**
         * 6: Recipient has opened the message, we know because of tracking
         */
             Opened = "Opened",        
        /**
         * 7: Test recipient, will be deleted after sent (enum name is not correct, but that cant be changed now)
         */
             Clicked = "Clicked",        
        /**
         * 8: Vacation notice, etc
         */
             SoftBounced = "SoftBounced",        
        /**
         * 9: Recipient does not have subscription for shipmentType
         */
             NoSubscription = "NoSubscription",        
        /**
         * 10: Recipient sent a complaint
         */
             Complained = "Complained",        
        /**
         * 11: Recipient address has bounced too many times
         */
             TooManyBounces = "TooManyBounces",        
  }


  /**
   * Communication channel leading to ticket being created
   * This Enum is used by:  ticket.origin
   */
  export enum TicketOrigin {
        /**
         * 0: We just have no idea. Maybe it was too long ago
         */
             Unknown = "Unknown",        
        /**
         * 1: Email communiqué
         */
             Email = "Email",        
        /**
         * 2: SMS 
         */
             SMS = "SMS",        
        /**
         * 3: Telefacsimile
         */
             Fax = "Fax",        
        /**
         * 4: Phone call to human operator
         */
             Phone = "Phone",        
        /**
         * 5: Facebook wall
         */
             Facebook = "Facebook",        
        /**
         * 6: Tweet tweet
         */
             Twitter = "Twitter",        
        /**
         * 7: Internal, by human operator
         */
             Internal = "Internal",        
        /**
         * 8: Direct by customer through CustomerCentre web pages
         */
             CustomerCentre = "CustomerCentre",        
        /**
         * 9: Auto-generated from Emarketing link
         */
             EMarketing = "EMarketing",        
        /**
         * 10: Automatic processes in CS
         */
             AutoGenerated = "AutoGenerated",        
        /**
         * 11: Chat
         */
             Chat = "Chat",        
        /**
         * 12: Form submission
         */
             Form = "Form",        
  }


  /**
   * Which field in ticket we count time spent on
   * This Enum is used by:  ticket_status.time_counter
   */
  export enum TicketStatusTimeCounter {
        /**
         * 0: Not counted
         */
             None = "None",        
        /**
         * 1: Time is counted in the _internally field
         */
             Internally = "Internally",        
        /**
         * 2: Time is counted in the _externally field
         */
             Externally = "Externally",        
        /**
         * 3: Time is counted in the _queue field
         */
             Queue = "Queue",        
  }


  /**
   * State of a Quote Version
   * This Enum is used by:  QuoteVersion.State; QuoteVersion.ArchivedState
   */
  export enum QuoteVersionState {
        /**
         * 0: State unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Draft that has been calculated, and ERP has verified it as OK
         */
             Draft = "Draft",        
        /**
         * 2: This is a draft that has not been calculated, prices and other attributes may be 'invalid' relative to ERP system rules
         */
             DraftNotCalculated = "DraftNotCalculated",        
        /**
         * 3: Draft has been checked, and there was a problem that needs (human) approval
         */
             DraftNeedsApproval = "DraftNeedsApproval",        
        /**
         * 4: Draft with potential problems has been human-approved
         */
             DraftApproved = "DraftApproved",        
        /**
         * 5: Draft has been checked, there was a problem, and (human) approval was NOT granted
         */
             DraftNotApproved = "DraftNotApproved",        
        /**
         * 6: Sent to customer, and is now a legally binding document
         */
             Sent = "Sent",        
        /**
         * 7: Archived - previous state saved in ArchivedState, newer versions exist
         */
             Archived = "Archived",        
        /**
         * 8: Version was rejected be the customer - we lost the sale
         */
             Lost = "Lost",        
        /**
         * 9: Quote was accepted and the sale is 'Sold'
         */
             Sold = "Sold",        
  }


  /**
   * The link type, often synonymous with the blob type, of a BinaryObjectLink row
   * This Enum is used by:  BinaryObjectLink.linkType
   */
  export enum BlobLinkType {
        /**
         * 1: Portrait picture of a person
         */
             PersonImage = "PersonImage",        
        /**
         * 2: Picture bound to a project
         */
             ProjectImage = "ProjectImage",        
        /**
         * 3: Picture used for a project of type Event, in the Audience client
         */
             EventImage = "EventImage",        
        /**
         * 4: Generic thumbnail - not used?
         */
             Thumbnail = "Thumbnail",        
        /**
         * 5: Watermark image for a Status Monitor in Sales Intelligence
         */
             StatusMonitorImage = "StatusMonitorImage",        
        /**
         * 6: Information related to a BatchTask in NetServer
         */
             BatchTask = "BatchTask",        
        /**
         * 7: Image (possibly more than one, differentiated by rank) of a Product; exactly one row should have rank=1
         */
             ProductImage = "ProductImage",        
        /**
         * 8: Thumbnail image for a product, only one allowed per product
         */
             ProductThumbnail = "ProductThumbnail",        
        /**
         * 9: Image (possibly more than one, differentiated by rank) of a Product image for a quote line, only one allowed per quote line
         */
             QuoteLineImage = "QuoteLineImage",        
        /**
         * 10: Thumbnail image for a quote line, only one allowed per quote line
         */
             QuoteLineThumbnail = "QuoteLineThumbnail",        
        /**
         * 11: Access Token used by Cloud Documents for accessing external services
         */
             AccessToken = "AccessToken",        
        /**
         * 12: Refresh Token used by Cloud Documents for accessing external services
         */
             RefreshToken = "RefreshToken",        
        /**
         * 13: Dashboard
         */
             Dashboard = "Dashboard",        
        /**
         * 14: Dashboard tile
         */
             DashboardTile = "DashboardTile",        
        /**
         * 15: Chat topic image
         */
             ChatTopicImage = "ChatTopicImage",        
        /**
         * 16: Forms background image
         */
             FormsBackgroundImage = "FormsBackgroundImage",        
        /**
         * 17: Contact logo image
         */
             ContactImage = "ContactImage",        
        /**
         * 18: HTML data for dashboard tiles
         */
             DashboardHtmlTileData = "DashboardHtmlTileData",        
  }


  /**
   * The state of a quote-alternative or line: Ok / OkWithInfo / Warn / Error
   * This Enum is used by:  QuoteVersion.Status; QuoteAlternative.Status; QuoteLine.Status
   */
  export enum QuoteStatus {
        /**
         * 0: OK, all is good
         */
             Ok = "Ok",        
        /**
         * 1: All is good, but there is some additional information that the user should be made aware of.
         */
             OkWithInfo = "OkWithInfo",        
        /**
         * 2: There is a problem that the user must be made aware of.
         */
             Warning = "Warning",        
        /**
         * 3: There is a problem that the system will not be able to get around. The user needs to do something. Example: The user has registered a quoteline that is discountinued and the amount the user has registered is not in stock.
         */
             Error = "Error",        
  }


  /**
   * Has a pre-calculated value been overriden in a QuoteAlternative or QuoteLine, and in that case what & how?
   * This Enum is used by:  QuoteAlternative.UserValueOverride; QuoteLine.UserValueOverride
   */
  export enum ValueOverride {
        /**
         * 0: Discount data is from the ERP system and has not been overridden
         */
             None = "None",        
        /**
         * 1: The Total amount has been overridden; all other fields should be recalculated
         */
             Total = "Total",        
        /**
         * 2: The discount percentage has been overriden; all other fields should be recalculated
         */
             DiscountPercent = "DiscountPercent",        
        /**
         * 3: The discount amount has been overridden; all other fields should be recalculated
         */
             DiscountAmount = "DiscountAmount",        
        /**
         * 4: The earning percent has been overridden; all other fields should be recalculated
         */
             EarningPercent = "EarningPercent",        
        /**
         * 5: The earning amount has been overridden; all other fields should be recalculated
         */
             EarningAmount = "EarningAmount",        
  }


  /**
   * Actor type within the ERP system, related but not identitcal to SuperOffice Entity type
   * This Enum is used by:  ErpConnectionActorType.ActorTypeErp; ErpExternalKey.ActorType
   */
  export enum ErpActorType {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: Erp Customer
         */
             Customer = "Customer",        
        /**
         * 2: Supplier or Vendor
         */
             Supplier = "Supplier",        
        /**
         * 3: Both customer and supplier
         */
             Partner = "Partner",        
        /**
         * 4: Contact person
         */
             Person = "Person",        
        /**
         * 5: Project
         */
             Project = "Project",        
        /**
         * 6: Employee, often mapped to associate
         */
             Employee = "Employee",        
        /**
         * 7: Sale
         */
             Sale = "Sale",        
  }


  /**
   * Actor type within the CRM, system, a subset of the SuperOffice Entity types
   * This Enum is used by:  ErpConnectionActorType.ActorTypeCrm
   */
  export enum CrmActorType {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: Crm Customer
         */
             Contact = "Contact",        
        /**
         * 2: Person
         */
             Person = "Person",        
        /**
         * 3: Project
         */
             Project = "Project",        
        /**
         * 4: Sale
         */
             Sale = "Sale",        
  }


  /**
   * Should this field be published by default?
   * This Enum is used by:  Task.defaultPublishType; DocTmpl.defaultPublishType; Publish.publishType
   */
  export enum PublishType {
        /**
         * 0: 
         */
             Undefined = "Undefined",        
        /**
         * 1: Published to external persons
         */
             External = "External",        
  }


  /**
   * Justification - 0 = default, left, right, center
   * This Enum is used by:  UDefField.justification
   */
  export enum UdefJustification {
        /**
         * 0: 
         */
             Default = "Default",        
        /**
         * 1: 
         */
             Left = "Left",        
        /**
         * 2: 
         */
             Center = "Center",        
        /**
         * 3: 
         */
             Right = "Right",        
  }


  /**
   * 0 = employee, 1 = external user, 2 = anonymous
   * This Enum is used by:  Role.roleType
   */
  export enum RoleType {
        /**
         * 0: Normal employee
         */
             Employee = "Employee",        
        /**
         * 1: External User
         */
             ExternalUser = "ExternalUser",        
        /**
         * 2: Anonymous (implicit) users.
         */
             Anonymous = "Anonymous",        
        /**
         * 3: System users - sentry is bypassed for this kind of users.
         */
             System = "System",        
  }


  /**
   * Enumerator describing the main pattern of recurrence.
   * This Enum is used by:  RecurrenceRule.pattern
   */
  export enum RecurrencePattern {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: Dayly recurrence.
         */
             Daily = "Daily",        
        /**
         * 2: Weekly recurrence.
         */
             Weekly = "Weekly",        
        /**
         * 3: Monthly recurrence.
         */
             Monthly = "Monthly",        
        /**
         * 4: Yearly recurrence.
         */
             Yearly = "Yearly",        
        /**
         * 5: Custom recurrence, only dates are specified.  See RecurrenceDayPattern for details
         */
             Custom = "Custom",        
  }


  /**
   * Normal = 0, Maximized = 1, Minimized = 2
   * This Enum is used by:  WinPosSize.state
   */
  export enum ShowWindowState {
        /**
         * 0: Show Normal
         */
             Normal = "Normal",        
        /**
         * 1: Show Maximized
         */
             Maximized = "Maximized",        
        /**
         * 2: Show Minimized
         */
             Minimized = "Minimized",        
  }


  /**
   * Entire system, per database, or per associate
   * This Enum is used by:  ModuleLicense.licenseType
   */
  export enum LicenseType {
        /**
         * 0: 
         */
             Unknown = "Unknown",        
        /**
         * 1: 
         */
             SiteLicense = "SiteLicense",        
        /**
         * 2: 
         */
             SatelliteLicense = "SatelliteLicense",        
        /**
         * 3: 
         */
             UserLicense = "UserLicense",        
  }


  /**
   * What type of delivery system to use for a mail merge
   */
  export enum EMailMergeTargetType {
        /**
         * 0: Default to mail, then fax, then printer
         */
             BestFit = "BestFit",        
        /**
         * 1: Default to mail, then fax
         */
             Electronic = "Electronic",        
        /**
         * 2: Deliver by mail
         */
             Mail = "Mail",        
        /**
         * 3: Deliver by fax
         */
             Fax = "Fax",        
        /**
         * 4: Deliver to printer
         */
             Printer = "Printer",        
        /**
         * 5: Output mails as xml. This target is used in conjunction with external mail clients.
         */
             Xml = "Xml",        
        /**
         * 6: Same as Xml, but use fax format for recipient addresses
         */
             XmlFax = "XmlFax",        
  }


  /**
   * Status if this appointment is in the process of being assigned to someone else
   */
  export enum AssignmentStatus {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: 1: All appointments not in the process of being assigned to someone else
         */
             None = "None",        
        /**
         * 11: 11: Appointment in the process of being assigned to someone else
         */
             Assigning = "Assigning",        
        /**
         * 12: 12: Appointment where the party requested to take ownership of the appointment or task has observed the request, but not yet accepted or declined the request.
         */
             Seen = "Seen",        
        /**
         * 13: 13: Appointment where the person requested to take ownership has declined.
         */
             Declined = "Declined",        
  }


  /**
   * Color index of tasks
   */
  export enum ColorIndex {
        /**
         * 0: Light blue
         */
             LightBlue = "LightBlue",        
        /**
         * 1: Dark blue
         */
             DarkBlue = "DarkBlue",        
        /**
         * 2: Light gray
         */
             LightGray = "LightGray",        
        /**
         * 3: Dark gray
         */
             DarkGray = "DarkGray",        
        /**
         * 4: Light green
         */
             LightGreen = "LightGreen",        
        /**
         * 5: Dark Green
         */
             DarkGreen = "DarkGreen",        
        /**
         * 6: Light Yellow
         */
             LightYellow = "LightYellow",        
        /**
         * 7: Dark Yellow
         */
             DarkYellow = "DarkYellow",        
        /**
         * 8: Light Red
         */
             LightRed = "LightRed",        
        /**
         * 9: Dark Red
         */
             DarkRed = "DarkRed",        
        /**
         * 10: Blue alternative 1
         */
             BlueAlt1 = "BlueAlt1",        
        /**
         * 11: Blue alternative 2
         */
             BlueAlt2 = "BlueAlt2",        
        /**
         * 12: Blue alternative 3
         */
             BlueAlt3 = "BlueAlt3",        
        /**
         * 13: Gray alternative 1
         */
             GrayAlt1 = "GrayAlt1",        
        /**
         * 14: Gray alternative 2
         */
             GrayAlt2 = "GrayAlt2",        
        /**
         * 15: Gray alternative 3
         */
             GrayAlt3 = "GrayAlt3",        
        /**
         * 16: Green alternative 1
         */
             GreenAlt1 = "GreenAlt1",        
        /**
         * 17: Green alternative 2
         */
             GreenAlt2 = "GreenAlt2",        
        /**
         * 18: Green alternative 3
         */
             GreenAlt3 = "GreenAlt3",        
        /**
         * 19: Yellow alternative 1
         */
             YellowAlt1 = "YellowAlt1",        
        /**
         * 20: Yellow alternative 2
         */
             YellowAlt2 = "YellowAlt2",        
        /**
         * 21: Yellow alternative 3
         */
             YellowAlt3 = "YellowAlt3",        
        /**
         * 22: Red alternative 1
         */
             RedAlt1 = "RedAlt1",        
        /**
         * 23: Red alternative 2
         */
             RedAlt2 = "RedAlt2",        
        /**
         * 24: Red alternative 3
         */
             RedAlt3 = "RedAlt3",        
  }


  /**
   * Url encoding
   * This Enum is used by:  TaskMenu.encoding
   */
  export enum UrlEncoding {
        /**
         * 0: Unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: No encoding
         */
             None = "None",        
        /**
         * 2: ANSI
         */
             ANSI = "ANSI",        
        /**
         * 3: Unicode
         */
             Unicode = "Unicode",        
  }


  /**
   * Target type of relation. (Contact, person or both)
   */
  export enum RelationTarget {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Contact
         */
             Contact = "Contact",        
        /**
         * 2: Person
         */
             Person = "Person",        
        /**
         * 3: Both
         */
             Both = "Both",        
  }


  /**
   * Same as the EUDefType enum in the C++ client, this is the DATABASE value used for udef definitions
   */
  export enum UDefType {
        /**
         * 0: Invalid
         */
             Invalid = "Invalid",        
        /**
         * 7: Owner is the contact table
         */
             Contact = "Contact",        
        /**
         * 8: Owner is the Person table
         */
             Person = "Person",        
        /**
         * 9: Owner is the project table
         */
             Project = "Project",        
        /**
         * 10: Owner is the sale table
         */
             Sale = "Sale",        
        /**
         * 11: This is the temporary udef table pair, used during reorganizations
         */
             Temp = "Temp",        
        /**
         * 12: Owner is the appointment table
         */
             Appointment = "Appointment",        
        /**
         * 13: Owner is the document table
         */
             Document = "Document",        
        /**
         * 14: Owner is undefined, you should not see this value in a real database
         */
             None = "None",        
  }


  /**
   * Field type: 1 = Number, 2 = Short text, 3 = Long text, 4 = Date, 5 = Unlimited date, 6 = Checkbox, 7 = Drop-down listbox, 8 = Decimal
   */
  export enum UDefFieldType {
        /**
         * 1: Number (32-bit integer)
         */
             Number = "Number",        
        /**
         * 2: Short text, up to 39 characters, stored in the 'small' table
         */
             ShortText = "ShortText",        
        /**
         * 3: Long text, up to 199 characters, stored in the 'large' table
         */
             LongText = "LongText",        
        /**
         * 4: Date, between 1.1.1970 and 2038, stored as seconds after 1.1.1970
         */
             Date = "Date",        
        /**
         * 5: Unlimited date (without time), stored as an YYYYMMDD string
         */
             UnlimitedDate = "UnlimitedDate",        
        /**
         * 6: Checkbox, stored as an integer, 1 = checked
         */
             Checkbox = "Checkbox",        
        /**
         * 7: List item reference, stored as an integer ID
         */
             List = "List",        
        /**
         * 8: Decimal / floating point, stored as a double
         */
             Decimal = "Decimal",        
  }


  /**
   * Custom Field type: 1 = Number, 2 = Short text, 3 = Long text, 4 = Date, 5 = Unlimited date, 6 = Checkbox, 7 = Drop-down listbox, 8 = Decimal, 9 = DateTime, 10 = Time, 11 = TimeSpan, 12 = Relation, 13 = Attachment
   */
  export enum CustomFieldType {
        /**
         * 0: Unknown/Undefined field type: 0
         */
             Unknown = "Unknown",        
        /**
         * 1: Number (32-bit integer): '[I:123]'
         */
             Integer = "Integer",        
        /**
         * 2: Short text, up to 255 characters: 'text value'
         */
             ShortText = "ShortText",        
        /**
         * 3: Long text: 'long text value'
         */
             LongText = "LongText",        
        /**
         * 4: Date: '[D:06/19/2019]'
         */
             Date = "Date",        
        /**
         * 5: Binary blob, encoded as Base64 string.
         */
             Blob = "Blob",        
        /**
         * 6: Boolean, stored as an integer, 1 = checked/true, 0 = unchecked/false: '[I:1]'
         */
             Checkbox = "Checkbox",        
        /**
         * 7: MDO List item reference, stored as an integer ID: '[I:123]'
         */
             MdoList = "MdoList",        
        /**
         * 8: Decimal / floating point, stored as a double: '[F:3.14159]'
         */
             Decimal = "Decimal",        
        /**
         * 9: Date + Time:  '[DT:06/19/2019 14:35:56.00000]'
         */
             DateTime = "DateTime",        
        /**
         * 10: Time only - ignore the date part: '[T:06/19/2019 14:35:56.00000]'
         */
             Time = "Time",        
        /**
         * 11: Timespan in some units (hours, minutes or seconds): '[I:1234]'
         */
             TimeSpan = "TimeSpan",        
        /**
         * 12: Relation to another table. Stored as an int (primary key of record in other table): '[I:1234]'
         */
             RelationTo = "RelationTo",        
        /**
         * 13: File Attachment: int record id in attachment table: '[I:1234]'
         */
             Attachment = "Attachment",        
        /**
         * 14: Dynamic Link: contains link redirect value: 'https://example.com/scripts/ticket.fcgi?action=redirectLink&extraFieldId=5&table=y_equipment&id=2'
         */
             DynamicLink = "DynamicLink",        
        /**
         * 15: Text with list items to choose between: 'text value'
         */
             ListText = "ListText",        
  }


  /**
   * Value for field 'scope' in table 'systemevent'.
   */
  export enum SystemEventScope {
        /**
         * 0: Undefined value (error)
         */
             Undefined = "Undefined",        
        /**
         * 1: Value valid for all databases under this licence
         */
             SystemWide = "SystemWide",        
        /**
         * 2: Value valid for this database
         */
             Database = "Database",        
        /**
         * 3: Value valid for this group
         */
             Group = "Group",        
        /**
         * 4: Value valid for this user
         */
             User = "User",        
  }


  /**
   * Entitys type for import
   * This Enum consists of flag values that can be combined.
   */
  export enum ImportEntityType {
        /**
         * 0: Unknown
         */
             Unknown = 0,        
        /**
         * 1: Person
         */
             Person = 1,        
        /**
         * 2: Contact
         */
             Contact = 2,        
        /**
         * 4: Product
         */
             Product = 4,        
  }


  /**
   * Action being done for the import row
   * This Enum consists of flag values that can be combined.
   */
  export enum ImportAction {
        /**
         * 0: Unknown
         */
             Unknown = 0,        
        /**
         * 1: PersonAdded
         */
             PersonAdded = 1,        
        /**
         * 2: PersonUpdated
         */
             PersonUpdated = 2,        
        /**
         * 4: PersonNoChange
         */
             PersonNoChange = 4,        
        /**
         * 8: ContactAdded
         */
             ContactAdded = 8,        
        /**
         * 16: ContactUpdated
         */
             ContactUpdated = 16,        
        /**
         * 32: ContactNoChange
         */
             ContactNoChange = 32,        
        /**
         * 64: ProductAdded
         */
             ProductAdded = 64,        
        /**
         * 128: ProductUpdated
         */
             ProductUpdated = 128,        
        /**
         * 256: ProductNoChange
         */
             ProductNoChange = 256,        
        /**
         * 512: Obs
         */
             Obs = 512,        
        /**
         * 1024: ObsERPDuplicate
         */
             ObsERPDuplicate = 1024,        
  }


  /**
   * Freetext operator
   */
  export enum FreeTextOperator {
        /**
         * 1: Contains
         */
             Contains = "Contains",        
        /**
         * 2: Starts with
         */
             StartsWith = "StartsWith",        
        /**
         * 3: Exact match
         */
             ExactMatch = "ExactMatch",        
  }


  /**
   * Response codes used by ErpSync services
   */
  export enum ErpSyncResponseCode {
        /**
         * 0: No error
         */
             NoError = "NoError",        
        /**
         * 1: Could not perform operation because connector has connections
         */
             ErrorConnectorHasConnections = "ErrorConnectorHasConnections",        
        /**
         * 2: Could not perform operation because the entity wasn't found
         */
             ErrorNotFound = "ErrorNotFound",        
  }


  /**
   * Status of a ticket/request
   * This Enum is used by:  ticket.status; ticket_status.status
   */
  export enum TicketBaseStatus {
        /**
         * 0: Unknown / uninitialized
         */
             Unknown = "Unknown",        
        /**
         * 1: Request is currently active
         */
             Active = "Active",        
        /**
         * 2: Request has been closed
         */
             Closed = "Closed",        
        /**
         * 3: Request has been postponed
         */
             Postponed = "Postponed",        
        /**
         * 4: Request has been deleted
         */
             Deleted = "Deleted",        
        /**
         * 5: Request has been merged with another request
         */
             Merged = "Merged",        
        /**
         * 1001: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             PostponedSpecific = "PostponedSpecific",        
        /**
         * 1002: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             Postponed1Hour = "Postponed1Hour",        
        /**
         * 1003: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             Postponed2Hours = "Postponed2Hours",        
        /**
         * 1004: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             Postponed3Hours = "Postponed3Hours",        
        /**
         * 1005: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             Postponed4Hours = "Postponed4Hours",        
        /**
         * 1006: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             PostponedDay = "PostponedDay",        
        /**
         * 1007: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             PostponedWeek = "PostponedWeek",        
        /**
         * 1008: Should only be used for UI, when saved to DB it will be converted to Postponed
         */
             PostponedMonth = "PostponedMonth",        
  }


  /**
   * The avaliable category
   */
  export enum ReportCategory {
        /**
         * 0: New/empty report
         */
             None = "None",        
        /**
         * 1: All reports
         */
             All = "All",        
        /**
         * 2: 2: Company reports
         */
             Contact = "Contact",        
        /**
         * 3: 3: Project reports
         */
             Project = "Project",        
        /**
         * 4: 4: Sale reports
         */
             Sale = "Sale",        
        /**
         * 5: 5: Appointment reports
         */
             Appointment = "Appointment",        
        /**
         * 6: 6: Selection reports
         */
             Selection = "Selection",        
        /**
         * 7: 7: Person reports
         */
             Person = "Person",        
        /**
         * 8: 8: Diary reports
         */
             Diary = "Diary",        
        /**
         * 9: Favorite reports for an associate
         */
             Favorites = "Favorites",        
  }


  /**
   * 
   * This Enum is used by:  email_item.priority
   */
  export enum EMailPriority {
        /**
         * 0: 
         */
             NoPriority = "NoPriority",        
        /**
         * 1: 
         */
             Highest = "Highest",        
        /**
         * 2: 
         */
             High = "High",        
        /**
         * 3: 
         */
             Normal = "Normal",        
        /**
         * 4: 
         */
             Low = "Low",        
        /**
         * 5: 
         */
             Lowest = "Lowest",        
  }


  /**
   * 
   * This Enum is used by:  email_item.flags
   * This Enum consists of flag values that can be combined.
   */
  export enum EMailFlags {
        /**
         * 0: 0: No flags set
         */
             None = 0,        
        /**
         * 1: 1: Mail has been read
         */
             Seen = 1,        
        /**
         * 2: 2: Mail has been deleted
         */
             Deleted = 2,        
        /**
         * 4: 4: Mail is recent
         */
             Recent = 4,        
        /**
         * 8: 8: Mail is flagged in the GUI
         */
             Flagged = 8,        
        /**
         * 16: 16: Mail is an unsent draft
         */
             Draft = 16,        
        /**
         * 32: 32: Mail has been replied to at least once
         */
             Answered = 32,        
  }


  /**
   * 
   */
  export enum BatchTaskState {
        /**
         * 0: Unknown state of the task.
         */
             Unknown = "Unknown",        
        /**
         * 1: The task is new, but not assigned or run by any party.
         */
             New = "New",        
        /**
         * 2: The task has been aquired by an instance that wants to execute the task, but has not been started.
         */
             Aquired = "Aquired",        
        /**
         * 3: The task is executing.
         */
             Started = "Started",        
        /**
         * 4: Executing the task has completed successfully.
         */
             Succeeded = "Succeeded",        
        /**
         * 5: Executing the task has halted and the task failed.   
         */
             Failed = "Failed",        
        /**
         * 6: Executing the task has completed successfully, and related data will be cleaned up by the responsible code.
         */
             SucceededManualCleanup = "SucceededManualCleanup",        
        /**
         * 99999: Special type. 
         */
             All = "All",        
  }


  /**
   * Batch task cancel support
   */
  export enum BatchTaskCancellationBehaviour {
        /**
         * 0: The batch task can be cancelled
         */
             CanCancel = "CanCancel",        
        /**
         * 1: The batch task can be cancelled, but the UI will display a warning (symbol SR_BATCH_<snakeCaseName>_CANCELLATION
         */
             CancelWithWarning = "CancelWithWarning",        
        /**
         * 2: The batch task cannot be cancelled
         */
             CannotCancel = "CannotCancel",        
  }


  /**
   * Selection type - static/dynamic/combined. Works together with targetTable to define what kind of Selection this is.
   */
  export enum SelectionType {
        /**
         * 0: Static selection
         */
             Static = "Static",        
        /**
         * 1: Dynamic selection
         */
             Dynamic = "Dynamic",        
        /**
         * 2: Combined selection
         */
             Combined = "Combined",        
  }


  /**
   * 
   */
  export enum ReportLayout {
        /**
         * 0: 
         */
             Unknown = "Unknown",        
        /**
         * 1: 
         */
             List = "List",        
        /**
         * 2: 
         */
             Label = "Label",        
        /**
         * 3: 
         */
             GroupList = "GroupList",        
        /**
         * 4: 
         */
             CrossTable = "CrossTable",        
        /**
         * 5: 
         */
             CalendarWeek5 = "CalendarWeek5",        
        /**
         * 6: 
         */
             CalendarWeek7 = "CalendarWeek7",        
        /**
         * 7: 
         */
             CalendarMonth = "CalendarMonth",        
        /**
         * 8: 
         */
             Text = "Text",        
  }


  /**
   * Describes the orientation of the paper when printing a report. The report layout must be "Label"
   */
  export enum ReportPaperOrientation {
        /**
         * 0: 
         */
             None = "None",        
        /**
         * 1: The taller orientation
         */
             Portrait = "Portrait",        
        /**
         * 2: The wider orientation
         */
             Landscape = "Landscape",        
  }


  /**
   * Locking semantics requested/applied to a document
   */
  export enum DocumentLockSemantics {
        /**
         * 0: Neither locking nor versioning is requested for the document
         */
             None = "None",        
        /**
         * 1: Locking, but not versioning, is requested for the document
         */
             Locking = "Locking",        
        /**
         * 2: Versioning is requested for the document (locking is implied)
         */
             Versioning = "Versioning",        
  }


  /**
   * 
   */
  export enum GeneratorEncoding {
        /**
         * 0: Plain text document.
         */
             Text = "Text",        
        /**
         * 1: Html encoded document, it is not strictly parsed.
         */
             Html = "Html",        
        /**
         * 2: XML Document, it is not strictly parsed
         */
             Xml = "Xml",        
        /**
         * 3: Microsoft word document.
         */
             MsWord = "MsWord",        
        /**
         * 4: Microsoft Excel document.
         */
             MsExcel = "MsExcel",        
        /**
         * 5: Microsoft Powerpoint document
         */
             MsPowerpoint = "MsPowerpoint",        
        /**
         * 6: Microsoft office documents zipped (Office 2007)
         */
             MsOffice2007 = "MsOffice2007",        
        /**
         * 7: Microsoft office xml documents (content of zipped x files)
         */
             MsOffice2007Xml = "MsOffice2007Xml",        
        /**
         * 8: The tags result are url-encoded (ansi), the source itself is treated as plain text.
         */
             Url = "Url",        
        /**
         * 9: The tags result are url-encoded (unicode), the source itself is treated as plain text.
         */
             UrlUnicode = "UrlUnicode",        
        /**
         * 10: Adobe pdf format
         */
             Pdf = "Pdf",        
        /**
         * 11: RFC822 Mime-content. Handles the actual mime part, but leaves a normal Text-generator to handle the actual content embedded within the MIME-layers.
         */
             Mime = "Mime",        
        /**
         * 12: OpenOffice/LibreOffice documents (zipped)
         */
             OpenDocument = "OpenDocument",        
        /**
         * 13: OpenOffice/LibreOffice documents (contents of zipped files)
         */
             OpenDocumentXml = "OpenDocumentXml",        
  }


  /**
   * Dashboard layout
   */
  export enum DashboardLayout {
        /**
         * 0: Unknown layout
         */
             None = "None",        
        /**
         * 1: One tile.
         */
             One = "One",        
        /**
         * 2: Two vertical split.
         */
             TwoVerticalSplit = "TwoVerticalSplit",        
        /**
         * 3: Three E split.
         */
             ThreeESplit = "ThreeESplit",        
        /**
         * 4: Two horizontal split.
         */
             TwoHorizontalSplit = "TwoHorizontalSplit",        
        /**
         * 5: Three T split.
         */
             ThreeTSplit = "ThreeTSplit",        
        /**
         * 7: Four tiles.
         */
             Four = "Four",        
  }


  /**
   * Dashboard tile type
   */
  export enum DashboardTileType {
        /**
         * 0: Unknown layout
         */
             None = "None",        
        /**
         * 1: Chart tile
         */
             Chart = "Chart",        
        /**
         * 2: Web tile
         */
             Web = "Web",        
        /**
         * 3: List tile
         */
             List = "List",        
        /**
         * 4: Number tile
         */
             Bignum = "Bignum",        
  }


  /**
   * Dashboard tile entity type
   */
  export enum DashboardTileEntityType {
        /**
         * 0: Unknown entity type
         */
             None = "None",        
        /**
         * 1: 
         */
             Company = "Company",        
        /**
         * 2: 
         */
             Project = "Project",        
        /**
         * 3: 
         */
             Sale = "Sale",        
        /**
         * 4: 
         */
             Product = "Product",        
        /**
         * 5: 
         */
             Activity = "Activity",        
        /**
         * 6: 
         */
             Document = "Document",        
        /**
         * 7: 
         */
             WebPanel = "WebPanel",        
        /**
         * 8: 
         */
             Followup = "Followup",        
  }


  /**
   * Dashboard tile option type
   */
  export enum DashboardTileOptionType {
        /**
         * 0: Unknown type
         */
             None = "None",        
        /**
         * 1: String
         */
             String = "String",        
        /**
         * 2: Integer
         */
             Integer = "Integer",        
        /**
         * 3: Boolean
         */
             Boolean = "Boolean",        
        /**
         * 4: List
         */
             List = "List",        
  }


  /**
   * Dashboard tile chart type (V2)
   * This Enum is used by:  dashboard_tile_definition.tile_type
   */
  export enum DashTileType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Pie
         */
             Pie = "Pie",        
        /**
         * 2: List
         */
             List = "List",        
        /**
         * 3: BigNum
         */
             BigNum = "BigNum",        
        /**
         * 4: Bar
         */
             Bar = "Bar",        
        /**
         * 5: Line
         */
             Line = "Line",        
        /**
         * 6: Area
         */
             Area = "Area",        
        /**
         * 7: Column
         */
             Column = "Column",        
        /**
         * 8: CombinedBarLine
         */
             CombinedBarLine = "CombinedBarLine",        
        /**
         * 9: CombinedColumnLine
         */
             CombinedColumnLine = "CombinedColumnLine",        
        /**
         * 10: WebPanel
         */
             WebPanel = "WebPanel",        
        /**
         * 11: HTML
         */
             HTML = "HTML",        
        /**
         * 12: Gauge
         */
             Gauge = "Gauge",        
  }


  /**
   * Dashboard tile currency mode
   * This Enum is used by:  dashboard_tile_definition.currency_mode
   */
  export enum DashTileCurrencyMode {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Base
         */
             Base = "Base",        
        /**
         * 2: Own
         */
             Own = "Own",        
        /**
         * 3: Specified
         */
             Specified = "Specified",        
  }


  /**
   * Dashboard tile measure
   * This Enum is used by:  dashboard_tile_definition.measure
   */
  export enum DashTileMeasure {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: CountAll
         */
             CountAll = "CountAll",        
        /**
         * 2: Sum
         */
             Sum = "Sum",        
        /**
         * 3: Average
         */
             Average = "Average",        
        /**
         * 4: Max
         */
             Max = "Max",        
        /**
         * 5: Min
         */
             Min = "Min",        
        /**
         * 6: Count
         */
             Count = "Count",        
  }


  /**
   * Dashboard tile entity type (V2)
   * This Enum is used by:  dashboard_tile_definition.entity_type
   */
  export enum DashTileEntityType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Contact
         */
             Contact = "Contact",        
        /**
         * 2: Sale
         */
             Sale = "Sale",        
        /**
         * 3: Project
         */
             Project = "Project",        
        /**
         * 4: Appointment
         */
             Appointment = "Appointment",        
        /**
         * 5: Person
         */
             Person = "Person",        
  }


  /**
   * Defines where a DashTileDefinition can be used
   * This Enum is used by:  dashboard_tile_definition.usage
   * This Enum consists of flag values that can be combined.
   */
  export enum DashTileUsage {
        /**
         * 0: Not visible anywhere
         */
             None = 0,        
        /**
         * 1: Visible in the tile library on the dashboard page
         */
             Dashboard = 1,        
        /**
         * 2: Visible in the tile library on the selection page
         */
             Selection = 2,        
  }


  /**
   * Target assignement levels
   * This Enum is used by:  target_dimension.assignment_level
   * This Enum consists of flag values that can be combined.
   */
  export enum TargetAssignementLevel {
        /**
         * 0: No level
         */
             None = 0,        
        /**
         * 1: Company
         */
             Company = 1,        
        /**
         * 2: Group
         */
             Group = 2,        
        /**
         * 4: Associate
         */
             Associate = 4,        
  }


  /**
   * Target entity type
   * This Enum is used by:  target_group.entity_type; target_revision_history.entity_type; target_dimension.entity_type
   */
  export enum TargetEntityType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Sale
         */
             Sale = "Sale",        
  }


  /**
   * Target measurement unit
   * This Enum is used by:  target_dimension.measurement_unit
   */
  export enum TargetMeasurementUnit {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Amount
         */
             Amount = "Amount",        
        /**
         * 2: Count
         */
             Count = "Count",        
        /**
         * 3: Profit
         */
             Profit = "Profit",        
  }


  /**
   * Target period type
   * This Enum is used by:  target_group.period_type
   */
  export enum TargetPeriodType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Year
         */
             Year = "Year",        
        /**
         * 2: HalfYear
         */
             HalfYear = "HalfYear",        
        /**
         * 3: Quarter
         */
             Quarter = "Quarter",        
        /**
         * 4: Month
         */
             Month = "Month",        
  }


  /**
   * Target level
   */
  export enum TargetLevel {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Global
         */
             Global = "Global",        
        /**
         * 2: Company
         */
             Company = "Company",        
        /**
         * 3: UserGroup
         */
             UserGroup = "UserGroup",        
        /**
         * 4: Associate
         */
             Associate = "Associate",        
  }


  /**
   * Log events
   * This Enum is used by:  target_revision.log_event
   */
  export enum LogEvent {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Create
         */
             Create = "Create",        
        /**
         * 2: Edit
         */
             Edit = "Edit",        
        /**
         * 3: Delete
         */
             Delete = "Delete",        
        /**
         * 4: Lock
         */
             Lock = "Lock",        
        /**
         * 5: Unlock
         */
             Unlock = "Unlock",        
  }


  /**
   * Workflow definition status
   * This Enum is used by:  workflow.status
   */
  export enum WorkflowDefinitionStatus {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Stopped
         */
             Stopped = "Stopped",        
        /**
         * 2: Paused
         */
             Paused = "Paused",        
        /**
         * 3: Running
         */
             Running = "Running",        
  }


  /**
   * Email/Mailing From field address algorithm
   * This Enum is used by:  email_flow.from_type
   */
  export enum EmailFromType {
        /**
         * 0: FromOnlySpecified
         */
             FromOnlySpecified = "FromOnlySpecified",        
        /**
         * 1: FromSalesContact
         */
             FromSalesContact = "FromSalesContact",        
        /**
         * 2: FromSupportContact
         */
             FromSupportContact = "FromSupportContact",        
  }


  /**
   * Email/Mailing ReplyTo field address algorithm
   * This Enum is used by:  email_flow.reply_to_type
   */
  export enum EmailReplyToType {
        /**
         * 0: ReplyToOnlySpecified
         */
             ReplyToOnlySpecified = "ReplyToOnlySpecified",        
        /**
         * 1: ReplyToSalesContact
         */
             ReplyToSalesContact = "ReplyToSalesContact",        
        /**
         * 2: ReplyToSupportContact
         */
             ReplyToSupportContact = "ReplyToSupportContact",        
        /**
         * 3: ReplyToEmpty (skip reply-to header, email agents will use from instead)
         */
             ReplyToEmpty = "ReplyToEmpty",        
  }


  /**
   * Workflow waiting step wait algorithm
   */
  export enum WorkflowTimeWaitAlgorithm {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Number of intervals (min, hour, workdays, days, weeks...)
         */
             NumIntervals = "NumIntervals",        
        /**
         * 2: Until specified datetime
         */
             UntilSpecified = "UntilSpecified",        
        /**
         * 3: Until date on person/company (a persons birthdate etc)
         */
             UntilDateField = "UntilDateField",        
  }


  /**
   * Workflow waiting time interval type
   */
  export enum WorkflowTimeWaitIntervalType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Seconds
         */
             Seconds = "Seconds",        
        /**
         * 2: Minutes
         */
             Minutes = "Minutes",        
        /**
         * 3: Hours
         */
             Hours = "Hours",        
        /**
         * 4: Working days
         */
             WorkingDays = "WorkingDays",        
        /**
         * 5: Days
         */
             Days = "Days",        
        /**
         * 6: Weeks
         */
             Weeks = "Weeks",        
  }


  /**
   * Workflow Goal Type
   * This Enum is used by:  workflow_goal.goal_type
   */
  export enum WorkflowGoalType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: ContactUpdated
         */
             ContactUpdated = "ContactUpdated",        
        /**
         * 2: SaleCreatedOnContact
         */
             SaleCreatedOnContact = "SaleCreatedOnContact",        
        /**
         * 3: SaleCreatedOnCompany
         */
             SaleCreatedOnCompany = "SaleCreatedOnCompany",        
        /**
         * 4: AppointmentCreated
         */
             AppointmentCreated = "AppointmentCreated",        
        /**
         * 5: AddedToProject
         */
             AddedToProject = "AddedToProject",        
        /**
         * 6: AddedToSelection
         */
             AddedToSelection = "AddedToSelection",        
        /**
         * 7: FormSubmitted
         */
             FormSubmitted = "FormSubmitted",        
        /**
         * 8: HadChat
         */
             HadChat = "HadChat",        
        /**
         * 9: LinkClicked
         */
             LinkClicked = "LinkClicked",        
        /**
         * 10: RequestCreated
         */
             RequestCreated = "RequestCreated",        
  }


  /**
   * Workflow Step Type
   * This Enum is used by:  workflow_step.step_type
   */
  export enum WorkflowStepType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: SendEmail
         */
             SendEmail = "SendEmail",        
        /**
         * 2: SendSMS
         */
             SendSMS = "SendSMS",        
        /**
         * 3: WaitForTime
         */
             WaitForTime = "WaitForTime",        
        /**
         * 4: WaitForAction
         */
             WaitForAction = "WaitForAction",        
        /**
         * 5: Split
         */
             Split = "Split",        
        /**
         * 6: UpdateContact
         */
             UpdateContact = "UpdateContact",        
        /**
         * 7: AddToList
         */
             AddToList = "AddToList",        
        /**
         * 8: RemoveFromList
         */
             RemoveFromList = "RemoveFromList",        
        /**
         * 9: CreateRequest
         */
             CreateRequest = "CreateRequest",        
        /**
         * 10: CreateFollowUp
         */
             CreateFollowUp = "CreateFollowUp",        
        /**
         * 11: CreateSale
         */
             CreateSale = "CreateSale",        
        /**
         * 12: NotifyByEmail
         */
             NotifyByEmail = "NotifyByEmail",        
        /**
         * 13: NotifyBySMS
         */
             NotifyBySMS = "NotifyBySMS",        
        /**
         * 14: RunScript
         */
             RunScript = "RunScript",        
  }


  /**
   * Workflow Trigger Type
   * This Enum is used by:  workflow_trigger.trigger_type
   */
  export enum WorkflowTriggerType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: ContactCreated
         */
             SendEmail = "SendEmail",        
        /**
         * 2: ContactUpdated
         */
             SendSMS = "SendSMS",        
        /**
         * 3: AddedToProject
         */
             WaitForTime = "WaitForTime",        
        /**
         * 4: CreatedRequest
         */
             WaitForAction = "WaitForAction",        
        /**
         * 5: LinkClicked
         */
             Split = "Split",        
        /**
         * 6: AddedToSelection
         */
             UpdateContact = "UpdateContact",        
        /**
         * 7: FormSubmitted
         */
             AddToList = "AddToList",        
        /**
         * 8: AddedMessage
         */
             RemoveFromList = "RemoveFromList",        
        /**
         * 9: StartedChat
         */
             CreateRequest = "CreateRequest",        
        /**
         * 10: OrderPlaced
         */
             CreateFollowUp = "CreateFollowUp",        
        /**
         * 11: SaleCreated
         */
             CreateSale = "CreateSale",        
        /**
         * 12: SaleSold
         */
             NotifyByEmail = "NotifyByEmail",        
        /**
         * 13: SaleLost
         */
             NotifyBySMS = "NotifyBySMS",        
  }


  /**
   * Completed status enum common to appointments, documents and sales
   */
  export enum ActivityStatus {
        /**
         * 0: 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: 1: Appointment has not been started
         */
             NotStarted = "NotStarted",        
        /**
         * 2: 2: Appointment has been started, but not completed
         */
             Started = "Started",        
        /**
         * 3: 3: Appointment is completed
         */
             Completed = "Completed",        
  }


  /**
   * Status if this appointment represents an invitation
   */
  export enum InvitationStatus {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: 1: Default value when the appointment is not a booking child
         */
             None = "None",        
        /**
         * 2: 2: The invitation has been accepted
         */
             Accepted = "Accepted",        
        /**
         * 4: 4: Hidden
         */
             Hidden = "Hidden",        
        /**
         * 5: 5: An invitation to a meeting not observed by the attending party
         */
             Invitation = "Invitation",        
        /**
         * 6: 6: The meeting has been moved
         */
             Moved = "Moved",        
        /**
         * 7: 7: An invitation to a meeting observed by the attending party. This can also be viewed upon as a tentative response to participate in the meeting.
         */
             Seen = "Seen",        
        /**
         * 8: 8: The meeting has been moved, and this is observed by the attending party
         */
             MovedSeen = "MovedSeen",        
        /**
         * 9: 9: The invitation has been declined/rejected
         */
             Declined = "Declined",        
        /**
         * 10: 10: The meeting has been cancelled.
         */
             Cancelled = "Cancelled",        
  }


  /**
   * The type of booking the appointment represents
   */
  export enum BookingType {
        /**
         * 0: 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: 1: Single appointment records
         */
             None = "None",        
        /**
         * 2: 2: Owner of a booking
         */
             Owner = "Owner",        
        /**
         * 3: 3: Participant of a booking
         */
             Participant = "Participant",        
  }


  /**
   * The type of the locale text strings
   */
  export enum LocalizedTextType {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = "Unknown",        
        /**
         * 1: 1 = fieldlabels 
         */
             Label = "Label",        
        /**
         * 2: 
         */
             Table = "Table",        
        /**
         * 3: 
         */
             Column = "Column",        
        /**
         * 4: 4 - import system field names 
         */
             ImportField = "ImportField",        
        /**
         * 5: 5 - import system table names
         */
             ImportObject = "ImportObject",        
        /**
         * 6: 6 - import system table groups
         */
             ImportGUICategory = "ImportGUICategory",        
        /**
         * 7: 7  (languageid > 0)
         */
             StartupHints = "StartupHints",        
        /**
         * 11: 11 (languageid > 0)
         */
             FunctionRightLabel = "FunctionRightLabel",        
        /**
         * 12: 12 (languageid > 0)
         */
             FunctionRightDesc = "FunctionRightDesc",        
        /**
         * 16: languageid = 0
         */
             StatusMonitorName = "StatusMonitorName",        
        /**
         * 19: 19  (languageid = 0)
         */
             UdefContactLabel = "UdefContactLabel",        
        /**
         * 20: 20  (languageid = 0)
         */
             UdefPersonLabel = "UdefPersonLabel",        
        /**
         * 21: 21  (languageid = 0)
         */
             UdefProjectLabel = "UdefProjectLabel",        
        /**
         * 22: 22 (languageid = 0) 
         */
             UdefSaleLabel = "UdefSaleLabel",        
        /**
         * 23: 23 (languageid = 0)
         */
             UdefAppointmentLabel = "UdefAppointmentLabel",        
        /**
         * 24: 24 (languageid = 0)
         */
             UdefDocumentLabel = "UdefDocumentLabel",        
        /**
         * 41: 41 - push notification tex
         */
             PushNotificationText = "PushNotificationText",        
  }


  /**
   * Describes what type of activity/action there has been on a contact
   * This Enum consists of flag values that can be combined.
   */
  export enum ContactAction {
        /**
         * 0: Unknown/undecided type
         */
             Unknown = 0,        
        /**
         * 1: The contact was created
         */
             Created = 1,        
        /**
         * 2: The contact was updated
         */
             Updated = 2,        
        /**
         * 4: A new activity was created on the contact
         */
             NewActivity = 4,        
        /**
         * 8: An activity belonging to the contact was completed
         */
             ActivityCompleted = 8,        
        /**
         * 16: A person was added to the contact
         */
             PersonAdded = 16,        
        /**
         * 32: A person belonging to the contact was updated
         */
             PersonUpdated = 32,        
        /**
         * 64: A document was added to the contact
         */
             DocumentAdded = 64,        
        /**
         * 127: Enumeration generated from the wsdl doesn't handle the [Flags] attribute (bit enumeration)
         */
             All = 127,        
        /**
         * 128: Undefined and generated value 128
         */
             UndefinedValue128 = 128,        
  }


  /**
   * Enum listing the possible checkout states of a document, as seen from outside the document plugin
   */
  export enum CheckoutState {
        /**
         * 0: Document is not checked out
         */
             NotCheckedOut = "NotCheckedOut",        
        /**
         * 1: Document is checked out, to the user making the query
         */
             CheckedOutOwn = "CheckedOutOwn",        
        /**
         * 2: Document is checked out, to a user different from the one making the query
         */
             CheckedOutOther = "CheckedOutOther",        
        /**
         * 3: Locking is not supported or selected for this document. The document is always 'writeable' in the sense of not being protected by any locks. The document may still be read-only due to Sentry constraints.
         */
             LockingNotSupported = "LockingNotSupported",        
  }


  /**
   * Enumeration that exposes what search types that are supported.
   */
  export enum StringSearchType {
        /**
         * 0: The Exact element searches for an exact string in your searchable content.
         */
             Exact = "Exact",        
        /**
         * 1: The BeginsWith element searches for a string at the start of your searchable content.
         */
             BeginsWith = "BeginsWith",        
        /**
         * 2: The EndsWith element searches for a string at the end of your searchable content.
         */
             EndsWith = "EndsWith",        
        /**
         * 3: The Exact element searches for a string in searchable content.
         */
             Contains = "Contains",        
  }


  /**
   * Type of responce
   */
  export enum ReturnType {
        /**
         * 0: here is no return value to be parsed
         */
             None = "None",        
        /**
         * 1: The return value is a string, to be shown to the user as a message by client
         */
             Message = "Message",        
        /**
         * 2: The return value is an soprotocol-compliant navigation string, to be parsed according to the rules for soprotocol strings. Client should navigate according to the given protocol.
         */
             SoProtocol = "SoProtocol",        
        /**
         * 3: The return value is the name of a custom GUI. The GetCustomGui method on the document plugin will be invoked, and the resulting GUI rendered by the client.
         */
             CustomGui = "CustomGui",        
        /**
         * 4: The return value is intended for custom processing; this implies that the document plugin has installed some component on the intended client that will intercept and handle this return value.  Note that this requires advanced knowledge of the client and its construction and navigation. Also note that this will make the functionality client-specific, so that it may not be available or work as intended on clients not specifically supported or known.
         */
             Other = "Other",        
        /**
         * 5: The return value is a URL that should be opened in a new window by the client.
         */
             URL = "URL",        
  }


  /**
   * Table right is a combination of bits representing permissions on a row.
   * This Enum consists of flag values that can be combined.
   */
  export enum ETableRight {
        /**
         * 0: The row should be hidden from the user.
         */
             None = 0,        
        /**
         * 1: The user is allowed to see the row.
         */
             Select = 1,        
        /**
         * 2: The user is allowed to modify the values in this row.
         */
             Update = 2,        
        /**
         * 3: Select or Update
         */
             UR = 3,        
        /**
         * 4: The user is allowed to add new rows to the table.
         */
             Insert = 4,        
        /**
         * 5: Select and Insert (Anonymous)
         */
             RI = 5,        
        /**
         * 7: Select, Update, Insert
         */
             URI = 7,        
        /**
         * 8: The user is allowed to delete this row.
         */
             Delete = 8,        
        /**
         * 11: Select, Update, Delete
         */
             UDR = 11,        
        /**
         * 15: Select, Update, Insert, Delete - no restrictions
         */
             FULL = 15,        
        /**
         * 16: A flag used to indicate that some fields may be hidden.
         */
             Filtering = 16,        
        /**
         * 17: Select or Filter
         */
             RF = 17,        
        /**
         * 20: Filtering and Insert (Anonymous)
         */
             FI = 20,        
        /**
         * 32: A hint flag used to indicate that some fields may not be updateable, even though the user has update access to the row.
         */
             RestrictedUpdate = 32,        
        /**
         * 64: Unused value only present for serializatrion purposes
         */
             Unused1 = 64,        
        /**
         * 128: Magic value to indicate not
         */
             Uninitialized = 128,        
        /**
         * 1: Select only
         */
             R = 1,        
        /**
         * 16: just Filtering
         */
             F = 16,        
  }


  /**
   * Used in the CheckLicenseStatusResult to describe how SOADMIN should handle
   */
  export enum LicenseStatus {
        /**
         * 0: License is ok. No changes detected
         */
             Ok = "Ok",        
        /**
         * 1: New license information available. Get new license and activate it.
         */
             NewLicenseAvailable = "NewLicenseAvailable",        
        /**
         * 2: Name change detected. Fetch new license and activate it.
         */
             NewCompanyNameAvailable = "NewCompanyNameAvailable",        
        /**
         * 3: Serial number change detected. Fetch new license and activate it.
         */
             NewSerialAvailable = "NewSerialAvailable",        
        /**
         * 4: See message for details.
         */
             UseCustomMessage = "UseCustomMessage",        
        /**
         * 5: See message and URL for details. Client should make URL clickable.
         */
             UseCustomMessageAndUrl = "UseCustomMessageAndUrl",        
        /**
         * 100: Something seems to be wrong. Cannot fetch new license.
         */
             ProblemWithLicense = "ProblemWithLicense",        
        /**
         * 5000: Bad things are happening.
         */
             UnknownError = "UnknownError",        
  }


  /**
   * Appointment recurrence change mode: only this, this and forward, stop
   */
  export enum RecurrenceUpdateMode {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: Update only this instance
         */
             OnlyThis = "OnlyThis",        
        /**
         * 2: Update this instance and all instances forward in time
         */
             ThisAndForward = "ThisAndForward",        
        /**
         * 9: Stop the recurrence from this instance and forward in time
         */
             StopRecurrence = "StopRecurrence",        
  }


  /**
   * Type of user (or resource). Mapps functionally to associate type - but with different binary values.
   */
  export enum UserType {
        /**
         * 0: 
         */
             Unknown = "Unknown",        
        /**
         * 1: Employee
         */
             InternalAssociate = "InternalAssociate",        
        /**
         * 2: Resource - meeting room, projector or other bookable resource
         */
             ResourceAssociate = "ResourceAssociate",        
        /**
         * 3: External person
         */
             ExternalAssociate = "ExternalAssociate",        
        /**
         * 4: Anonymous user - used by applications that does not require logins
         */
             AnonymousAssociate = "AnonymousAssociate",        
        /**
         * 5: System user - mainly used for integrations or administration.  This user is not goverened sentry
         */
             SystemAssociate = "SystemAssociate",        
  }


  /**
   * Describes what source the contacts should be retrieved from.
   * This Enum consists of flag values that can be combined.
   */
  export enum ContactSourceType {
        /**
         * 0: Unknown/undecided type
         */
             Unknown = 0,        
        /**
         * 1: Contact is retrieved from the history list
         */
             History = 1,        
        /**
         * 2: Contact is retrieved from the diary
         */
             Diary = 2,        
        /**
         * 4: Contact is retrieved from the favorites list
         */
             Favorites = 4,        
        /**
         * 7: Enumeration generated from the wsdl doesn't handle the [Flags] attribute (bit enumeration)
         */
             All = 7,        
        /**
         * 8: Undefined and generated value 8
         */
             UndefinedValue8 = 8,        
  }


  /**
   * The visibility of the record
   */
  export enum Visibility {
        /**
         * 0: 0 = Visible for all users
         */
             All = "All",        
        /**
         * 1: 1 = The visibility is set for an associate
         */
             Associate = "Associate",        
        /**
         * 2: 2 = The visibility is set for a group
         */
             Group = "Group",        
  }


  /**
   * Types of control used with user administration work with credentials
   */
  export enum CredentialControlType {
        /**
         * 0: The value is static and can not be modified by the user interface.
         */
             Static = "Static",        
        /**
         * 1: The value is a normal edit field
         */
             Edit = "Edit",        
        /**
         * 2: The value is a password field and a pasword control should be used.
         */
             Password = "Password",        
        /**
         * 3: The GetCredentialsGroups and GetCredentialUsers needs to be used to select a value.
         */
             Link = "Link",        
        /**
         * 4: The value should be hidden in the user interface.
         */
             Hidden = "Hidden",        
  }


  /**
   * Enumerator describing how the series of recurring activities are terminated.
   */
  export enum RecurrenceEndType {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: The end of the recurrence is described by an end date.
         */
             EndDate = "EndDate",        
        /**
         * 2: The end of the recurrence is described with a number of recurring activities.
         */
             Counter = "Counter",        
  }


  /**
   * Describes what source the associates should be retrieved from. 
   * This Enum consists of flag values that can be combined.
   */
  export enum AssociateSourceType {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = 0,        
        /**
         * 1: Associate is retrieved from the history list
         */
             History = 1,        
        /**
         * 2: Associate is retrieved from the diary view list
         */
             DiaryViewList = 2,        
        /**
         * 4: Associate is retrieved from the department (usergroup) list
         */
             Department = 4,        
        /**
         * 7: Enumeration generated from the wsdl doesn't handle the [Flags] attribute (bit enumeration)
         */
             All = 7,        
        /**
         * 8: Undefined and generated value 8
         */
             UndefinedValue8 = 8,        
  }


  /**
   * Describes the available relation to owner types available for roles
   */
  export enum RoleRelationToOwner {
        /**
         * 0: Data owned by my own user
         */
             MyOwn = "MyOwn",        
        /**
         * 1: Data owned by users in my primary group
         */
             PrimaryGroup = "PrimaryGroup",        
        /**
         * 2: Data owned by users in my other groups
         */
             OtherGroups = "OtherGroups",        
        /**
         * 3: Data owned by users that has no common user groups with the current user
         */
             OtherAssociates = "OtherAssociates",        
        /**
         * 4: Data owned by external users
         */
             ExternalUser = "ExternalUser",        
        /**
         * 5: Data owned by anonymous users
         */
             Anonymous = "Anonymous",        
        /**
         * 6: Data owned by my company
         */
             MyCompany = "MyCompany",        
        /**
         * 7: Data owned by users on the same project
         */
             SameProject = "SameProject",        
  }


  /**
   * Describes what part of the external user should be changed.
   * This Enum consists of flag values that can be combined.
   */
  export enum ExternalUserInfoModification {
        /**
         * 0: Unknown - used when initializing
         */
             Unknown = 0,        
        /**
         * 1: Change the UserName
         */
             UserName = 1,        
        /**
         * 2: Change the Password
         */
             Password = 2,        
        /**
         * 4: Change role id.
         */
             Role = 4,        
        /**
         * 8: Change isActive
         */
             Active = 8,        
        /**
         * 15: Enumeration generated from the wsdl doesn't handle the [Flags] attribute (bit enumeration)
         */
             All = 15,        
        /**
         * 16: Undefined and generated value 16
         */
             UndefinedValue16 = 16,        
  }


  /**
   * Access levels to a single field. Read and/or write.
   * This Enum consists of flag values that can be combined.
   */
  export enum EFieldRight {
        /**
         * 0: Hidden from user.
         */
             None = 0,        
        /**
         * 1: Read-only right
         */
             Read = 1,        
        /**
         * 2: Write-only (not terribly useful)
         */
             Write = 2,        
        /**
         * 3: Read + Write rights combined
         */
             Update = 3,        
        /**
         * 4: 
         */
             Unused1 = 4,        
        /**
         * 8: 
         */
             Unused2 = 8,        
        /**
         * 16: 
         */
             Unused3 = 16,        
        /**
         * 32: 
         */
             Unused4 = 32,        
        /**
         * 64: Field must be filled out (not blank)
         */
             UIHintMandatory = 64,        
        /**
         * 128: Field can only be read, not updated, in the user interface.
         */
             UIHintReadOnly = 128,        
        /**
         * 192: Rights only applied in the user interface.  These rights are threated as deny rights.
         */
             UIHints = 192,        
        /**
         * 256: Undefined and generated value 256
         */
             UndefinedValue256 = 256,        
        /**
         * 32: Field can have NULL as a valid value.
         */
             Nullable = 32,        
  }


  /**
   * Operators to be used between restrictions, describes how this restriction is related to the next one in an array
   */
  export enum InterRestrictionOperator {
        /**
         * 0: 
         */
             None = "None",        
        /**
         * 1: 
         */
             And = "And",        
        /**
         * 2: 
         */
             Or = "Or",        
  }


  /**
   * Enumerator for the days of the week
   * This Enum is used by:  email_flow.selected_days
   * This Enum consists of flag values that can be combined.
   */
  export enum Weekday {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = 0,        
        /**
         * 1: Monday
         */
             Monday = 1,        
        /**
         * 2: Tuesday
         */
             Tuesday = 2,        
        /**
         * 4: Wednesday
         */
             Wednesday = 4,        
        /**
         * 8: Thursday
         */
             Thursday = 8,        
        /**
         * 16: Friday
         */
             Friday = 16,        
        /**
         * 32: Saturday
         */
             Saturday = 32,        
        /**
         * 64: Sunday
         */
             Sunday = 64,        
  }


  /**
   * Enumerator describing the different patterns for a yearly recurrence
   */
  export enum RecurrenceYearlyPattern {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 6: The recurrence is for a numbered day of the month, e.g. the 17. day of August
         */
             DayOfMonth = "DayOfMonth",        
        /**
         * 7: The recurrence is for weekdays in the month, e.g. the 3. thursday of August
         */
             WeekdayOfMonth = "WeekdayOfMonth",        
  }


  /**
   * Enumerator describing the different patterns for a monthly recurrence
   */
  export enum RecurrenceMonthlyPattern {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 4: The recurrence is for a numbered day of the month, e.g. the 17. day of every 2 months
         */
             DayOfMonth = "DayOfMonth",        
        /**
         * 5: The recurrence is for weekdays in the month, e.g. the 3. thursday of every 3 months
         */
             WeekdayOfMonth = "WeekdayOfMonth",        
  }


  /**
   * Enumerator describing the different patterns for a daily recurrence
   */
  export enum RecurrenceDailyPattern {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: The recurrence is for every workday (Mon-Fri)
         */
             EveryWorkday = "EveryWorkday",        
        /**
         * 2: The recurrence is for every workday (Mon-Sun)
         */
             EveryWeekday = "EveryWeekday",        
        /**
         * 3: The recurrence is for every cyclic interval of days.
         */
             EveryCyclicDay = "EveryCyclicDay",        
  }


  /**
   * Enumerator representing a week of the month
   */
  export enum WeekOfMonth {
        /**
         * 0: 0: Unknown/undecided type
         */
             Unknown = "Unknown",        
        /**
         * 1: The first week of the month
         */
             First = "First",        
        /**
         * 2: The second week of the month
         */
             Second = "Second",        
        /**
         * 3: The third week of the month
         */
             Third = "Third",        
        /**
         * 4: The fourth week of the month
         */
             Fourth = "Fourth",        
        /**
         * 5: The last week of the month
         */
             Last = "Last",        
  }


  /**
   * How to sort the Order by statement.
   */
  export enum OrderBySortType {
        /**
         * 0: Sort the Order by Ascending
         */
             ASC = "ASC",        
        /**
         * 1: Sort the Order by Descending
         */
             DESC = "DESC",        
  }


  /**
   * 
   */
  export enum EjUserStatus {
        /**
         * 0: 
         */
             StatusNone = "StatusNone",        
        /**
         * 1: 
         */
             StatusNormal = "StatusNormal",        
        /**
         * 2: 
         */
             StatusNotAvailable = "StatusNotAvailable",        
        /**
         * 3: 
         */
             StatusDeleted = "StatusDeleted",        
        /**
         * 4: 
         */
             StatusReadOnly = "StatusReadOnly",        
        /**
         * 5: 
         */
             StatusSpm = "StatusSpm",        
        /**
         * 127: 
         */
             StatusSystem = "StatusSystem",        
  }


  /**
   * 
   */
  export enum StatusScreenPanelType {
        /**
         * 0: 
         */
             None = "None",        
        /**
         * 1: 
         */
             OpenRequestsPerCategory = "OpenRequestsPerCategory",        
        /**
         * 2: 
         */
             OpenRequestsPerTopCategory = "OpenRequestsPerTopCategory",        
        /**
         * 3: 
         */
             OpenRequestsPerUser = "OpenRequestsPerUser",        
        /**
         * 4: 
         */
             OpenRequestsPerUserGroup = "OpenRequestsPerUserGroup",        
        /**
         * 5: 
         */
             OpenRequestsPerPriority = "OpenRequestsPerPriority",        
        /**
         * 6: 
         */
             OpenRequestsPerStatus = "OpenRequestsPerStatus",        
        /**
         * 7: 
         */
             OpenRequestsPerOrigin = "OpenRequestsPerOrigin",        
        /**
         * 11: 
         */
             NumberOfCreatedRequestsPerDay = "NumberOfCreatedRequestsPerDay",        
        /**
         * 12: 
         */
             NumberOfCreatedRequestsPerWeek = "NumberOfCreatedRequestsPerWeek",        
        /**
         * 13: 
         */
             NumberOfCreatedRequestsPerMonth = "NumberOfCreatedRequestsPerMonth",        
        /**
         * 21: 
         */
             UserResponseTimePerDay = "UserResponseTimePerDay",        
        /**
         * 22: 
         */
             UserReponseTimePerWeek = "UserReponseTimePerWeek",        
        /**
         * 23: 
         */
             UserResponseTimePerMonth = "UserResponseTimePerMonth",        
        /**
         * 101: 
         */
             OwnOpenRequests = "OwnOpenRequests",        
        /**
         * 102: 
         */
             UnassignedRequests = "UnassignedRequests",        
        /**
         * 103: 
         */
             Hotlist = "Hotlist",        
        /**
         * 111: 
         */
             StatusScreen = "StatusScreen",        
  }


  /**
   * 
   */
  export enum NetServerBuildType {
        /**
         * 0: SuperOffice Research and Product Development confidential version that never shall run on machines not owned by SuperOffice Research and Product Development.
         */
             Feature = "Feature",        
        /**
         * 1: SuperOffice Research and Product Development confidential version that never shall run on machines not owned by SuperOffice Research and Product Development.
         */
             Stable = "Stable",        
        /**
         * 2: SuperOffice Confidential version that shall only be used by SuperOffice or partners with a signed and valid agreement with SuperOffice.
         */
             Alpha = "Alpha",        
        /**
         * 3: Beta version intended for SuperOffice with beta partners
         */
             Beta = "Beta",        
        /**
         * 4: Release candidate intended for customers participating in a ramp-up or beta program.
         */
             ReleaseCandidate = "ReleaseCandidate",        
        /**
         * 5: Release intended for customers.
         */
             Release = "Release",        
  }


  /**
   * Possible actions for QuoteVersionButton states.
   */
  export enum QuoteVersionButtonAction {
        /**
         * 0: No action
         */
             None = "None",        
        /**
         * 1: Validate the quote version.
         */
             ValidateVersion = "ValidateVersion",        
        /**
         * 2: Edit the quote version.
         */
             Edit = "Edit",        
        /**
         * 3: Send the quote version.
         */
             Send = "Send",        
        /**
         * 4: Approve the quote version.
         */
             Approve = "Approve",        
        /**
         * 5: Reject the quote version.
         */
             Reject = "Reject",        
        /**
         * 7: Validate the quote version and send if possible.
         */
             ValidateVersionAndSendIfPossible = "ValidateVersionAndSendIfPossible",        
        /**
         * 8: Validate the quote version and place the order if possible.
         */
             ValidateVersionAndPlaceOrderIfPossible = "ValidateVersionAndPlaceOrderIfPossible",        
        /**
         * 9: Make a copy of the quote version.
         */
             CloneVersion = "CloneVersion",        
        /**
         * 10: Send confirmation.
         */
             SendConfirmation = "SendConfirmation",        
        /**
         * 11: Fetch updated values from the ERP system and show the results in the dialog.
         */
             GetOrderState = "GetOrderState",        
        /**
         * 12: Ask the quote connector to update the prices for the entire version.
         */
             UpdatePrices = "UpdatePrices",        
  }


  /**
   * Domains for temporary keys.
   * This Enum is used by:  TemporaryKey.domain
   */
  export enum TemporaryKeyDomain {
        /**
         * 0: Unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Mailing recipient.
         */
             MailingRecipient = "MailingRecipient",        
        /**
         * 2: Form submission recipient
         */
             FormSubmitterRecipient = "FormSubmitterRecipient",        
        /**
         * 3: View request in customer center
         */
             ViewTicketInCustomerCenter = "ViewTicketInCustomerCenter",        
        /**
         * 4: Customer center login key
         */
             LoginCustomerCenter = "LoginCustomerCenter",        
        /**
         * 5: Customer center public faq key
         */
             PublicFaq = "PublicFaq",        
        /**
         * 6: Customer center change password key
         */
             ChangePasswordCustomerCenter = "ChangePasswordCustomerCenter",        
  }


  /**
   * Is this document template some kind of quote document, and if so what
   * This Enum is used by:  DocTmpl.privacyDocType
   */
  export enum DocTmplPrivacyType {
        /**
         * 0: This document template is in no way related to GDPR
         */
             None = "None",        
        /**
         * 1: This is a template for the 'you have been added to our system' kind of email; must be an .html email template
         */
             PersonRegistered = "PersonRegistered",        
  }


  /**
   * Webhook status: active, stopped, or tooManyErrors
   * This Enum is used by:  Webhook.state
   */
  export enum WebhookState {
        /**
         * 0: 0: Unknown state
         */
             Unknown = "Unknown",        
        /**
         * 1: 1: Active - webhook is posting events to receiver
         */
             Active = "Active",        
        /**
         * 2: 2: Stopped - webhook has been turned off. Not posting events to receiver.
         */
             Stopped = "Stopped",        
        /**
         * 3: 3: Too many errors received - webhook has been turned off automatically.
         */
             TooManyErrors = "TooManyErrors",        
  }


  /**
   * Command action
   */
  export enum CommandAction {
        /**
         * 0: Command should be implicitly executed.
         */
             Implicit = "Implicit",        
        /**
         * 1: The client must display a dialog with buttons Yes and No.
         */
             YesNo = "YesNo",        
        /**
         * 2: The client must display a dialog with buttons Ok and Cancel.
         */
             OkCancel = "OkCancel",        
        /**
         * 3: The client must display a dialog with Ok-button
         */
             Ok = "Ok",        
  }


  /**
   * Result of an action
   */
  export enum CommandActionResult {
        /**
         * 0: Command should be implicitly executed.
         */
             Implicit = "Implicit",        
        /**
         * 1: The user clicked Yes.
         */
             Yes = "Yes",        
        /**
         * 2: The client clicked Ok
         */
             Ok = "Ok",        
        /**
         * 3: The client clicked No
         */
             No = "No",        
        /**
         * 4: The client clicked Cancel
         */
             Cancel = "Cancel",        
  }


  /**
   * Value for field 'DevicePlatform' in table 'PushNotificationService'
   * This Enum is used by:  PushNotificationService.DevicePlatform
   */
  export enum NotificationPlatform {
        /**
         * 0: Device is by Apple
         */
             Apple = "Apple",        
        /**
         * 1: Device is by Google
         */
             Google = "Google",        
        /**
         * 2: Device is by Microsoft
         */
             Microsoft = "Microsoft",        
        /**
         * 3: Device is by Apple, used for development and debugging
         */
             AppleDeveloper = "AppleDeveloper",        
        /**
         * 4: Device is by Apple, used for testing
         */
             AppleAdHoc = "AppleAdHoc",        
        /**
         * 5: Device is by Google, used for testing, development and debugging
         */
             GoogleDeveloper = "GoogleDeveloper",        
  }


  /**
   * Defines what type of content the notification contains
   */
  export enum NotificationMessageType {
        /**
         * 0: A plain message
         */
             Message = "Message",        
        /**
         * 1: A message with high priority
         */
             ImportantMessage = "ImportantMessage",        
        /**
         * 2: Trigger an action
         */
             RemoteAction = "RemoteAction",        
        /**
         * 3: Ask user a yes or no question
         */
             YesNoQuestion = "YesNoQuestion",        
        /**
         * 4: Show a web page
         */
             ShowWebPage = "ShowWebPage",        
  }


  /**
   * Valid iCal methods
   */
  export enum CalMethod {
        /**
         * 0: Method is not known
         */
             Unknown = "Unknown",        
        /**
         * 1: Add one or more instances to an existing event.
         */
             Add = "Add",        
        /**
         * 2: Cancel one or more instances of an existing event.
         */
             Cancel = "Cancel",        
        /**
         * 3: Counter a REQUEST with an alternative proposal, Sent by an 'Attendee' to the 'Organizer'. 
         */
             Counter = "Counter",        
        /**
         * 4: Decline a counter proposal. Sent to an 'Attendee' by the 'Organizer'.
         */
             DeclineCounter = "DeclineCounter",        
        /**
         * 5: Post notification of an event. Used primarily as a method of advertising the existence of an event.
         */
             Publish = "Publish",        
        /**
         * 6: A request is sent to an 'Organizer' by an 'Attendee' asking for the latest version of an event to be resent to the requester.
         */
             Refresh = "Refresh",        
        /**
         * 7: Reply to an event request. Clients may set their status ('partstat') to ACCEPTED, DECLINED, TENTATIVE, or DELEGATED.
         */
             Reply = "Reply",        
        /**
         * 8: Make a request for an event. This is an explicit invitation to one or more 'Attendees'. Event Requests are also used to update or change an existing event. Clients that cannot handle REQUEST may degrade the event to view it as an PUBLISH.
         */
             Request = "Request",        
  }


  /**
   * iCal reply status
   */
  export enum CalReplyStatus {
        /**
         * 0: Status is not known
         */
             Unknown = "Unknown",        
        /**
         * 1: The sender accepted the meeting.
         */
             Accepted = "Accepted",        
        /**
         * 2: The sender declined the meeting.
         */
             Declined = "Declined",        
        /**
         * 3: The sender tentatively accepted the meeting.
         */
             Tentative = "Tentative",        
  }


  /**
   * Sender email mode
   * This Enum is used by:  DocTmpl.senderEmailMode
   */
  export enum SenderMailMode {
        /**
         * 0: Use default sender, saved in template SenderEmailAddress
         */
             UseDefaultSender = "UseDefaultSender",        
        /**
         * 1: Use persons our contact email address
         */
             UseOurContact = "UseOurContact",        
        /**
         * 2: Use persons support contact.
         */
             UseSupportContact = "UseSupportContact",        
        /**
         * 3: Use logged in user.
         */
             UseLoggedInUser = "UseLoggedInUser",        
  }


  /**
   * Forms recaptcha mode
   */
  export enum FormsRecaptchaMode {
        /**
         * 0: Recpatcha is turned off
         */
             NotAvailable = "NotAvailable",        
        /**
         * 1: Global recaptcha keys are configured
         */
             GlobalKeysExist = "GlobalKeysExist",        
        /**
         * 2: Recaptcha keys need to be supplied
         */
             KeysNeeded = "KeysNeeded",        
  }


  /**
   * Appointment and invitation synchronization info
   * This Enum is used by:  appointment.cautionWarning
   */
  export enum AppointmentCautionWarning {
        /**
         * 0: OK - no special warnings
         */
             OK = "OK",        
        /**
         * 1: Possibly not in sync with external systems
         */
             NotInSync = "NotInSync",        
        /**
         * 2: Updated without sending email notifications
         */
             NotNotifiedByEmail = "NotNotifiedByEmail",        
        /**
         * 3: This recurrence pattern might not be supported in the external participants calendars
         */
             RecurrencePatternNotSupported = "RecurrencePatternNotSupported",        
        /**
         * 4: A requested change to a recurrence appointment was not supported
         */
             IncomingRecurrenceChangeNotSupported = "IncomingRecurrenceChangeNotSupported",        
        /**
         * 5: Time and date may be wrong for external participants due to unsupported changes
         */
             ExternalParticipantsDateTimeMismatch = "ExternalParticipantsDateTimeMismatch",        
  }


  /**
   * Is this document template some kind of appointment document, and if so what
   * This Enum is used by:  DocTmpl.invitationDocType
   */
  export enum DocTmplInvitationType {
        /**
         * 0: None - Not an appointment type
         */
             None = "None",        
        /**
         * 1: 1 - New appointment
         */
             New = "New",        
        /**
         * 2: 2 - Changed appointment
         */
             Changed = "Changed",        
        /**
         * 3: 3 - Cancelled appointment
         */
             Cancelled = "Cancelled",        
  }


  /**
   * Default video-meeting status for meetings created in SuperOffice CRM
   * This Enum is used by:  Task.default_videomeeting_status
   */
  export enum VideoMeetingStatus {
        /**
         * 0: No Change
         */
             NoChange = "NoChange",        
        /**
         * 1: Appointments created from this task-type should enable video-meetings by default.
         */
             VideoMeetingOn = "VideoMeetingOn",        
        /**
         * 2: Appointments created from this task-type should disable video-meetings by default.
         */
             VideoMeetingOff = "VideoMeetingOff",        
  }


  /**
   * Upsert: Action to take on target table rows that do not match any incoming keys
   */
  export enum UpsertNomatchAction {
        /**
         * 0: No action, leave target table row unchanged
         */
             NoChange = "NoChange",        
        /**
         * 1: Zero/default-value the targeted 'columns' in all non-matching rows
         */
             ZeroColumns = "ZeroColumns",        
        /**
         * 2: Delete the entire row, for all non-matching-row
         */
             DeleteRow = "DeleteRow",        
  }


  /**
   * Upsert: Action status for each row
   */
  export enum UpsertRowActionStatus {
        /**
         * 1: Row not found in target, inserted; primary key assigned as required
         */
             Inserted = "Inserted",        
        /**
         * 2: Row found in target, update needed
         */
             Updated = "Updated",        
        /**
         * 3: Row found in target, no updated needed
         */
             NoUpdateNeeded = "NoUpdateNeeded",        
        /**
         * 4: Row not matched in incoming keys and DELETED
         */
             Deleted = "Deleted",        
        /**
         * 5: Row not matched in incoming keys and COLUMNS ZEROED
         */
             ColumnsZeroed = "ColumnsZeroed",        
  }


  /**
   * Service mailbox type
   * This Enum is used by:  mail_in_filter.server_type
   */
  export enum MailboxType {
        /**
         * 0: Unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Pop
         */
             Pop = "Pop",        
        /**
         * 2: Imap - without OAuth support. Use ImapOAuth instead if OAuth is needed
         */
             Imap = "Imap",        
        /**
         * 3: Mapi - Obsolete
         */
             Mapi = "Mapi",        
        /**
         * 4: TypeFacebook
         */
             Facebook = "Facebook",        
        /**
         * 5: Pop SSL
         */
             Pops = "Pops",        
        /**
         * 6: Imap SSL
         */
             Imaps = "Imaps",        
        /**
         * 7: Incoming SMS, using a plugin
         */
             SmsPlugin = "SmsPlugin",        
        /**
         * 8: Used by mailboxes set up as a Mailgun route
         */
             Mailgun = "Mailgun",        
        /**
         * 9: This is a Imap mailbox using OAuth authentication
         */
             ImapOAuth = "ImapOAuth",        
  }


  /**
   * Type and content of delta 
   * This Enum is used by:  ConfigurableScreenDelta.deltaType
   */
  export enum DeltaType {
        /**
         * 0: Unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: System
         */
             System = "System",        
        /**
         * 2: WebPanel from list table extapp, generated by code
         */
             WebPanel = "WebPanel",        
        /**
         * 3: UDefs and extrafields, generated by code
         */
             CustomFields = "CustomFields",        
        /**
         * 4: Customized, added by admin user
         */
             Customized = "Customized",        
  }


  /**
   * State of delta
   * This Enum is used by:  ConfigurableScreenDelta.deltaState
   */
  export enum DeltaState {
        /**
         * 0: No state applicable for this delta, or not yet decided
         */
             Unknown = "Unknown",        
        /**
         * 1: Draft
         */
             Draft = "Draft",        
        /**
         * 2: Published
         */
             Published = "Published",        
  }


  /**
   * Types of events that are sent through the Pocket Notification API, where they are paired with an entity id
   */
  export enum NotificationEventType {
        /**
         * 0: Unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: New ticket
         */
             NewTicket = "NewTicket",        
        /**
         * 2: Appointment invitation
         */
             AppointmentInvitation = "AppointmentInvitation",        
        /**
         * 3: Appointment moved or changed
         */
             AppointmentMoved = "AppointmentMoved",        
        /**
         * 4: Appointment cancelled
         */
             AppointmentCancelled = "AppointmentCancelled",        
        /**
         * 5: New ticket message
         */
             NewTicketMessage = "NewTicketMessage",        
        /**
         * 6: Ticket activated
         */
             TicketActivated = "TicketActivated",        
        /**
         * 7: Ticket escalated
         */
             TicketEscalated = "TicketEscalated",        
        /**
         * 8: Quote approval requested
         */
             QuoteApprovalRequest = "QuoteApprovalRequest",        
        /**
         * 9: Quote approval approved
         */
             QuoteApprovalApproved = "QuoteApprovalApproved",        
        /**
         * 10: Quote approval denied
         */
             QuoteApprovalDenied = "QuoteApprovalDenied",        
        /**
         * 11: Invitation or assignment declined
         */
             AppointmentDeclined = "AppointmentDeclined",        
  }


  /**
   * Different methods the user may use to change password
   */
  export enum ChangePasswordType {
        /**
         * 0: The user can change password by triggering a change password email.
         */
             Email = "Email",        
        /**
         * 1: The user can change password by providing a new password
         */
             Password = "Password",        
  }


  /**
   * Standard/predefined values for ticket message header
   */
  export enum MessageHeaderStdItem {
        /**
         * 0: A standard header without predefined meaning
         */
             None = "None",        
        /**
         * 1: Forward
         */
             Forward = "Forward",        
        /**
         * 2: Message contains certain attachments we simply do not import
         */
             UnnamedAttachmentBlocked = "UnnamedAttachmentBlocked",        
        /**
         * 3: NoAutoReply
         */
             NoAutoReply = "NoAutoReply",        
        /**
         * 4: Auto-reply sent to customers A
         */
             SentAutoReplyToCustomersA = "SentAutoReplyToCustomersA",        
        /**
         * 5: Auto-reply sent to customers B
         */
             SentAutoReplyToCustomersB = "SentAutoReplyToCustomersB",        
        /**
         * 6: Customer read FAQ
         */
             CustomerReadFAQ = "CustomerReadFAQ",        
        /**
         * 7: This message was generated from reply template
         */
             ReplyTemplate = "ReplyTemplate",        
  }


  /**
   * Ticket message header modification options
   */
  export enum MessageHeaderStdItemCol {
        /**
         * 0: The header should be returned/treated as-is, no modifications
         */
             None = "None",        
        /**
         * 1: The localization resource should be added to the Name of a header
         */
             Name = "Name",        
        /**
         * 2: The localization resource should be added to the Value of a header
         */
             Value = "Value",        
  }


  /**
   * Ticket message actions
   */
  export enum MessageActionType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Reply to message
         */
             Reply = "Reply",        
        /**
         * 2: Reply to all
         */
             ReplyAll = "ReplyAll",        
        /**
         * 3: Forward message
         */
             Forward = "Forward",        
        /**
         * 4: Add internal commnet
         */
             InternalComment = "InternalComment",        
  }


  /**
   * Recipient type
   */
  export enum RecipientType {
        /**
         * 0: To
         */
             To = "To",        
        /**
         * 1: CC
         */
             CC = "CC",        
        /**
         * 3: BCC
         */
             BCC = "BCC",        
  }


  /**
   * Embedded images type
   */
  export enum ImageEmbedType {
        /**
         * 0: Link image via href.
         */
             Link = "Link",        
        /**
         * 1: Inline image via base64 encoded data in document.
         */
             Inline = "Inline",        
  }


  /**
   * Task List Item Type
   * This Enum is used by:  TaskMenu.taskType
   */
  export enum TaskListItemType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Url
         */
             Url = "Url",        
        /**
         * 2: Crm Script
         */
             CrmScript = "CrmScript",        
        /**
         * 3: SoProtocol
         */
             SoProtocol = "SoProtocol",        
  }


  /**
   * Show Task Item In Client
   * This Enum is used by:  TaskMenu.showInClient
   * This Enum consists of flag values that can be combined.
   */
  export enum ShowTaskItemInClient {
        /**
         * 0: None
         */
             None = 0,        
        /**
         * 1: SuperOffice Web
         */
             Web = 1,        
        /**
         * 2: Mobile
         */
             Mobile = 2,        
  }


  /**
   * Behaviour in archives
   * This Enum is used by:  TaskMenu.archiveBehaviour
   * This Enum consists of flag values that can be combined.
   */
  export enum ArchiveBehaviour {
        /**
         * 0: None
         */
             None = 0,        
        /**
         * 1: RightClick in archives
         */
             InArchives = 1,        
        /**
         * 2: RightClick with multiselect in archives
         */
             MultiSelectInArchives = 2,        
  }


  /**
   * CrmScript Event triggers - when CRMScript is run based on user or system actions. 
   */
  export enum EventHandlerType {
        /**
         * 0: Unknown trigger
         */
             Unknown = "Unknown",        
        /**
         * 100: NewTicket
         */
             NewTicket = "NewTicket",        
        /**
         * 101: NewTicketFromCustomerCenter
         */
             NewTicketFromCustomerCenter = "NewTicketFromCustomerCenter",        
        /**
         * 102: NewTicketFromEmail
         */
             NewTicketFromEmail = "NewTicketFromEmail",        
        /**
         * 103: NewTicketFromCustomerCenterBeforeSave
         */
             NewTicketFromCustomerCenterBeforeSave = "NewTicketFromCustomerCenterBeforeSave",        
        /**
         * 104: NewTicketFromSpmLink
         */
             NewTicketFromSpmLink = "NewTicketFromSpmLink",        
        /**
         * 105: NewNotifyTicketFromForm
         */
             NewNotifyTicketFromForm = "NewNotifyTicketFromForm",        
        /**
         * 106: NewTicketFromForm
         */
             NewTicketFromForm = "NewTicketFromForm",        
        /**
         * 109: TicketSave
         */
             TicketSave = "TicketSave",        
        /**
         * 110: TicketClosed
         */
             TicketClosed = "TicketClosed",        
        /**
         * 111: TicketPostponed
         */
             TicketPostponed = "TicketPostponed",        
        /**
         * 112: TicketDeleted
         */
             TicketDeleted = "TicketDeleted",        
        /**
         * 113: TicketActivated
         */
             TicketActivated = "TicketActivated",        
        /**
         * 120: TicketReopened
         */
             TicketReopened = "TicketReopened",        
        /**
         * 121: TicketReopenedFromCustomerCenter
         */
             TicketReopenedFromCustomerCenter = "TicketReopenedFromCustomerCenter",        
        /**
         * 122: TicketReopenedFromEmail
         */
             TicketReopenedFromEmail = "TicketReopenedFromEmail",        
        /**
         * 140: TicketChangedPriority
         */
             TicketChangedPriority = "TicketChangedPriority",        
        /**
         * 141: TicketChangedCategory
         */
             TicketChangedCategory = "TicketChangedCategory",        
        /**
         * 142: TicketChangedOwnedBy
         */
             TicketChangedOwnedBy = "TicketChangedOwnedBy",        
        /**
         * 143: TicketChangedPrimaryCustomer
         */
             TicketChangedPrimaryCustomer = "TicketChangedPrimaryCustomer",        
        /**
         * 144: TicketChangedTicketStatus
         */
             TicketChangedTicketStatus = "TicketChangedTicketStatus",        
        /**
         * 145: TicketChangedSlevel
         */
             TicketChangedSlevel = "TicketChangedSlevel",        
        /**
         * 150: TicketMessageAdded
         */
             TicketMessageAdded = "TicketMessageAdded",        
        /**
         * 151: TicketInternalMessageAdded
         */
             TicketInternalMessageAdded = "TicketInternalMessageAdded",        
        /**
         * 152: TicketExternalMessageAdded
         */
             TicketExternalMessageAdded = "TicketExternalMessageAdded",        
        /**
         * 153: TicketMessageSentimentCalculated
         */
             TicketMessageSentimentCalculated = "TicketMessageSentimentCalculated",        
        /**
         * 200: CompactModeInjection
         */
             CompactModeInjection = "CompactModeInjection",        
        /**
         * 210: CustomerCenterAuthentication
         */
             CustomerCenterAuthentication = "CustomerCenterAuthentication",        
        /**
         * 300: ScheduledTaskFailed
         */
             ScheduledTaskFailed = "ScheduledTaskFailed",        
        /**
         * 301: DbiTaskFailed
         */
             DbiTaskFailed = "DbiTaskFailed",        
        /**
         * 302: CustomerSetSubscriptions
         */
             CustomerSetSubscriptions = "CustomerSetSubscriptions",        
        /**
         * 303: ImportMailBeforeProcessing
         */
             ImportMailBeforeProcessing = "ImportMailBeforeProcessing",        
        /**
         * 304: ImportMailAfterProcessing
         */
             ImportMailAfterProcessing = "ImportMailAfterProcessing",        
        /**
         * 305: MainMenu
         */
             MainMenu = "MainMenu",        
        /**
         * 400: ChatNewSession
         */
             ChatNewSession = "ChatNewSession",        
        /**
         * 401: ChatSessionChangedStatus
         */
             ChatSessionChangedStatus = "ChatSessionChangedStatus",        
        /**
         * 402: ChatBeforeSaveNewMessage
         */
             ChatBeforeSaveNewMessage = "ChatBeforeSaveNewMessage",        
        /**
         * 403: ChatAfterSaveNewMessage
         */
             ChatAfterSaveNewMessage = "ChatAfterSaveNewMessage",        
        /**
         * 500: ServiceScreenNewTicketLoad
         */
             ServiceScreenNewTicketLoad = "ServiceScreenNewTicketLoad",        
        /**
         * 501: ServiceScreenNewQuickTicketLoad
         */
             ServiceScreenNewQuickTicketLoad = "ServiceScreenNewQuickTicketLoad",        
        /**
         * 502: ServiceScreenListTicketMessagesLoad
         */
             ServiceScreenListTicketMessagesLoad = "ServiceScreenListTicketMessagesLoad",        
        /**
         * 503: ServiceScreenAddMessageLoad
         */
             ServiceScreenAddMessageLoad = "ServiceScreenAddMessageLoad",        
        /**
         * 504: ServiceScreenEditTicketLoad
         */
             ServiceScreenEditTicketLoad = "ServiceScreenEditTicketLoad",        
        /**
         * 505: ServiceScreenViewCustomerLoad
         */
             ServiceScreenViewCustomerLoad = "ServiceScreenViewCustomerLoad",        
        /**
         * 506: ServiceScreenEditCustomerLoad
         */
             ServiceScreenEditCustomerLoad = "ServiceScreenEditCustomerLoad",        
        /**
         * 507: ServiceScreenViewCompanyLoad
         */
             ServiceScreenViewCompanyLoad = "ServiceScreenViewCompanyLoad",        
        /**
         * 508: ServiceScreenEditCompanyLoad
         */
             ServiceScreenEditCompanyLoad = "ServiceScreenEditCompanyLoad",        
        /**
         * 509: ServiceScreenForwardLoad
         */
             ServiceScreenForwardLoad = "ServiceScreenForwardLoad",        
        /**
         * 510: ServiceScreenEditExtraTableLoad
         */
             ServiceScreenEditExtraTableLoad = "ServiceScreenEditExtraTableLoad",        
        /**
         * 600: ServiceScreenNewTicketBeforeSubmit
         */
             ServiceScreenNewTicketBeforeSubmit = "ServiceScreenNewTicketBeforeSubmit",        
        /**
         * 601: ServiceScreenNewQuickTicketBeforeSubmit
         */
             ServiceScreenNewQuickTicketBeforeSubmit = "ServiceScreenNewQuickTicketBeforeSubmit",        
        /**
         * 602: ServiceScreenListTicketMessagesBeforeSubmit
         */
             ServiceScreenListTicketMessagesBeforeSubmit = "ServiceScreenListTicketMessagesBeforeSubmit",        
        /**
         * 603: ServiceScreenAddMessageBeforeSubmit
         */
             ServiceScreenAddMessageBeforeSubmit = "ServiceScreenAddMessageBeforeSubmit",        
        /**
         * 604: ServiceScreenEditTicketBeforeSubmit
         */
             ServiceScreenEditTicketBeforeSubmit = "ServiceScreenEditTicketBeforeSubmit",        
        /**
         * 605: ServiceScreenViewCustomerBeforeSubmit
         */
             ServiceScreenViewCustomerBeforeSubmit = "ServiceScreenViewCustomerBeforeSubmit",        
        /**
         * 606: ServiceScreenEditCustomerBeforeSubmit
         */
             ServiceScreenEditCustomerBeforeSubmit = "ServiceScreenEditCustomerBeforeSubmit",        
        /**
         * 607: ServiceScreenViewCompanyBeforeSubmit
         */
             ServiceScreenViewCompanyBeforeSubmit = "ServiceScreenViewCompanyBeforeSubmit",        
        /**
         * 608: ServiceScreenEditCompanyBeforeSubmit
         */
             ServiceScreenEditCompanyBeforeSubmit = "ServiceScreenEditCompanyBeforeSubmit",        
        /**
         * 609: ServiceScreenForwardBeforeSubmit
         */
             ServiceScreenForwardBeforeSubmit = "ServiceScreenForwardBeforeSubmit",        
        /**
         * 610: ServiceScreenEditExtraTableBeforeSubmit
         */
             ServiceScreenEditExtraTableBeforeSubmit = "ServiceScreenEditExtraTableBeforeSubmit",        
        /**
         * 700: ServiceScreenNewTicketAfterSubmit
         */
             ServiceScreenNewTicketAfterSubmit = "ServiceScreenNewTicketAfterSubmit",        
        /**
         * 701: ServiceScreenNewQuickTicketAfterSubmit
         */
             ServiceScreenNewQuickTicketAfterSubmit = "ServiceScreenNewQuickTicketAfterSubmit",        
        /**
         * 702: ServiceScreenListTicketMessagesAfterSubmit
         */
             ServiceScreenListTicketMessagesAfterSubmit = "ServiceScreenListTicketMessagesAfterSubmit",        
        /**
         * 703: ServiceScreenAddMessageAfterSubmit
         */
             ServiceScreenAddMessageAfterSubmit = "ServiceScreenAddMessageAfterSubmit",        
        /**
         * 704: ServiceScreenEditTicketAfterSubmit
         */
             ServiceScreenEditTicketAfterSubmit = "ServiceScreenEditTicketAfterSubmit",        
        /**
         * 705: ServiceScreenViewCustomerAfterSubmit
         */
             ServiceScreenViewCustomerAfterSubmit = "ServiceScreenViewCustomerAfterSubmit",        
        /**
         * 706: ServiceScreenEditCustomerAfterSubmit
         */
             ServiceScreenEditCustomerAfterSubmit = "ServiceScreenEditCustomerAfterSubmit",        
        /**
         * 707: ServiceScreenViewCompanyAfterSubmit
         */
             ServiceScreenViewCompanyAfterSubmit = "ServiceScreenViewCompanyAfterSubmit",        
        /**
         * 708: ServiceScreenEditCompanyAfterSubmit
         */
             ServiceScreenEditCompanyAfterSubmit = "ServiceScreenEditCompanyAfterSubmit",        
        /**
         * 709: ServiceScreenForwardAfterSubmit
         */
             ServiceScreenForwardAfterSubmit = "ServiceScreenForwardAfterSubmit",        
        /**
         * 710: ServiceScreenEditExtraTableAfterSubmit
         */
             ServiceScreenEditExtraTableAfterSubmit = "ServiceScreenEditExtraTableAfterSubmit",        
        /**
         * 1001: SalesBeforeSaveAppointment
         */
             SalesBeforeSaveAppointment = "SalesBeforeSaveAppointment",        
        /**
         * 1002: SalesBeforeSaveStakeholder
         */
             SalesBeforeSaveStakeholder = "SalesBeforeSaveStakeholder",        
        /**
         * 1003: SalesBeforeSaveQuote
         */
             SalesBeforeSaveQuote = "SalesBeforeSaveQuote",        
        /**
         * 1004: SalesBeforeSaveDocument
         */
             SalesBeforeSaveDocument = "SalesBeforeSaveDocument",        
        /**
         * 1005: SalesBeforeSaveContact
         */
             SalesBeforeSaveContact = "SalesBeforeSaveContact",        
        /**
         * 1006: SalesBeforeSavePerson
         */
             SalesBeforeSavePerson = "SalesBeforeSavePerson",        
        /**
         * 1007: SalesBeforeSaveRelation
         */
             SalesBeforeSaveRelation = "SalesBeforeSaveRelation",        
        /**
         * 1008: SalesBeforeSaveSale
         */
             SalesBeforeSaveSale = "SalesBeforeSaveSale",        
        /**
         * 1009: SalesBeforeSaveProject
         */
             SalesBeforeSaveProject = "SalesBeforeSaveProject",        
        /**
         * 1010: SalesBeforeSaveSelection
         */
             SalesBeforeSaveSelection = "SalesBeforeSaveSelection",        
        /**
         * 1011: SalesBeforeSaveProjectMember
         */
             SalesBeforeSaveProjectMember = "SalesBeforeSaveProjectMember",        
        /**
         * 1012: SalesBeforeSaveSelectionMember
         */
             SalesBeforeSaveSelectionMember = "SalesBeforeSaveSelectionMember",        
        /**
         * 1013: SalesBeforeSaveQuoteLine
         */
             SalesBeforeSaveQuoteLine = "SalesBeforeSaveQuoteLine",        
        /**
         * 1014: SalesBeforeSaveApproveQuote
         */
             SalesBeforeSaveApproveQuote = "SalesBeforeSaveApproveQuote",        
        /**
         * 1015: SalesBeforeSaveRejectQuote
         */
             SalesBeforeSaveRejectQuote = "SalesBeforeSaveRejectQuote",        
        /**
         * 1016: SalesBeforeSaveTicket
         */
             SalesBeforeSaveTicket = "SalesBeforeSaveTicket",        
        /**
         * 1101: SalesAfterSaveAppointment
         */
             SalesAfterSaveAppointment = "SalesAfterSaveAppointment",        
        /**
         * 1102: SalesAfterSaveStakeholder
         */
             SalesAfterSaveStakeholder = "SalesAfterSaveStakeholder",        
        /**
         * 1103: SalesAfterSaveQuote
         */
             SalesAfterSaveQuote = "SalesAfterSaveQuote",        
        /**
         * 1104: SalesAfterSaveDocument
         */
             SalesAfterSaveDocument = "SalesAfterSaveDocument",        
        /**
         * 1105: SalesAfterSaveContact
         */
             SalesAfterSaveContact = "SalesAfterSaveContact",        
        /**
         * 1106: SalesAfterSavePerson
         */
             SalesAfterSavePerson = "SalesAfterSavePerson",        
        /**
         * 1107: SalesAfterSaveRelation
         */
             SalesAfterSaveRelation = "SalesAfterSaveRelation",        
        /**
         * 1108: SalesAfterSaveSale
         */
             SalesAfterSaveSale = "SalesAfterSaveSale",        
        /**
         * 1109: SalesAfterSaveProject
         */
             SalesAfterSaveProject = "SalesAfterSaveProject",        
        /**
         * 1110: SalesAfterSaveSelection
         */
             SalesAfterSaveSelection = "SalesAfterSaveSelection",        
        /**
         * 1111: SalesAfterSaveProjectMember
         */
             SalesAfterSaveProjectMember = "SalesAfterSaveProjectMember",        
        /**
         * 1112: SalesAfterSaveSelectionMember
         */
             SalesAfterSaveSelectionMember = "SalesAfterSaveSelectionMember",        
        /**
         * 1113: SalesAfterSaveQuoteLine
         */
             SalesAfterSaveQuoteLine = "SalesAfterSaveQuoteLine",        
        /**
         * 1114: SalesAfterSaveApproveQuote
         */
             SalesAfterSaveApproveQuote = "SalesAfterSaveApproveQuote",        
        /**
         * 1115: SalesAfterSaveRejectQuote
         */
             SalesAfterSaveRejectQuote = "SalesAfterSaveRejectQuote",        
        /**
         * 1116: SalesAfterSaveTicket
         */
             SalesAfterSaveTicket = "SalesAfterSaveTicket",        
  }


  /**
   * Hierarchy domain
   */
  export enum Domain {
        /**
         * 0: Unknown folders
         */
             Unknown = "Unknown",        
        /**
         * 1: ExtraTable folders
         */
             ExtraTables = "ExtraTables",        
        /**
         * 2: ScreenDefinition folders
         */
             ScreenDefinitions = "ScreenDefinitions",        
        /**
         * 3: Script folders
         */
             Scripts = "Scripts",        
        /**
         * 4: Selection folders
         */
             Selections = "Selections",        
        /**
         * 5: ExternalDocument folders
         */
             ExternalDocuments = "ExternalDocuments",        
        /**
         * 6: UserGroup folders
         */
             UserGroups = "UserGroups",        
        /**
         * 7: ExternalDocuments related to spam messages folders
         */
             ExternalDocumentRelatedToSpmMessage = "ExternalDocumentRelatedToSpmMessage",        
        /**
         * 8: Dashboard folders
         */
             Dashboards = "Dashboards",        
        /**
         * 9: EmailFlow folders
         */
             EmailFlows = "EmailFlows",        
  }


  /**
   * What is the status of this submission
   * This Enum is used by:  form_submission.status
   */
  export enum FormSubmissionStatus {
        /**
         * 0: Unknown status
         */
             Unknown = "Unknown",        
        /**
         * 1: In progress, the customer has started filling out fields but has not yet submitted
         */
             InProgress = "InProgress",        
        /**
         * 2: EmailVerification, the customer has submitted but we are waiting for click from sent email
         */
             EmailVerification = "EmailVerification",        
        /**
         * 3: Submitted, the form is submitted but has not yet been executed
         */
             Submitted = "Submitted",        
        /**
         * 4: Processed, the actions of the form have been executed
         */
             Processed = "Processed",        
        /**
         * 5: Failed, there was an error when processing the form
         */
             Failed = "Failed",        
  }


  /**
   * Status of the ticket priority
   * This Enum is used by:  ticket_priority.status
   */
  export enum TicketPriorityStatus {
        /**
         * 0: The status is unknown, should normally not occure in the database
         */
             Unknown = "Unknown",        
        /**
         * 1: The status is normal
         */
             Normal = "Normal",        
        /**
         * 2: The status is marked as deleted
         */
             Deleted = "Deleted",        
  }


  /**
   * The delegate method used for request assigned to this category
   * This Enum is used by:  ej_category.delegate_method
   */
  export enum TicketCategoryDelegateMethod {
        /**
         * 0: The delegate method is unknown, should normally not occure in the database
         */
             Unknown = "Unknown",        
        /**
         * 1: Delegate request even among the members of the category
         */
             Even = "Even",        
        /**
         * 2: Use weights on each member of the category when delegating requests
         */
             Weighted = "Weighted",        
        /**
         * 3: The user with the least active request will be delegated the request
         */
             Unfair = "Unfair",        
        /**
         * 4: Do not delegate request. Instead set them as unassigned
         */
             Not = "Not",        
  }


  /**
   * Default status for requests. Note that this is a preference which should be respected if making UI. It is not enforced by the APIs
   */
  export enum TicketCategoryClosingStatus {
        /**
         * 0: The ticket status should respect the user defined behaviour
         */
             UserDefined = "UserDefined",        
        /**
         * 1: Indicates that the ticket status should be active as default
         */
             Active = "Active",        
        /**
         * 2: Indicates that the ticket status should be closed as default
         */
             Closed = "Closed",        
        /**
         * 3: Indicates that the ticket status should be postponed as default
         */
             Postponed = "Postponed",        
  }


  /**
   * Flags for the ticket category
   * This Enum is used by:  ej_category.flags
   * This Enum consists of flag values that can be combined.
   */
  export enum TicketCategoryFlags {
        /**
         * 0: No flags set
         */
             Unknown = 0,        
        /**
         * 1: The category is set to internal and is not available in the Customer Centre
         */
             Internal = 1,        
        /**
         * 2: Only assign requests to logged in users
         */
             OnlyLoggedInUsers = 2,        
        /**
         * 4: Assign to category administrator if no one is available
         */
             DefaultToCategoryMaster = 4,        
        /**
         * 8: Show category in status view
         */
             ListInMain = 8,        
        /**
         * 16: Reassign requests on new message if user is not present
         */
             RedelegateNewMsg = 16,        
        /**
         * 32: Assign to our service contact if available
         */
             DelegateToOurContact = 32,        
        /**
         * 64: Merge outbound e-mails for this category with reply template
         */
             PropagateReplyTemplate = 64,        
        /**
         * 128: Only members have access to requests in this category
         */
             MembersOnly = 128,        
        /**
         * 256: Send notification to members about new request
         */
             NotifyMembers = 256,        
        /**
         * 512: Take ownership by default when replying
         */
             AcceptWhenReplying = 512,        
  }


  /**
   * Flags for the ticket priority
   * This Enum is used by:  ticket_priority.flags
   * This Enum consists of flag values that can be combined.
   */
  export enum TicketPriorityFlags {
        /**
         * 0: No flags set
         */
             Unknown = 0,        
        /**
         * 1: The ticket priority will be avilable in the Customer Centre
         */
             External = 1,        
        /**
         * 2: This is the ticket priority that should default be set in the UI when creating a request
         */
             Default = 2,        
        /**
         * 4: Escalate requests according to time frame
         */
             AlertSchedule = 4,        
  }


  /**
   * The different types of escalate events that can be set
   * This Enum is used by:  ticket_priority.ticket_read; ticket_priority.changed_owner; ticket_priority.ticket_newinfo; ticket_priority.ticket_closed; ticket_priority.ticket_changed_priority; ticket_priority.ticket_new
   */
  export enum TicketPriorityEscalateEvent {
        /**
         * -1: No esclate event set
         */
             None = "None",        
        /**
         * 0: This event will stop the escalation chain
         */
             Stop = "Stop",        
        /**
         * 1: Continue the escalation event from where it was previously stopped
         */
             Continue = "Continue",        
        /**
         * 2: Start the escalation chain over from the beginning
         */
             Restart = "Restart",        
  }


  /**
   * Indicates if a ticket is external or internal
   * This Enum is used by:  ej_message.slevel; ticket.slevel
   */
  export enum TicketSecurityLevel {
        /**
         * 0: Unknown level
         */
             Unknown = "Unknown",        
        /**
         * 1: Internal tickets will not be shown on the Customer Centre
         */
             Internal = "Internal",        
        /**
         * 2: External tickets will be shown on the Customer Centre
         */
             External = "External",        
  }


  /**
   * Whether the owner has read the ticket or not (red, yellow, green)
   * This Enum is used by:  ticket.read_status
   */
  export enum TicketReadStatus {
        /**
         * 0: Unknown status
         */
             Unknown = "Unknown",        
        /**
         * 1: The ticket has been read by the owner
         */
             Green = "Green",        
        /**
         * 2: The ticket has received new information since last read
         */
             Yellow = "Yellow",        
        /**
         * 3: The ticket is unread by the owner
         */
             Red = "Red",        
  }


  /**
   * Indicates if the field body contains plain or html text
   * This Enum is used by:  ej_message.type
   */
  export enum TicketMessageType {
        /**
         * 0: Unknown type
         */
             Unknown = "Unknown",        
        /**
         * 1: The body field contains plain text
         */
             Plain = "Plain",        
        /**
         * 2: The body field contains html text
         */
             Html = "Html",        
  }


  /**
   * Classification of ticket messages
   * This Enum is used by:  ej_message.message_category
   */
  export enum TicketMessageCategory {
        /**
         * 0: A normal message
         */
             Message = "Message",        
        /**
         * 1: Email bounce
         */
             Bounce = "Bounce",        
        /**
         * 2: This message is the error message returned by the sending system (for example a mail server)
         */
             OutboxFailed = "OutboxFailed",        
  }


  /**
   * Indicates what design type this message is created with
   * This Enum is used by:  s_message.designtype
   */
  export enum DesignType {
        /**
         * 0: Unknown type
         */
             Unknown = "Unknown",        
        /**
         * 1: Created with SOEditor
         */
             SOEditor = "SOEditor",        
        /**
         * 2: Created with Unlayer editor
         */
             Unlayer = "Unlayer",        
  }


  /**
   * Bitmask defining what type of shipment message this is
   * This Enum is used by:  s_message.flags
   * This Enum consists of flag values that can be combined.
   */
  export enum ShipmentMessageType {
        /**
         * 1: Contains a plain part
         */
             Plain = 1,        
        /**
         * 2: Contains a HTML part
         */
             Html = 2,        
        /**
         * 4: Obsolete
         */
             SourceView = 4,        
        /**
         * 8: Contains a SMS
         */
             Sms = 8,        
        /**
         * 16: Contains a document. The document_message field contains the id of the actual document
         */
             Document = 16,        
  }


  /**
   * Bitmask available ticket alert action
   * This Enum consists of flag values that can be combined.
   */
  export enum TicketAlertAction {
        /**
         * 1: Controls Reassign
         */
             ActionDelegate = 1,        
        /**
         * 2: Undefined and generated value 2
         */
             UndefinedValue2 = 2,        
        /**
         * 4: Controls Email others
         */
             ActionEmail = 4,        
        /**
         * 8: Controls Email Contact
         */
             ActionEmailCustomer = 8,        
        /**
         * 16: Controls Email User
         */
             ActionEmailUser = 16,        
        /**
         * 32: Controls Email Category administrator
         */
             ActionEmailCategoryMaster = 32,        
        /**
         * 64: Controls SMS others
         */
             ActionSms = 64,        
        /**
         * 128: Controls SMS Contact
         */
             ActionSmsCustomer = 128,        
        /**
         * 256: Controls SMS User
         */
             ActionSmsUser = 256,        
        /**
         * 512: Controls SMS Category administrator
         */
             ActionSmsCategoryMaster = 512,        
        /**
         * 1024: Controls Script
         */
             ActionScript = 1024,        
  }


  /**
   * Value for field 'badge' in table 'ej_message'. Defines the initial source of a message.
   * This Enum is used by:  ej_message.badge
   */
  export enum BadgeType {
        /**
         * 0: Initial message source is unknown
         */
             Unknown = "Unknown",        
        /**
         * 1: Message created as a reply
         */
             Reply = "Reply",        
        /**
         * 2: Message created as a forward
         */
             Forward = "Forward",        
        /**
         * 3: Message created as an internal comment
         */
             Comment = "Comment",        
        /**
         * 4: Message created as a result of other events processing
         */
             Incoming = "Incoming",        
        /**
         * 5: Message created as first outgoing message on ticket
         */
             Outgoing = "Outgoing",        
  }


  /**
   * Chat Topic widget_theme: classic(0) or modern(1).
   * This Enum is used by:  chat_topic.widget_theme
   */
  export enum WidgetTheme {
        /**
         * 0: Set widget theme to classic
         */
             Classic = "Classic",        
        /**
         * 1: Set widget theme to modern
         */
             Modern = "Modern",        
  }


  /**
   * Chat Topic required fields in pre-form.
   * This Enum is used by:  chat_topic.widget_offline_fields; chat_topic.widget_required_fields
   * This Enum consists of flag values that can be combined.
   */
  export enum WidgetRequiredFields {
        /**
         * 0: No required fields
         */
             None = 0,        
        /**
         * 1: Email field is required
         */
             Email = 1,        
        /**
         * 2: Name field is required
         */
             Name = 2,        
        /**
         * 4: Company field is required
         */
             Company = 4,        
        /**
         * 8: Phone field is required
         */
             Phone = 8,        
        /**
         * 16: Country field is required
         */
             Country = 16,        
  }


  /**
   * Value for field 'flags' in table 'chat_session'.
   * This Enum is used by:  chat_session.flags
   * This Enum consists of flag values that can be combined.
   */
  export enum ChatSessionFlags {
        /**
         * 0: Set if the customer is currently typing a chat message
         */
             CustomerIsTyping = 0,        
        /**
         * 1: Set if the agent is currently typing a chat message
         */
             UserIsTyping = 1,        
  }


  /**
   * Value for field 'type' in table 'cust_config'.
   * This Enum is used by:  cust_config.type
   */
  export enum ConfigType {
        /**
         * 0: The row contains style data
         */
             Style = "Style",        
        /**
         * 1: The row contains options data
         */
             Options = "Options",        
  }


  /**
   * What kind of form is this
   * This Enum is used by:  form.type
   */
  export enum FormType {
        /**
         * 0: A form instance
         */
             Normal = "Normal",        
        /**
         * 1: A form template
         */
             Template = "Template",        
  }


  /**
   * Chat Session status
   * This Enum is used by:  chat_session.status
   */
  export enum ChatSessionStatus {
        /**
         * 0: Unknown
         */
             Invalid = "Invalid",        
        /**
         * 1: PreChatForm
         */
             PreChatForm = "PreChatForm",        
        /**
         * 2: FAQ
         */
             Faq = "Faq",        
        /**
         * 3: OfflineForm
         */
             OfflineForm = "OfflineForm",        
        /**
         * 4: InQueue - waiting to be picked up
         */
             InQueue = "InQueue",        
        /**
         * 5: CustomerLast - customer wrote last message
         */
             CustomerLast = "CustomerLast",        
        /**
         * 6: UserLast - user agent wrote last message
         */
             UserLast = "UserLast",        
        /**
         * 7: Finished - may be deleted later
         */
             Finished = "Finished",        
        /**
         * 8: Deleted - ready to delete. Usually cleaned within 24 hrs
         */
             Deleted = "Deleted",        
        /**
         * 9: Closed
         */
             Closed = "Closed",        
        /**
         * 10: RequestPosted
         */
             RequestPosted = "RequestPosted",        
        /**
         * 11: ClosedFromQueue
         */
             ClosedFromQueue = "ClosedFromQueue",        
  }


  /**
   * Chat Message type
   * This Enum is used by:  chat_message.type
   */
  export enum ChatMessageType {
        /**
         * 0: Invalid
         */
             Invalid = "Invalid",        
        /**
         * 1: Message is from user agent to customer
         */
             ToCustomer = "ToCustomer",        
        /**
         * 2: Message is from customer to user agent
         */
             ToUser = "ToUser",        
        /**
         * 3: Special message - see SpecialType field for details
         */
             Special = "Special",        
  }


  /**
   * Chat Message Special type
   * This Enum is used by:  chat_message.special_type
   */
  export enum ChatMessageSpecialType {
        /**
         * 0: None
         */
             None = "None",        
        /**
         * 1: Welcome
         */
             Welcome = "Welcome",        
        /**
         * 2: Url
         */
             Url = "Url",        
        /**
         * 3: Block
         */
             Block = "Block",        
        /**
         * 5: NewSession
         */
             NewSession = "NewSession",        
        /**
         * 6: TransferedSession
         */
             TransferedSession = "TransferedSession",        
        /**
         * 7: Error
         */
             Error = "Error",        
        /**
         * 8: SessionDeleted
         */
             SessionDeleted = "SessionDeleted",        
        /**
         * 9: FaqSuccessQuestion
         */
             FaqSuccessQuestion = "FaqSuccessQuestion",        
        /**
         * 10: ClosedByUser
         */
             ClosedByUser = "ClosedByUser",        
        /**
         * 11: ClosedByCustomer
         */
             ClosedByCustomer = "ClosedByCustomer",        
        /**
         * 12: ClosedByIdle
         */
             ClosedByIdle = "ClosedByIdle",        
        /**
         * 13: TransferRejected
         */
             TransferRejected = "TransferRejected",        
        /**
         * 14: ReopenedByCustomer
         */
             ReopenedByCustomer = "ReopenedByCustomer",        
        /**
         * 15: ClickedOption
         */
             ClickedOption = "ClickedOption",        
        /**
         * 16: Bot message
         */
             BotMessage = "BotMessage",        
        /**
         * 17: Message has a list of clickable options. Options are as JSON array in specialParam
         */
             Options = "Options",        
        /**
         * 18: Uploaded file has illegal type
         */
             IllegalFileType = "IllegalFileType",        
  }


  /**
   * Chat widget size: normal or large
   */
  export enum ChatWidgetSize {
        /**
         * 0: Normal chat widget size.
         */
             Normal = "Normal",        
        /**
         * 1: Large size chat widget.
         */
             Large = "Large",        
  }
