// @superoffice/tsclient.webapi/src/FindAgent.ts
// <auto-generated>
//      SuperOffice Code Generator
//      modelling\templates\servicesnew\tsagents.tt
// </auto-generated>


/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          WebApi/FindAgent.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Find functions
 */
export interface IFindAgent {
    GetCriteriaInformation(request: carriers.Find_GetCriteriaInformationRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetCriteriaInformation(storageType: string, providerName: string, storageKey: string, staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetCriteriaInformationWithContext(request: carriers.Find_GetCriteriaInformationWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetCriteriaInformationWithContext(storageType: string, providerName: string, storageKey: string, staticColumns: string[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    SaveRestrictions(request: carriers.Find_SaveRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictions(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsWithContext(request: carriers.Find_SaveRestrictionsWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsWithContext(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionGroup(request: carriers.Find_SaveRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionGroup(storageType: string, providerName: string, storageKey: string, restrictionGroup: carriers.ArchiveRestrictionGroup, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionGroups(request: carriers.Find_SaveRestrictionGroupsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;
    SaveRestrictionGroups(storageType: string, providerName: string, storageKey: string, restrictionGroups: carriers.ArchiveRestrictionGroup[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;
    GetRestrictionGroups(request: carriers.Find_GetRestrictionGroupsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;
    GetRestrictionGroups(storageType: string, providerName: string, storageKey: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;
    GetRestrictionGroup(request: carriers.Find_GetRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;
    GetRestrictionGroup(storageType: string, providerName: string, storageKey: string, rank: number, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;
    CreateRestrictionGroup(request: carriers.Find_CreateRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;
    CreateRestrictionGroup(storageType: string, providerName: string, storageKey: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;
    DeleteRestrictionGroup(request: carriers.Find_DeleteRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    DeleteRestrictionGroup(storageType: string, providerName: string, storageKey: string, rank: number, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsAndGetCriteriaInformation(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    SaveRestrictionsAndGetCriteriaInformation(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    Find(request: carriers.Find_FindRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    Find(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictions(request: carriers.Find_FindFromRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictions(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumns(request: carriers.Find_FindFromRestrictionsColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumns(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, desiredColumns: string[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    GetAvailableRestrictionColumns(request: carriers.Find_GetAvailableRestrictionColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< string[] >;
    GetAvailableRestrictionColumns(storageType: string, providerName: string, webapi_options?:base.WebApiRequestOptions) : Promise< string[] >;
    GetSpecifiedCriteriaInformationWithDefaults(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetSpecifiedCriteriaInformationWithDefaults(storageType: string, providerName: string, storageKey: string, desiredColumnNames: string[], staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetSpecifiedCriteriaInformationWithDefaultsWithContext(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    GetSpecifiedCriteriaInformationWithDefaultsWithContext(storageType: string, providerName: string, storageKey: string, desiredColumnNames: string[], staticColumns: string[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    FindFromRestrictionsColumnsOrderBy(request: carriers.Find_FindFromRestrictionsColumnsOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumnsOrderBy(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, desiredColumns: string[], orderBy: carriers.ArchiveOrderByInfo[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    GetDefaultDesiredColumns(request: carriers.Find_GetDefaultDesiredColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    GetDefaultDesiredColumns(storageType: string, providerName: string, storageKey: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    GetDefaultOrderBy(request: carriers.Find_GetDefaultOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;
    GetDefaultOrderBy(storageType: string, providerName: string, storageKey: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;
    PopulateRestrictions(request: carriers.Find_PopulateRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;
    PopulateRestrictions(providerName: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;
    PopulateRestrictionValues(request: carriers.Find_PopulateRestrictionValuesRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;
    PopulateRestrictionValues(restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;
    GetDefaultDesiredColumnsFromRestrictions(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    GetDefaultDesiredColumnsFromRestrictions(providerName: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    GetDefaultOrderByFromDesiredColumns(request: carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;
    GetDefaultOrderByFromDesiredColumns(providerName: string, desiredColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;
    FindOrderBy(request: carriers.Find_FindOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindOrderBy(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, orderBy: carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithExtraRestrictions(request: carriers.Find_FindWithExtraRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithExtraRestrictions(storageType: string, providerName: string, storageKey: string, extraRestrictions: carriers.ArchiveRestrictionInfo[], orderBy: carriers.ArchiveOrderByInfo[], desiredColumns: string[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithColumns(request: carriers.Find_FindWithColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithColumns(storageType: string, providerName: string, storageKey: string, desiredColumns: string[], pageSize: number, pageNumber: number, orderBy: carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    SaveRestrictions2(request: carriers.Find_SaveRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictions2(storageType: string, providerName: string, storageKey: string, restrictions: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsWithContext2(request: carriers.Find_SaveRestrictionsWithContext2Request, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsWithContext2(storageType: string, providerName: string, storageKey: string, restrictions: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;
    SaveRestrictionsAndGetCriteriaInformation2(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    SaveRestrictionsAndGetCriteriaInformation2(storageType: string, providerName: string, storageKey: string, restrictions: string, staticColumns: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;
    FindFromRestrictions2(request: carriers.Find_FindFromRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictions2(restrictions: string, providerName: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumns2(request: carriers.Find_FindFromRestrictionsColumns2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumns2(restrictions: string, providerName: string, desiredColumns: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumnsOrderBy2(request: carriers.Find_FindFromRestrictionsColumnsOrderBy2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindFromRestrictionsColumnsOrderBy2(restrictions: string, providerName: string, desiredColumns: string, orderBy: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    GetDefaultDesiredColumnsFromRestrictions2(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    GetDefaultDesiredColumnsFromRestrictions2(providerName: string, restrictions: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;
    FindOrderBy2(request: carriers.Find_FindOrderBy2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindOrderBy2(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, orderBy: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithExtraRestrictions2(request: carriers.Find_FindWithExtraRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
    FindWithExtraRestrictions2(storageType: string, providerName: string, storageKey: string, extraRestrictions: string, orderBy: string, desiredColumns: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;
} // interface IFindAgent




/**
 * Find functions
 */
export class FindAgent extends base.WebApiBase implements IFindAgent {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Agents/Find/";
        super(options);
    }



    private async _GetCriteriaInformation(request: carriers.Find_GetCriteriaInformationRequest, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetCriteriaInformation", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Get criteria information from a set of saved criteria. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     * @param request - StorageType, ProviderName, StorageKey, StaticColumns
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async GetCriteriaInformation(request: carriers.Find_GetCriteriaInformationRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Get criteria information from a set of saved criteria. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param staticColumns - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async GetCriteriaInformation(storageType: string, providerName: string, storageKey: string, staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async GetCriteriaInformation(storageType:string|carriers.Find_GetCriteriaInformationRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, staticColumns?:string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_GetCriteriaInformationRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.StaticColumns)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetCriteriaInformationRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       StaticColumns: staticColumns as string[],
                       };   
            _options = webapi_options;
        }
        return this._GetCriteriaInformation( _request, _options );
    }


    private async _GetCriteriaInformationWithContext(request: carriers.Find_GetCriteriaInformationWithContextRequest, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetCriteriaInformationWithContext", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Get criteria information from a set of saved criteria. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     * @param request - StorageType, ProviderName, StorageKey, StaticColumns, Context
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async GetCriteriaInformationWithContext(request: carriers.Find_GetCriteriaInformationWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Get criteria information from a set of saved criteria. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param staticColumns - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @param context - Optional context that can be used by FindProvider
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async GetCriteriaInformationWithContext(storageType: string, providerName: string, storageKey: string, staticColumns: string[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async GetCriteriaInformationWithContext(storageType:string|carriers.Find_GetCriteriaInformationWithContextRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, staticColumns?:string[], context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_GetCriteriaInformationWithContextRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.StaticColumns)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetCriteriaInformationWithContextRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       StaticColumns: staticColumns as string[],
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._GetCriteriaInformationWithContext( _request, _options );
    }


    private async _SaveRestrictions(request: carriers.Find_SaveRestrictionsRequest, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictions", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param request - StorageType, ProviderName, StorageKey, Restrictions
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictions(request: carriers.Find_SaveRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - Array of restrictions. The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictions(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async SaveRestrictions(storageType:string|carriers.Find_SaveRestrictionsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_SaveRestrictionsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictions( _request, _options );
    }


    private async _SaveRestrictionsWithContext(request: carriers.Find_SaveRestrictionsWithContextRequest, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionsWithContext", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param request - StorageType, ProviderName, StorageKey, Restrictions, Context
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictionsWithContext(request: carriers.Find_SaveRestrictionsWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - Array of restrictions. The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @param context - Optional context that can be used by FindProvider
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictionsWithContext(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async SaveRestrictionsWithContext(storageType:string|carriers.Find_SaveRestrictionsWithContextRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:carriers.ArchiveRestrictionInfo[], context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_SaveRestrictionsWithContextRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionsWithContextRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionsWithContext( _request, _options );
    }


    private async _SaveRestrictionGroup(request: carriers.Find_SaveRestrictionGroupRequest, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionGroup", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions as a restriction group for later use as search criteria (including as dynamic selection and Find). 
     * @param request - StorageType, ProviderName, StorageKey, RestrictionGroup, Context
     * @returns This service call just saves the restrictions.
     */
    public async SaveRestrictionGroup(request: carriers.Find_SaveRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Save an array of restrictions as a restriction group for later use as search criteria (including as dynamic selection and Find). 
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictionGroup - Information about a group of restrictions
     * @param context - Optional context that can be used by FindProvider
     * @returns This service call just saves the restrictions.
     */
    public async SaveRestrictionGroup(storageType: string, providerName: string, storageKey: string, restrictionGroup: carriers.ArchiveRestrictionGroup, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async SaveRestrictionGroup(storageType:string|carriers.Find_SaveRestrictionGroupRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictionGroup?:carriers.ArchiveRestrictionGroup, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_SaveRestrictionGroupRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.RestrictionGroup)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionGroupRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       RestrictionGroup: restrictionGroup as carriers.ArchiveRestrictionGroup,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionGroup( _request, _options );
    }


    private async _SaveRestrictionGroups(request: carriers.Find_SaveRestrictionGroupsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionGroup[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionGroups", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionGroup[] = converters.Create_ArchiveRestrictionGroupArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save and rerank an array of restriction groups, returning the possibly modified array.
     * @param request - StorageType, ProviderName, StorageKey, RestrictionGroups, Context
     * @returns The new restriction group.
     */
    public async SaveRestrictionGroups(request: carriers.Find_SaveRestrictionGroupsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;

    /**
     * Save and rerank an array of restriction groups, returning the possibly modified array.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictionGroups - Information about a group of restrictions
     * @param context - Optional context that can be used by FindProvider
     * @returns The new restriction group.
     */
    public async SaveRestrictionGroups(storageType: string, providerName: string, storageKey: string, restrictionGroups: carriers.ArchiveRestrictionGroup[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;

    public async SaveRestrictionGroups(storageType:string|carriers.Find_SaveRestrictionGroupsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictionGroups?:carriers.ArchiveRestrictionGroup[], context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] > {
        let _request : carriers.Find_SaveRestrictionGroupsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.RestrictionGroups)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionGroupsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       RestrictionGroups: restrictionGroups as carriers.ArchiveRestrictionGroup[],
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionGroups( _request, _options );
    }


    private async _GetRestrictionGroups(request: carriers.Find_GetRestrictionGroupsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionGroup[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetRestrictionGroups", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionGroup[] = converters.Create_ArchiveRestrictionGroupArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Return all the restriction groups.
     * @param request - StorageType, ProviderName, StorageKey, Context
     * @returns The restriction groups.
     */
    public async GetRestrictionGroups(request: carriers.Find_GetRestrictionGroupsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;

    /**
     * Return all the restriction groups.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param context - Optional context that can be used by FindProvider
     * @returns The restriction groups.
     */
    public async GetRestrictionGroups(storageType: string, providerName: string, storageKey: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] >;

    public async GetRestrictionGroups(storageType:string|carriers.Find_GetRestrictionGroupsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup[] > {
        let _request : carriers.Find_GetRestrictionGroupsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetRestrictionGroupsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._GetRestrictionGroups( _request, _options );
    }


    private async _GetRestrictionGroup(request: carriers.Find_GetRestrictionGroupRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionGroup >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetRestrictionGroup", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionGroup = converters.Create_ArchiveRestrictionGroup_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Return the restriction group with given rank or a blank carrier.
     * @param request - StorageType, ProviderName, StorageKey, Rank, Context
     * @returns The new restriction group.
     */
    public async GetRestrictionGroup(request: carriers.Find_GetRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;

    /**
     * Return the restriction group with given rank or a blank carrier.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param rank - Rank of the group to be deleted.
     * @param context - Optional context that can be used by FindProvider
     * @returns The new restriction group.
     */
    public async GetRestrictionGroup(storageType: string, providerName: string, storageKey: string, rank: number, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;

    public async GetRestrictionGroup(storageType:string|carriers.Find_GetRestrictionGroupRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, rank?:number, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup > {
        let _request : carriers.Find_GetRestrictionGroupRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Rank)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetRestrictionGroupRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Rank: rank as number,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._GetRestrictionGroup( _request, _options );
    }


    private async _CreateRestrictionGroup(request: carriers.Find_CreateRestrictionGroupRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionGroup >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("CreateRestrictionGroup", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionGroup = converters.Create_ArchiveRestrictionGroup_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Create a restriction group, initialized with next rank etc.
     * @param request - StorageType, ProviderName, StorageKey, Context
     * @returns The new restriction group.
     */
    public async CreateRestrictionGroup(request: carriers.Find_CreateRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;

    /**
     * Create a restriction group, initialized with next rank etc.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param context - Optional context that can be used by FindProvider
     * @returns The new restriction group.
     */
    public async CreateRestrictionGroup(storageType: string, providerName: string, storageKey: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup >;

    public async CreateRestrictionGroup(storageType:string|carriers.Find_CreateRestrictionGroupRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionGroup > {
        let _request : carriers.Find_CreateRestrictionGroupRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_CreateRestrictionGroupRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._CreateRestrictionGroup( _request, _options );
    }


    private async _DeleteRestrictionGroup(request: carriers.Find_DeleteRestrictionGroupRequest, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("DeleteRestrictionGroup", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Create a restriction group, initialized with next rank etc.
     * @param request - StorageType, ProviderName, StorageKey, Rank, Context
     * @returns This service call just saves the restrictions.
     */
    public async DeleteRestrictionGroup(request: carriers.Find_DeleteRestrictionGroupRequest, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Create a restriction group, initialized with next rank etc.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param rank - Rank of the group to be deleted.
     * @param context - Optional context that can be used by FindProvider
     * @returns This service call just saves the restrictions.
     */
    public async DeleteRestrictionGroup(storageType: string, providerName: string, storageKey: string, rank: number, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async DeleteRestrictionGroup(storageType:string|carriers.Find_DeleteRestrictionGroupRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, rank?:number, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_DeleteRestrictionGroupRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Rank)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_DeleteRestrictionGroupRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Rank: rank as number,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._DeleteRestrictionGroup( _request, _options );
    }


    private async _SaveRestrictionsAndGetCriteriaInformation(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionsAndGetCriteriaInformation", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). Then, return the same result as a call to GetCriteriaInformation would have done. The purpose is to encapsulate saving and updating of a GUI in one round trip.
     * @param request - StorageType, ProviderName, StorageKey, Restrictions, StaticColumns
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async SaveRestrictionsAndGetCriteriaInformation(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). Then, return the same result as a call to GetCriteriaInformation would have done. The purpose is to encapsulate saving and updating of a GUI in one round trip.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - Array of restrictions. The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @param staticColumns - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async SaveRestrictionsAndGetCriteriaInformation(storageType: string, providerName: string, storageKey: string, restrictions: carriers.ArchiveRestrictionInfo[], staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async SaveRestrictionsAndGetCriteriaInformation(storageType:string|carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:carriers.ArchiveRestrictionInfo[], staticColumns?:string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' || typeof(first.StaticColumns)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionsAndGetCriteriaInformationRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       StaticColumns: staticColumns as string[],
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionsAndGetCriteriaInformation( _request, _options );
    }


    private async _Find(request: carriers.Find_FindRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("Find", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby columns are also calculated by the system.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param request - StorageType, ProviderName, StorageKey, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async Find(request: carriers.Find_FindRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby columns are also calculated by the system.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async Find(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async Find(storageType:string|carriers.Find_FindRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._Find( _request, _options );
    }


    private async _FindFromRestrictions(request: carriers.Find_FindFromRestrictionsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictions", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are passed in directly, not fetched by a restriction storage provider. The columns of the result are calculated based on the restriction.
     * @param request - Restrictions, ProviderName, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictions(request: carriers.Find_FindFromRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are passed in directly, not fetched by a restriction storage provider. The columns of the result are calculated based on the restriction.
     * @param restrictions - Array of restrictions specifying the search. Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictions(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictions(restrictions:carriers.ArchiveRestrictionInfo[]|carriers.Find_FindFromRestrictionsRequest, providerName?:string|base.WebApiRequestOptions, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictionsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictionsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       ProviderName: providerName as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictions( _request, _options );
    }


    private async _FindFromRestrictionsColumns(request: carriers.Find_FindFromRestrictionsColumnsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictionsColumns", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is calculated by the system.&lt;para/&gt;Use the GetCriteriaInformation and GetDefaultDesiredColumns service methods to let the system calculate these values, if you want to use or modify them.
     * @param request - Restrictions, ProviderName, DesiredColumns, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumns(request: carriers.Find_FindFromRestrictionsColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is calculated by the system.&lt;para/&gt;Use the GetCriteriaInformation and GetDefaultDesiredColumns service methods to let the system calculate these values, if you want to use or modify them.
     * @param restrictions - Array of restrictions specifying the search. Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param desiredColumns - Array of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumns(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, desiredColumns: string[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictionsColumns(restrictions:carriers.ArchiveRestrictionInfo[]|carriers.Find_FindFromRestrictionsColumnsRequest, providerName?:string|base.WebApiRequestOptions, desiredColumns?:string[], pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictionsColumnsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictionsColumnsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       ProviderName: providerName as string,
                       DesiredColumns: desiredColumns as string[],
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictionsColumns( _request, _options );
    }


    private async _GetAvailableRestrictionColumns(request: carriers.Find_GetAvailableRestrictionColumnsRequest, options?:base.WebApiRequestOptions) : Promise< string[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< string[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetAvailableRestrictionColumns", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : string[] = response.data as string[];
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Get a list of the column names corresponding to available restrictions for a certain archive provider and restriction storage provider. Such columns have CanRestrict set to true, and are supported by the given restriction storage provider.
     * @param request - StorageType, ProviderName
     * @returns Array of column names, corresponding to support restrictions for the given archive and restriction storage providers.
     */
    public async GetAvailableRestrictionColumns(request: carriers.Find_GetAvailableRestrictionColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< string[] >;

    /**
     * Get a list of the column names corresponding to available restrictions for a certain archive provider and restriction storage provider. Such columns have CanRestrict set to true, and are supported by the given restriction storage provider.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @returns Array of column names, corresponding to support restrictions for the given archive and restriction storage providers.
     */
    public async GetAvailableRestrictionColumns(storageType: string, providerName: string, webapi_options?:base.WebApiRequestOptions) : Promise< string[] >;

    public async GetAvailableRestrictionColumns(storageType:string|carriers.Find_GetAvailableRestrictionColumnsRequest, providerName?:string|base.WebApiRequestOptions, webapi_options?:base.WebApiRequestOptions) : Promise< string[] > {
        let _request : carriers.Find_GetAvailableRestrictionColumnsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetAvailableRestrictionColumnsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       };   
            _options = webapi_options;
        }
        return this._GetAvailableRestrictionColumns( _request, _options );
    }


    private async _GetSpecifiedCriteriaInformationWithDefaults(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetSpecifiedCriteriaInformationWithDefaults", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Get criteria information from a set of saved criteria, for a specific set of columns. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     * @param request - StorageType, ProviderName, StorageKey, DesiredColumnNames, StaticColumns
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     */
    public async GetSpecifiedCriteriaInformationWithDefaults(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Get criteria information from a set of saved criteria, for a specific set of columns. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param desiredColumnNames - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @param staticColumns - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     */
    public async GetSpecifiedCriteriaInformationWithDefaults(storageType: string, providerName: string, storageKey: string, desiredColumnNames: string[], staticColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async GetSpecifiedCriteriaInformationWithDefaults(storageType:string|carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, desiredColumnNames?:string[], staticColumns?:string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.DesiredColumnNames)!='undefined' || typeof(first.StaticColumns)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       DesiredColumnNames: desiredColumnNames as string[],
                       StaticColumns: staticColumns as string[],
                       };   
            _options = webapi_options;
        }
        return this._GetSpecifiedCriteriaInformationWithDefaults( _request, _options );
    }


    private async _GetSpecifiedCriteriaInformationWithDefaultsWithContext(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetSpecifiedCriteriaInformationWithDefaultsWithContext", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Get criteria information from a set of saved criteria, for a specific set of columns. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     * @param request - StorageType, ProviderName, StorageKey, DesiredColumnNames, StaticColumns, Context
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     */
    public async GetSpecifiedCriteriaInformationWithDefaultsWithContext(request: carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Get criteria information from a set of saved criteria, for a specific set of columns. The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param desiredColumnNames - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @param staticColumns - Optional array of restrictions that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @param context - Optional context that can be used by FindProvider
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control. ALL columns specified in the call will be present in the results; those that do not have corresponding criteria set will have empty values and the default (first) operator, with the IsActive flag set to false.
     */
    public async GetSpecifiedCriteriaInformationWithDefaultsWithContext(storageType: string, providerName: string, storageKey: string, desiredColumnNames: string[], staticColumns: string[], context: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async GetSpecifiedCriteriaInformationWithDefaultsWithContext(storageType:string|carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, desiredColumnNames?:string[], staticColumns?:string[], context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.DesiredColumnNames)!='undefined' || typeof(first.StaticColumns)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetSpecifiedCriteriaInformationWithDefaultsWithContextRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       DesiredColumnNames: desiredColumnNames as string[],
                       StaticColumns: staticColumns as string[],
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._GetSpecifiedCriteriaInformationWithDefaultsWithContext( _request, _options );
    }


    private async _FindFromRestrictionsColumnsOrderBy(request: carriers.Find_FindFromRestrictionsColumnsOrderByRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictionsColumnsOrderBy", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is also passed in directly.&lt;para/&gt;Use the GetCriteriaInformation, GetDefaultDesiredColumns and GetDefaultOrderBy service methods to let the system calculate these values, if you want to use or modify them.
     * @param request - Restrictions, ProviderName, DesiredColumns, OrderBy, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumnsOrderBy(request: carriers.Find_FindFromRestrictionsColumnsOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is also passed in directly.&lt;para/&gt;Use the GetCriteriaInformation, GetDefaultDesiredColumns and GetDefaultOrderBy service methods to let the system calculate these values, if you want to use or modify them.
     * @param restrictions - Array of restrictions specifying the search. Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param desiredColumns - Array of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param orderBy - Array of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumnsOrderBy(restrictions: carriers.ArchiveRestrictionInfo[], providerName: string, desiredColumns: string[], orderBy: carriers.ArchiveOrderByInfo[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictionsColumnsOrderBy(restrictions:carriers.ArchiveRestrictionInfo[]|carriers.Find_FindFromRestrictionsColumnsOrderByRequest, providerName?:string|base.WebApiRequestOptions, desiredColumns?:string[], orderBy?:carriers.ArchiveOrderByInfo[], pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictionsColumnsOrderByRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.OrderBy)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictionsColumnsOrderByRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       ProviderName: providerName as string,
                       DesiredColumns: desiredColumns as string[],
                       OrderBy: orderBy as carriers.ArchiveOrderByInfo[],
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictionsColumnsOrderBy( _request, _options );
    }


    private async _GetDefaultDesiredColumns(request: carriers.Find_GetDefaultDesiredColumnsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveColumnInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetDefaultDesiredColumns", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveColumnInfo[] = converters.Create_ArchiveColumnInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a storage type, provider name and storage key, which are used to fetch the corresponding restrictions from the database (in the same way as Find does). If you want to specify the restriction directly, use the GetDefaultDesiredColumnsFromRestrictions method instead. This is the algorithm that is used by the Find service method.
     * @param request - StorageType, ProviderName, StorageKey
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumns(request: carriers.Find_GetDefaultDesiredColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a storage type, provider name and storage key, which are used to fetch the corresponding restrictions from the database (in the same way as Find does). If you want to specify the restriction directly, use the GetDefaultDesiredColumnsFromRestrictions method instead. This is the algorithm that is used by the Find service method.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of the provider to calculate default desired columns for
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumns(storageType: string, providerName: string, storageKey: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    public async GetDefaultDesiredColumns(storageType:string|carriers.Find_GetDefaultDesiredColumnsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        let _request : carriers.Find_GetDefaultDesiredColumnsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetDefaultDesiredColumnsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       };   
            _options = webapi_options;
        }
        return this._GetDefaultDesiredColumns( _request, _options );
    }


    private async _GetDefaultOrderBy(request: carriers.Find_GetDefaultOrderByRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveOrderByInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetDefaultOrderBy", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveOrderByInfo[] = converters.Create_ArchiveOrderByInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Calculate the default orderby columns for a given provider and a search. The search is specified by a storage type, provider name and storage key, and is fetched from the database. Default desired columns are then calculated for the search, and those columns are then used as the basis for calculating an order by. If you want to specify the desired columns directly, use the GetDefaultOrderByFromDesiredColumns method instead.  This is the same algorithm that is used by the Find service method.
     * @param request - StorageType, ProviderName, StorageKey
     * @returns Orderby information
     */
    public async GetDefaultOrderBy(request: carriers.Find_GetDefaultOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;

    /**
     * Calculate the default orderby columns for a given provider and a search. The search is specified by a storage type, provider name and storage key, and is fetched from the database. Default desired columns are then calculated for the search, and those columns are then used as the basis for calculating an order by. If you want to specify the desired columns directly, use the GetDefaultOrderByFromDesiredColumns method instead.  This is the same algorithm that is used by the Find service method.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Provider name to calculate default orderby for
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @returns Orderby information
     */
    public async GetDefaultOrderBy(storageType: string, providerName: string, storageKey: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;

    public async GetDefaultOrderBy(storageType:string|carriers.Find_GetDefaultOrderByRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] > {
        let _request : carriers.Find_GetDefaultOrderByRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' ) )
        {
            _request = storageType as carriers.Find_GetDefaultOrderByRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       };   
            _options = webapi_options;
        }
        return this._GetDefaultOrderBy( _request, _options );
    }


    private async _PopulateRestrictions(request: carriers.Find_PopulateRestrictionsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("PopulateRestrictions", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionInfo[] = converters.Create_ArchiveRestrictionInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Take an incoming set of minimally populated restrictions (name + operator is required), and populate all the other parts of the ArchiveRestrictionInfo structure. This includes column information, display values (including list value lookup), and calculated/default values where the value hints specify read-only (R).
     * @param request - ProviderName, Restrictions
     * @returns Fully populated restrictions in the same order as the incoming restrictions.
     */
    public async PopulateRestrictions(request: carriers.Find_PopulateRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;

    /**
     * Take an incoming set of minimally populated restrictions (name + operator is required), and populate all the other parts of the ArchiveRestrictionInfo structure. This includes column information, display values (including list value lookup), and calculated/default values where the value hints specify read-only (R).
     * @param providerName - Provider name to use for populating column information
     * @param restrictions - Restrictions to populate. The Name and Operator fields have to have valid content, and Values should be set as appropriate. Other fields can be left blank or null. If a ColumnInfo is already set, it will not be overwritten.
     * @returns Fully populated restrictions in the same order as the incoming restrictions.
     */
    public async PopulateRestrictions(providerName: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;

    public async PopulateRestrictions(providerName:string|carriers.Find_PopulateRestrictionsRequest, restrictions?:carriers.ArchiveRestrictionInfo[]|base.WebApiRequestOptions, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] > {
        let _request : carriers.Find_PopulateRestrictionsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = providerName;
        if( typeof( first ) === "object" && first && ( typeof(first.ProviderName)!='undefined' || typeof(first.Restrictions)!='undefined' ) )
        {
            _request = providerName as carriers.Find_PopulateRestrictionsRequest;
            _options = restrictions as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       ProviderName: providerName as string,
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       };   
            _options = webapi_options;
        }
        return this._PopulateRestrictions( _request, _options );
    }


    private async _PopulateRestrictionValues(request: carriers.Find_PopulateRestrictionValuesRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveRestrictionInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("PopulateRestrictionValues", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveRestrictionInfo[] = converters.Create_ArchiveRestrictionInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Take an incoming set of Restrictions (name + operator + any user-entered values), and populate/expand all values as specified by the operator&apos;s ValueHints, taking into account any values already there. Used for dynamic date periods; perhaps others in the future
     * @param request - Restrictions
     * @returns Restrictions in the same order as the incoming restrictions, with all values expanded.
     */
    public async PopulateRestrictionValues(request: carriers.Find_PopulateRestrictionValuesRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;

    /**
     * Take an incoming set of Restrictions (name + operator + any user-entered values), and populate/expand all values as specified by the operator&apos;s ValueHints, taking into account any values already there. Used for dynamic date periods; perhaps others in the future
     * @param restrictions - Restrictions to populate. The Name and Operator fields have to have valid content, and Values should be set as appropriate. Other fields can be left blank or null and will not be changed.
     * @returns Restrictions in the same order as the incoming restrictions, with all values expanded.
     */
    public async PopulateRestrictionValues(restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] >;

    public async PopulateRestrictionValues(restrictions:carriers.ArchiveRestrictionInfo[]|carriers.Find_PopulateRestrictionValuesRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveRestrictionInfo[] > {
        let _request : carriers.Find_PopulateRestrictionValuesRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_PopulateRestrictionValuesRequest;
            _options = webapi_options as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       };   
            _options = webapi_options;
        }
        return this._PopulateRestrictionValues( _request, _options );
    }


    private async _GetDefaultDesiredColumnsFromRestrictions(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveColumnInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetDefaultDesiredColumnsFromRestrictions", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveColumnInfo[] = converters.Create_ArchiveColumnInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a provider name and a set of restrictions. This is the algorithm that is used by the Find service method.
     * @param request - ProviderName, Restrictions
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumnsFromRestrictions(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a provider name and a set of restrictions. This is the algorithm that is used by the Find service method.
     * @param providerName - Name of the provider to calculate default desired columns for
     * @param restrictions - Restriction to use in the calculation of default desired columns
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumnsFromRestrictions(providerName: string, restrictions: carriers.ArchiveRestrictionInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    public async GetDefaultDesiredColumnsFromRestrictions(providerName:string|carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest, restrictions?:carriers.ArchiveRestrictionInfo[]|base.WebApiRequestOptions, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        let _request : carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = providerName;
        if( typeof( first ) === "object" && first && ( typeof(first.ProviderName)!='undefined' || typeof(first.Restrictions)!='undefined' ) )
        {
            _request = providerName as carriers.Find_GetDefaultDesiredColumnsFromRestrictionsRequest;
            _options = restrictions as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       ProviderName: providerName as string,
                       Restrictions: restrictions as carriers.ArchiveRestrictionInfo[],
                       };   
            _options = webapi_options;
        }
        return this._GetDefaultDesiredColumnsFromRestrictions( _request, _options );
    }


    private async _GetDefaultOrderByFromDesiredColumns(request: carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveOrderByInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetDefaultOrderByFromDesiredColumns", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveOrderByInfo[] = converters.Create_ArchiveOrderByInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Calculate the default orderby columns for a given provider and a set of desired columns. This is the same algorithm that is used by the Find service method.
     * @param request - ProviderName, DesiredColumns
     * @returns Orderby information
     */
    public async GetDefaultOrderByFromDesiredColumns(request: carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;

    /**
     * Calculate the default orderby columns for a given provider and a set of desired columns. This is the same algorithm that is used by the Find service method.
     * @param providerName - Provider name to calculate default orderby for
     * @param desiredColumns - Desired columns (return fields), used in the orderby calculation. You can generally only order by columns that have been set as 'desired'.
     * @returns Orderby information
     */
    public async GetDefaultOrderByFromDesiredColumns(providerName: string, desiredColumns: string[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] >;

    public async GetDefaultOrderByFromDesiredColumns(providerName:string|carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest, desiredColumns?:string[]|base.WebApiRequestOptions, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveOrderByInfo[] > {
        let _request : carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = providerName;
        if( typeof( first ) === "object" && first && ( typeof(first.ProviderName)!='undefined' || typeof(first.DesiredColumns)!='undefined' ) )
        {
            _request = providerName as carriers.Find_GetDefaultOrderByFromDesiredColumnsRequest;
            _options = desiredColumns as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       ProviderName: providerName as string,
                       DesiredColumns: desiredColumns as string[],
                       };   
            _options = webapi_options;
        }
        return this._GetDefaultOrderByFromDesiredColumns( _request, _options );
    }


    private async _FindOrderBy(request: carriers.Find_FindOrderByRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindOrderBy", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby parameter is used for sorting the results.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param request - StorageType, ProviderName, StorageKey, PageSize, PageNumber, OrderBy
     * @returns Results from search, containing column information and result rows.
     */
    public async FindOrderBy(request: carriers.Find_FindOrderByRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby parameter is used for sorting the results.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @param orderBy - Array of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindOrderBy(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, orderBy: carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindOrderBy(storageType:string|carriers.Find_FindOrderByRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, pageSize?:number, pageNumber?:number, orderBy?:carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindOrderByRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' || typeof(first.OrderBy)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindOrderByRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       OrderBy: orderBy as carriers.ArchiveOrderByInfo[],
                       };   
            _options = webapi_options;
        }
        return this._FindOrderBy( _request, _options );
    }


    private async _FindWithExtraRestrictions(request: carriers.Find_FindWithExtraRestrictionsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindWithExtraRestrictions", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. In addition an extra set of restrictions can be added to the search. These restrictions will not be saved, they are only valid for the current search. Extra restrictions will override restrictions with the same key already stored on the storagekey.
     * @param request - StorageType, ProviderName, StorageKey, ExtraRestrictions, OrderBy, DesiredColumns, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithExtraRestrictions(request: carriers.Find_FindWithExtraRestrictionsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. In addition an extra set of restrictions can be added to the search. These restrictions will not be saved, they are only valid for the current search. Extra restrictions will override restrictions with the same key already stored on the storagekey.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param extraRestrictions - Extra restrictions to append to the the search. These will override saved restrictions with the same key.
     * @param orderBy - Array of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @param desiredColumns - Array of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithExtraRestrictions(storageType: string, providerName: string, storageKey: string, extraRestrictions: carriers.ArchiveRestrictionInfo[], orderBy: carriers.ArchiveOrderByInfo[], desiredColumns: string[], pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindWithExtraRestrictions(storageType:string|carriers.Find_FindWithExtraRestrictionsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, extraRestrictions?:carriers.ArchiveRestrictionInfo[], orderBy?:carriers.ArchiveOrderByInfo[], desiredColumns?:string[], pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindWithExtraRestrictionsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.ExtraRestrictions)!='undefined' || typeof(first.OrderBy)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindWithExtraRestrictionsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       ExtraRestrictions: extraRestrictions as carriers.ArchiveRestrictionInfo[],
                       OrderBy: orderBy as carriers.ArchiveOrderByInfo[],
                       DesiredColumns: desiredColumns as string[],
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindWithExtraRestrictions( _request, _options );
    }


    private async _FindWithColumns(request: carriers.Find_FindWithColumnsRequest, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindWithColumns", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters.
     * @param request - StorageType, ProviderName, StorageKey, DesiredColumns, PageSize, PageNumber, OrderBy
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithColumns(request: carriers.Find_FindWithColumnsRequest, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param desiredColumns - Array of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @param orderBy - Array of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithColumns(storageType: string, providerName: string, storageKey: string, desiredColumns: string[], pageSize: number, pageNumber: number, orderBy: carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindWithColumns(storageType:string|carriers.Find_FindWithColumnsRequest, providerName?:string|base.WebApiRequestOptions, storageKey?:string, desiredColumns?:string[], pageSize?:number, pageNumber?:number, orderBy?:carriers.ArchiveOrderByInfo[], webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindWithColumnsRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' || typeof(first.OrderBy)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindWithColumnsRequest;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       DesiredColumns: desiredColumns as string[],
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       OrderBy: orderBy as carriers.ArchiveOrderByInfo[],
                       };   
            _options = webapi_options;
        }
        return this._FindWithColumns( _request, _options );
    }


    private async _SaveRestrictions2(request: carriers.Find_SaveRestrictions2Request, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictions2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param request - StorageType, ProviderName, StorageKey, Restrictions
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictions2(request: carriers.Find_SaveRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - String of restrictions. "name = 'SuperOffice'" The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictions2(storageType: string, providerName: string, storageKey: string, restrictions: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async SaveRestrictions2(storageType:string|carriers.Find_SaveRestrictions2Request, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:string, webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_SaveRestrictions2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictions2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictions2( _request, _options );
    }


    private async _SaveRestrictionsWithContext2(request: carriers.Find_SaveRestrictionsWithContext2Request, options?:base.WebApiRequestOptions) : Promise< void > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionsWithContext2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param request - StorageType, ProviderName, StorageKey, Restrictions, Context
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictionsWithContext2(request: carriers.Find_SaveRestrictionsWithContext2Request, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). 
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - String of restrictions. The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @param context - Optional context that can be used by FindProvider
     * @returns This service call just saves the restrictions. See SaveRestrictionsAndGetCriteriaInformation if you would like the restrictions returned as criteria immediately, in one roundtrip
     */
    public async SaveRestrictionsWithContext2(storageType: string, providerName: string, storageKey: string, restrictions: string, context: string, webapi_options?:base.WebApiRequestOptions) : Promise< void >;

    public async SaveRestrictionsWithContext2(storageType:string|carriers.Find_SaveRestrictionsWithContext2Request, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:string, context?:string, webapi_options?:base.WebApiRequestOptions) : Promise< void > {
        let _request : carriers.Find_SaveRestrictionsWithContext2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' || typeof(first.Context)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionsWithContext2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as string,
                       Context: context as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionsWithContext2( _request, _options );
    }


    private async _SaveRestrictionsAndGetCriteriaInformation2(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request, options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.CriteriaInformation >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("SaveRestrictionsAndGetCriteriaInformation2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.CriteriaInformation = converters.Create_CriteriaInformation_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). Then, return the same result as a call to GetCriteriaInformation would have done. The purpose is to encapsulate saving and updating of a GUI in one round trip.
     * @param request - StorageType, ProviderName, StorageKey, Restrictions, StaticColumns
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async SaveRestrictionsAndGetCriteriaInformation2(request: carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    /**
     * Save an array of restrictions for later use as search criteria (including as dynamic selection and Find). Then, return the same result as a call to GetCriteriaInformation would have done. The purpose is to encapsulate saving and updating of a GUI in one round trip.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is the intended consumer of the restrictions
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it saves the restrictions as criteria
     * @param restrictions - String of restrictions. The ColumnInfo member and the DisplayValues members need NOT be populated; it is enough to provide a name, operator and any values the operator may need. The IsActive is also saved. Values should be encoded using the CultureDataFormatter to ensure compatibility across cultures.
     * @param staticColumns - Optional string of comma-separated columns that are to be EXCLUDED from the CriteriaArchiveRows part of the result. In the Find dialogs, that corresponds to the 'static' fields, to avoid duplicating them in the 'Match also' criteria list. This array can be null, indicating that all restrictions should be included in the criteria list.
     * @returns The result contains the restrictions in two forms: fully populated ArchiveRestrictionInfo objects, used to display details and for saving changes; and as a list suitable for an Archive control
     */
    public async SaveRestrictionsAndGetCriteriaInformation2(storageType: string, providerName: string, storageKey: string, restrictions: string, staticColumns: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation >;

    public async SaveRestrictionsAndGetCriteriaInformation2(storageType:string|carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request, providerName?:string|base.WebApiRequestOptions, storageKey?:string, restrictions?:string, staticColumns?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.CriteriaInformation > {
        let _request : carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.Restrictions)!='undefined' || typeof(first.StaticColumns)!='undefined' ) )
        {
            _request = storageType as carriers.Find_SaveRestrictionsAndGetCriteriaInformation2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       Restrictions: restrictions as string,
                       StaticColumns: staticColumns as string,
                       };   
            _options = webapi_options;
        }
        return this._SaveRestrictionsAndGetCriteriaInformation2( _request, _options );
    }


    private async _FindFromRestrictions2(request: carriers.Find_FindFromRestrictions2Request, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictions2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are passed in directly, not fetched by a restriction storage provider. The columns of the result are calculated based on the restriction.
     * @param request - Restrictions, ProviderName, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictions2(request: carriers.Find_FindFromRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are passed in directly, not fetched by a restriction storage provider. The columns of the result are calculated based on the restriction.
     * @param restrictions - String of restrictions specifying the search. e.g. "name='SuperOffice'" Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictions2(restrictions: string, providerName: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictions2(restrictions:string|carriers.Find_FindFromRestrictions2Request, providerName?:string|base.WebApiRequestOptions, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictions2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictions2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as string,
                       ProviderName: providerName as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictions2( _request, _options );
    }


    private async _FindFromRestrictionsColumns2(request: carriers.Find_FindFromRestrictionsColumns2Request, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictionsColumns2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is calculated by the system.&lt;para/&gt;Use the GetCriteriaInformation and GetDefaultDesiredColumns service methods to let the system calculate these values, if you want to use or modify them.
     * @param request - Restrictions, ProviderName, DesiredColumns, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumns2(request: carriers.Find_FindFromRestrictionsColumns2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is calculated by the system.&lt;para/&gt;Use the GetCriteriaInformation and GetDefaultDesiredColumns service methods to let the system calculate these values, if you want to use or modify them.
     * @param restrictions - String of restrictions specifying the search. e.g.:"name begins 'Super'". Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param desiredColumns - Array of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumns2(restrictions: string, providerName: string, desiredColumns: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictionsColumns2(restrictions:string|carriers.Find_FindFromRestrictionsColumns2Request, providerName?:string|base.WebApiRequestOptions, desiredColumns?:string, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictionsColumns2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictionsColumns2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as string,
                       ProviderName: providerName as string,
                       DesiredColumns: desiredColumns as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictionsColumns2( _request, _options );
    }


    private async _FindFromRestrictionsColumnsOrderBy2(request: carriers.Find_FindFromRestrictionsColumnsOrderBy2Request, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindFromRestrictionsColumnsOrderBy2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is also passed in directly.&lt;para/&gt;Use the GetCriteriaInformation, GetDefaultDesiredColumns and GetDefaultOrderBy service methods to let the system calculate these values, if you want to use or modify them.
     * @param request - Restrictions, ProviderName, DesiredColumns, OrderBy, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumnsOrderBy2(request: carriers.Find_FindFromRestrictionsColumnsOrderBy2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. &lt;para/&gt;The criteria for the Find are passed in directly, not fetched by a restriction storage provider. &lt;para/&gt;The desired columns of the result set are also passed in directly.&lt;para/&gt;The orderby information is also passed in directly.&lt;para/&gt;Use the GetCriteriaInformation, GetDefaultDesiredColumns and GetDefaultOrderBy service methods to let the system calculate these values, if you want to use or modify them.
     * @param restrictions - String of restrictions specifying the search. Each restriction must match a column of the  given archive provider, and that column must have its CanRestrictBy property set to true.
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param desiredColumns - Comma separated list of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param orderBy - String of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindFromRestrictionsColumnsOrderBy2(restrictions: string, providerName: string, desiredColumns: string, orderBy: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindFromRestrictionsColumnsOrderBy2(restrictions:string|carriers.Find_FindFromRestrictionsColumnsOrderBy2Request, providerName?:string|base.WebApiRequestOptions, desiredColumns?:string, orderBy?:string, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindFromRestrictionsColumnsOrderBy2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = restrictions;
        if( typeof( first ) === "object" && first && ( typeof(first.Restrictions)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.OrderBy)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = restrictions as carriers.Find_FindFromRestrictionsColumnsOrderBy2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       Restrictions: restrictions as string,
                       ProviderName: providerName as string,
                       DesiredColumns: desiredColumns as string,
                       OrderBy: orderBy as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindFromRestrictionsColumnsOrderBy2( _request, _options );
    }


    private async _GetDefaultDesiredColumnsFromRestrictions2(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request, options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ArchiveColumnInfo[] >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetDefaultDesiredColumnsFromRestrictions2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ArchiveColumnInfo[] = converters.Create_ArchiveColumnInfoArray_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a provider name and a set of restrictions. This is the algorithm that is used by the Find service method.
     * @param request - ProviderName, Restrictions
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumnsFromRestrictions2(request: carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    /**
     * Calculate the default desired columns, i.e., the result columns for a given search. The search is defined by a provider name and a set of restrictions. This is the algorithm that is used by the Find service method.
     * @param providerName - Name of the provider to calculate default desired columns for
     * @param restrictions - Restriction to use in the calculation of default desired columns
     * @returns Column information for the default desired columns, fully populated. Percentage-specified column widths sum to exactly 100.
     */
    public async GetDefaultDesiredColumnsFromRestrictions2(providerName: string, restrictions: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] >;

    public async GetDefaultDesiredColumnsFromRestrictions2(providerName:string|carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request, restrictions?:string|base.WebApiRequestOptions, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.ArchiveColumnInfo[] > {
        let _request : carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = providerName;
        if( typeof( first ) === "object" && first && ( typeof(first.ProviderName)!='undefined' || typeof(first.Restrictions)!='undefined' ) )
        {
            _request = providerName as carriers.Find_GetDefaultDesiredColumnsFromRestrictions2Request;
            _options = restrictions as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       ProviderName: providerName as string,
                       Restrictions: restrictions as string,
                       };   
            _options = webapi_options;
        }
        return this._GetDefaultDesiredColumnsFromRestrictions2( _request, _options );
    }


    private async _FindOrderBy2(request: carriers.Find_FindOrderBy2Request, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindOrderBy2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby parameter is used for sorting the results.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param request - StorageType, ProviderName, StorageKey, PageSize, PageNumber, OrderBy
     * @returns Results from search, containing column information and result rows.
     */
    public async FindOrderBy2(request: carriers.Find_FindOrderBy2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. The columns of the result are calculated based on the restriction. The orderby parameter is used for sorting the results.&lt;para/&gt;The other variants of the Find method allow you greater control over the individual aspects of the process.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @param orderBy - Comma separated list of order by specifications. "name asc, dept desc" If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindOrderBy2(storageType: string, providerName: string, storageKey: string, pageSize: number, pageNumber: number, orderBy: string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindOrderBy2(storageType:string|carriers.Find_FindOrderBy2Request, providerName?:string|base.WebApiRequestOptions, storageKey?:string, pageSize?:number, pageNumber?:number, orderBy?:string, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindOrderBy2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' || typeof(first.OrderBy)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindOrderBy2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       OrderBy: orderBy as string,
                       };   
            _options = webapi_options;
        }
        return this._FindOrderBy2( _request, _options );
    }


    private async _FindWithExtraRestrictions2(request: carriers.Find_FindWithExtraRestrictions2Request, options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.FindResults >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("FindWithExtraRestrictions2", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.FindResults = converters.Create_FindResults_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. In addition an extra set of restrictions can be added to the search. These restrictions will not be saved, they are only valid for the current search. Extra restrictions will override restrictions with the same key already stored on the storagekey.
     * @param request - StorageType, ProviderName, StorageKey, ExtraRestrictions, OrderBy, DesiredColumns, PageSize, PageNumber
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithExtraRestrictions2(request: carriers.Find_FindWithExtraRestrictions2Request, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    /**
     * Execute a Find operation and return a page of results. The criteria for the Find are fetched from the restriction storage provider according to the given parameters. In addition an extra set of restrictions can be added to the search. These restrictions will not be saved, they are only valid for the current search. Extra restrictions will override restrictions with the same key already stored on the storagekey.
     * @param storageType - Restriction storage type specification, either 'Criteria' or 'Reporter' (or possible extensions)
     * @param providerName - Name of archive provider that is to execute the search and return the result columns/rows
     * @param storageKey - Storage key to be interpreted by the restriction storage provider, when it fetches criteria for the search
     * @param extraRestrictions - Extra restrictions to append to the the search. These will override saved restrictions with the same key.
     * @param orderBy - Comma separated list of order by specifications. If it is null or empty, the row order is unspecified, database dependent, and might not be the same from call to call, depending on query execution plans. The unspecified order willgenerally not vary within pages of the same query.
     * @param desiredColumns - Comma separated list of column names desired for the result. Each name must match a column offered by the given archive provider.
     * @param pageSize - Size of result set pages
     * @param pageNumber - Result set page to return, 0 is the first page. When a call returns no rows, no further pages are available. Negative page numbers are interpreted as number of rows to skip.
     * @returns Results from search, containing column information and result rows.
     */
    public async FindWithExtraRestrictions2(storageType: string, providerName: string, storageKey: string, extraRestrictions: string, orderBy: string, desiredColumns: string, pageSize: number, pageNumber: number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults >;

    public async FindWithExtraRestrictions2(storageType:string|carriers.Find_FindWithExtraRestrictions2Request, providerName?:string|base.WebApiRequestOptions, storageKey?:string, extraRestrictions?:string, orderBy?:string, desiredColumns?:string, pageSize?:number, pageNumber?:number, webapi_options?:base.WebApiRequestOptions) : Promise< carriers.FindResults > {
        let _request : carriers.Find_FindWithExtraRestrictions2Request = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = storageType;
        if( typeof( first ) === "object" && first && ( typeof(first.StorageType)!='undefined' || typeof(first.ProviderName)!='undefined' || typeof(first.StorageKey)!='undefined' || typeof(first.ExtraRestrictions)!='undefined' || typeof(first.OrderBy)!='undefined' || typeof(first.DesiredColumns)!='undefined' || typeof(first.PageSize)!='undefined' || typeof(first.PageNumber)!='undefined' ) )
        {
            _request = storageType as carriers.Find_FindWithExtraRestrictions2Request;
            _options = providerName as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       StorageType: storageType as string,
                       ProviderName: providerName as string,
                       StorageKey: storageKey as string,
                       ExtraRestrictions: extraRestrictions as string,
                       OrderBy: orderBy as string,
                       DesiredColumns: desiredColumns as string,
                       PageSize: pageSize as number,
                       PageNumber: pageNumber as number,
                       };   
            _options = webapi_options;
        }
        return this._FindWithExtraRestrictions2( _request, _options );
    }

    } // class v1FindAgentController
