// @superoffice/tsclient.webapi/src/HierarchyRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          HierarchyRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Folder structures
 * Collection of all services that works with Lists. These are typical lists of data shown in dropdown list, checkbox lists, etc.
 *
 * `api/v1/Hierarchy`
 */
export interface IHierarchyRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.HierarchyEntity >;
    Post(entity: carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity >;
    PutById(id:number, entity: carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity >;
    DeleteById(HierarchyEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(HierarchyEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.HierarchyEntity >;
} // interface IHierarchyRestApi




/**
 * Folder structures
 * Collection of all services that works with Lists. These are typical lists of data shown in dropdown list, checkbox lists, etc.
 *
 * `api/v1/Hierarchy`
 */
export class HierarchyRestApi extends base.WebApiBase implements IHierarchyRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Hierarchy/";
        super(options);
    }


    /**
     * `GET /api/v1/Hierarchy/default`
     * Set default values into a new HierarchyEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.HierarchyEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.HierarchyEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.HierarchyEntity = converters.Create_HierarchyEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Hierarchy/`
     * Creates a new HierarchyEntity 
     * @param newHierarchy - The HierarchyEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New HierarchyEntity
     */
    public async Post(newHierarchy: carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.HierarchyEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newHierarchy, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.HierarchyEntity = converters.Create_HierarchyEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Hierarchy/123`
     * Updates the existing HierarchyEntity
     * @param hierarchy - The HierarchyEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated HierarchyEntity
     */
    public async PutById(id:number, hierarchy: carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.HierarchyEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), hierarchy, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.HierarchyEntity = null;
                if( response.status != 304 )
                    res = converters.Create_HierarchyEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Hierarchy/123`
     * Updates an existing HierarchyEntity
     * @param hierarchy - The partial HierarchyEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated HierarchyEntity
     */
    public async PatchById(id:number, hierarchy: base.WebApiPatchOperation[] | carriers.HierarchyEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.HierarchyEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.HierarchyEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), hierarchy, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.HierarchyEntity = null;
                if( response.status != 304 )
                    res = converters.Create_HierarchyEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Hierarchy/123`
     * Deletes the HierarchyEntity
     * @param hierarchyId - The id of the HierarchyEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(hierarchyId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( hierarchyId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Hierarchy/123`
     * Get an existing HierarchyEntity
     * @param hierarchyId - The HierarchyEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The HierarchyEntity
     */
    public async GetById( hierarchyId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.HierarchyEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.HierarchyEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( hierarchyId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.HierarchyEntity = null;
                if( response.status != 304 )
                    res = converters.Create_HierarchyEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class HierarchyRestApi
