// @superoffice/tsclient.webapi/src/PersonRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          PersonRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The Person Service. The service implements all services working with the Person object.
 * Person data services.
 *
 * `api/v1/Person`
 */
export interface IPersonRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.PersonEntity >;
    Post(entity: carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity >;
    PutById(id:number, entity: carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity >;
    DeleteById(PersonEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(PersonEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.PersonEntity >;
} // interface IPersonRestApi




/**
 * The Person Service. The service implements all services working with the Person object.
 * Person data services.
 *
 * `api/v1/Person`
 */
export class PersonRestApi extends base.WebApiBase implements IPersonRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Person/";
        super(options);
    }


    /**
     * `GET /api/v1/Person/default`
     * Set default values into a new PersonEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.PersonEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PersonEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.PersonEntity = converters.Create_PersonEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Person/`
     * Creates a new PersonEntity 
     * @param newPerson - The PersonEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New PersonEntity
     */
    public async Post(newPerson: carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PersonEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newPerson, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.PersonEntity = converters.Create_PersonEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Person/123`
     * Updates the existing PersonEntity
     * @param person - The PersonEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated PersonEntity
     */
    public async PutById(id:number, person: carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PersonEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), person, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PersonEntity = null;
                if( response.status != 304 )
                    res = converters.Create_PersonEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Person/123`
     * Updates an existing PersonEntity
     * @param person - The partial PersonEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated PersonEntity
     */
    public async PatchById(id:number, person: base.WebApiPatchOperation[] | carriers.PersonEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.PersonEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PersonEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), person, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PersonEntity = null;
                if( response.status != 304 )
                    res = converters.Create_PersonEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Person/123`
     * Deletes the PersonEntity
     * @param personId - The id of the PersonEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(personId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( personId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Person/123`
     * Get an existing PersonEntity
     * @param personId - The PersonEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The PersonEntity
     */
    public async GetById( personId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.PersonEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PersonEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( personId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PersonEntity = null;
                if( response.status != 304 )
                    res = converters.Create_PersonEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class PersonRestApi
