// @superoffice/tsclient.webapi/src/PreviewsAgent.ts
// <auto-generated>
//      SuperOffice Code Generator
//      modelling\templates\servicesnew\tsagents.tt
// </auto-generated>


/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          WebApi/PreviewsAgent.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Get preview strings from a hint
 */
export interface IPreviewsAgent {
    GetPreview(request: carriers.Previews_GetPreviewRequest, webapi_options?:base.WebApiRequestOptions) : Promise< string >;
    GetPreview(previewHint: string, webapi_options?:base.WebApiRequestOptions) : Promise< string >;
} // interface IPreviewsAgent




/**
 * Get preview strings from a hint
 */
export class PreviewsAgent extends base.WebApiBase implements IPreviewsAgent {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Agents/Previews/";
        super(options);
    }



    private async _GetPreview(request: carriers.Previews_GetPreviewRequest, options?:base.WebApiRequestOptions) : Promise< string > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< string >(async (resolve, reject) => {
            try {
                config.baseURL = this._options.baseUrl;
                const response = await this._axios.post("GetPreview", request , config);
                base.WebApiBase.SetCompleted(options);
                const res : string = response.data as string;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }



    /**
     * Parse a Preview hint and return a Preview. The returned Preview string may contain resource string identifiers (in square brackets), to be processed by the resource manager.&lt;para /&gt;The Preview hint is either a literal text, representing itself, or a set of key/value pairs enclosed in curly braces. Each key is separated from its value by an equals sign, and each pair from the next by an ampersand, according to usual conventions.&lt;para /&gt;A typical Preview hint could be {contact_id=123} or {appointment_id=222&amp;amp;mode=simple}
     * @param request - PreviewHint
     * @returns 
     */
    public async GetPreview(request: carriers.Previews_GetPreviewRequest, webapi_options?:base.WebApiRequestOptions) : Promise< string >;

    /**
     * Parse a Preview hint and return a Preview. The returned Preview string may contain resource string identifiers (in square brackets), to be processed by the resource manager.&lt;para /&gt;The Preview hint is either a literal text, representing itself, or a set of key/value pairs enclosed in curly braces. Each key is separated from its value by an equals sign, and each pair from the next by an ampersand, according to usual conventions.&lt;para /&gt;A typical Preview hint could be {contact_id=123} or {appointment_id=222&amp;amp;mode=simple}
     * @param previewHint - 
     * @returns 
     */
    public async GetPreview(previewHint: string, webapi_options?:base.WebApiRequestOptions) : Promise< string >;

    public async GetPreview(previewHint:string|carriers.Previews_GetPreviewRequest, webapi_options?:base.WebApiRequestOptions) : Promise< string > {
        let _request : carriers.Previews_GetPreviewRequest = null;
        let _options : base.WebApiRequestOptions = null;
        const first : any = previewHint;
        if( typeof( first ) === "object" && first && ( typeof(first.PreviewHint)!='undefined' ) )
        {
            _request = previewHint as carriers.Previews_GetPreviewRequest;
            _options = webapi_options as base.WebApiRequestOptions;
        }
        else
        {
            _request = { 
                       PreviewHint: previewHint as string,
                       };   
            _options = webapi_options;
        }
        return this._GetPreview( _request, _options );
    }

    } // class v1PreviewsAgentController
