// @superoffice/tsclient.webapi/src/PricelistRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          PricelistRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * A pricelist is basically a collection of products. It can be valid in a time period, and outright deactivated. All prices in the product list is in a specific currency. We have decieded not to separate prices and products, which means that we get a simpler data model, but some redundancy.
 * Services for the Quote Management feature, part of the Sale module
 *
 * `api/v1/Pricelist`
 */
export interface IPricelistRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.PriceList >;
    Post(entity: carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList >;
    PutById(id:number, entity: carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList >;
    DeleteById(PriceListId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(PriceListId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.PriceList >;
} // interface IPricelistRestApi




/**
 * A pricelist is basically a collection of products. It can be valid in a time period, and outright deactivated. All prices in the product list is in a specific currency. We have decieded not to separate prices and products, which means that we get a simpler data model, but some redundancy.
 * Services for the Quote Management feature, part of the Sale module
 *
 * `api/v1/Pricelist`
 */
export class PricelistRestApi extends base.WebApiBase implements IPricelistRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Pricelist/";
        super(options);
    }


    /**
     * `GET /api/v1/Pricelist/default`
     * Set default values into a new PriceList.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.PriceList > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PriceList >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.PriceList = converters.Create_PriceList_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Pricelist/`
     * Creates a new PriceList 
     * @param newPricelist - The PriceList to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New PriceList
     */
    public async Post(newPricelist: carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PriceList >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newPricelist, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.PriceList = converters.Create_PriceList_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Pricelist/123`
     * Updates the existing PriceList
     * @param pricelist - The PriceList to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated PriceList
     */
    public async PutById(id:number, pricelist: carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PriceList >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), pricelist, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PriceList = null;
                if( response.status != 304 )
                    res = converters.Create_PriceList_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Pricelist/123`
     * Updates an existing PriceList
     * @param pricelist - The partial PriceList containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated PriceList
     */
    public async PatchById(id:number, pricelist: base.WebApiPatchOperation[] | carriers.PriceList, options?:base.WebApiPutRequestOptions) : Promise< carriers.PriceList > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PriceList >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), pricelist, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PriceList = null;
                if( response.status != 304 )
                    res = converters.Create_PriceList_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Pricelist/123`
     * Deletes the PriceList
     * @param pricelistId - The id of the PriceList to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(pricelistId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( pricelistId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Pricelist/123`
     * Get an existing PriceList
     * @param pricelistId - The PriceList to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The PriceList
     */
    public async GetById( pricelistId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.PriceList > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.PriceList >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( pricelistId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.PriceList = null;
                if( response.status != 304 )
                    res = converters.Create_PriceList_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class PricelistRestApi
