// @superoffice/tsclient.webapi/src/ProductRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          ProductRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Information about a Product from the ERP system. This may be fetched from SuperOffice Product table through the SO Connector
 * Services for the Quote Management feature, part of the Sale module
 *
 * `api/v1/Product`
 */
export interface IProductRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Product >;
    Post(entity: carriers.Product, options?:base.WebApiPutRequestOptions) : Promise< carriers.Product >;
    PutById(id:number, entity: carriers.Product, options?:base.WebApiPutRequestOptions) : Promise< carriers.Product >;
    DeleteById(ProductId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
} // interface IProductRestApi




/**
 * Information about a Product from the ERP system. This may be fetched from SuperOffice Product table through the SO Connector
 * Services for the Quote Management feature, part of the Sale module
 *
 * `api/v1/Product`
 */
export class ProductRestApi extends base.WebApiBase implements IProductRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Product/";
        super(options);
    }


    /**
     * `GET /api/v1/Product/default`
     * Set default values into a new Product.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Product > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Product >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Product = converters.Create_Product_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Product/`
     * Creates a new Product 
     * @param newProduct - The Product to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New Product
     */
    public async Post(newProduct: carriers.Product, options?:base.WebApiPutRequestOptions) : Promise< carriers.Product > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Product >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newProduct, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Product = converters.Create_Product_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Product/123`
     * Updates the existing Product
     * @param product - The Product to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated Product
     */
    public async PutById(id:number, product: carriers.Product, options?:base.WebApiPutRequestOptions) : Promise< carriers.Product > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Product >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), product, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Product = null;
                if( response.status != 304 )
                    res = converters.Create_Product_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `DELETE /api/v1/Product/123`
     * Deletes the Product
     * @param productId - The id of the Product to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(productId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( productId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

} // class ProductRestApi
