// @superoffice/tsclient.webapi/src/ProjectMemberRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          ProjectMemberRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * A person can be a member of multiple projects, with a different comment and member type in each project.
 * Collection of all services that works with Project data.
 *
 * `api/v1/ProjectMember`
 */
export interface IProjectMemberRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ProjectMember >;
    Post(entity: carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember >;
    PutById(id:number, entity: carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember >;
    DeleteById(ProjectMemberId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(ProjectMemberId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ProjectMember >;
} // interface IProjectMemberRestApi




/**
 * A person can be a member of multiple projects, with a different comment and member type in each project.
 * Collection of all services that works with Project data.
 *
 * `api/v1/ProjectMember`
 */
export class ProjectMemberRestApi extends base.WebApiBase implements IProjectMemberRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/ProjectMember/";
        super(options);
    }


    /**
     * `GET /api/v1/ProjectMember/default`
     * Set default values into a new ProjectMember.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ProjectMember > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectMember >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ProjectMember = converters.Create_ProjectMember_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/ProjectMember/`
     * Creates a new ProjectMember 
     * @param newProjectMember - The ProjectMember to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New ProjectMember
     */
    public async Post(newProjectMember: carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectMember >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newProjectMember, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ProjectMember = converters.Create_ProjectMember_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/ProjectMember/123`
     * Updates the existing ProjectMember
     * @param projectMember - The ProjectMember to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ProjectMember
     */
    public async PutById(id:number, projectMember: carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectMember >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), projectMember, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectMember = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectMember_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/ProjectMember/123`
     * Updates an existing ProjectMember
     * @param projectMember - The partial ProjectMember containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ProjectMember
     */
    public async PatchById(id:number, projectMember: base.WebApiPatchOperation[] | carriers.ProjectMember, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectMember > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectMember >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), projectMember, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectMember = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectMember_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/ProjectMember/123`
     * Deletes the ProjectMember
     * @param projectMemberId - The id of the ProjectMember to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(projectMemberId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( projectMemberId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/ProjectMember/123`
     * Get an existing ProjectMember
     * @param projectMemberId - The ProjectMember to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The ProjectMember
     */
    public async GetById( projectMemberId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ProjectMember > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectMember >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( projectMemberId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectMember = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectMember_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class ProjectMemberRestApi
