// @superoffice/tsclient.webapi/src/ProjectRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          ProjectRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The Project Service. The service implements all services working with the Project object
 * Collection of all services that works with Project data.
 *
 * `api/v1/Project`
 */
export interface IProjectRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ProjectEntity >;
    Post(entity: carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity >;
    PutById(id:number, entity: carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity >;
    DeleteById(ProjectEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(ProjectEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ProjectEntity >;
} // interface IProjectRestApi




/**
 * The Project Service. The service implements all services working with the Project object
 * Collection of all services that works with Project data.
 *
 * `api/v1/Project`
 */
export class ProjectRestApi extends base.WebApiBase implements IProjectRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Project/";
        super(options);
    }


    /**
     * `GET /api/v1/Project/default`
     * Set default values into a new ProjectEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ProjectEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ProjectEntity = converters.Create_ProjectEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Project/`
     * Creates a new ProjectEntity 
     * @param newProject - The ProjectEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New ProjectEntity
     */
    public async Post(newProject: carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newProject, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ProjectEntity = converters.Create_ProjectEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Project/123`
     * Updates the existing ProjectEntity
     * @param project - The ProjectEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ProjectEntity
     */
    public async PutById(id:number, project: carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), project, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Project/123`
     * Updates an existing ProjectEntity
     * @param project - The partial ProjectEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ProjectEntity
     */
    public async PatchById(id:number, project: base.WebApiPatchOperation[] | carriers.ProjectEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ProjectEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), project, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Project/123`
     * Deletes the ProjectEntity
     * @param projectId - The id of the ProjectEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(projectId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( projectId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Project/123`
     * Get an existing ProjectEntity
     * @param projectId - The ProjectEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The ProjectEntity
     */
    public async GetById( projectId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ProjectEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ProjectEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( projectId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ProjectEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ProjectEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class ProjectRestApi
