// @superoffice/tsclient.webapi/src/RelationDefinitionRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          RelationDefinitionRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * A relation definition entity with source and destination tables
 * Collection of all services that works with Lists. These are typical lists of data shown in dropdown list, checkbox lists, etc.
 *
 * `api/v1/RelationDefinition`
 */
export interface IRelationDefinitionRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.RelationDefinitionEntity >;
    Post(entity: carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity >;
    PutById(id:number, entity: carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity >;
    GetById(RelationDefinitionEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.RelationDefinitionEntity >;
} // interface IRelationDefinitionRestApi




/**
 * A relation definition entity with source and destination tables
 * Collection of all services that works with Lists. These are typical lists of data shown in dropdown list, checkbox lists, etc.
 *
 * `api/v1/RelationDefinition`
 */
export class RelationDefinitionRestApi extends base.WebApiBase implements IRelationDefinitionRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/RelationDefinition/";
        super(options);
    }


    /**
     * `GET /api/v1/RelationDefinition/default`
     * Set default values into a new RelationDefinitionEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.RelationDefinitionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RelationDefinitionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.RelationDefinitionEntity = converters.Create_RelationDefinitionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/RelationDefinition/`
     * Creates a new RelationDefinitionEntity 
     * @param newRelationDefinition - The RelationDefinitionEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New RelationDefinitionEntity
     */
    public async Post(newRelationDefinition: carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RelationDefinitionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newRelationDefinition, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.RelationDefinitionEntity = converters.Create_RelationDefinitionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/RelationDefinition/123`
     * Updates the existing RelationDefinitionEntity
     * @param relationDefinition - The RelationDefinitionEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated RelationDefinitionEntity
     */
    public async PutById(id:number, relationDefinition: carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RelationDefinitionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), relationDefinition, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RelationDefinitionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RelationDefinitionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/RelationDefinition/123`
     * Updates an existing RelationDefinitionEntity
     * @param relationDefinition - The partial RelationDefinitionEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated RelationDefinitionEntity
     */
    public async PatchById(id:number, relationDefinition: base.WebApiPatchOperation[] | carriers.RelationDefinitionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RelationDefinitionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RelationDefinitionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), relationDefinition, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RelationDefinitionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RelationDefinitionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/RelationDefinition/123`
     * Get an existing RelationDefinitionEntity
     * @param relationDefinitionId - The RelationDefinitionEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The RelationDefinitionEntity
     */
    public async GetById( relationDefinitionId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.RelationDefinitionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RelationDefinitionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( relationDefinitionId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RelationDefinitionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RelationDefinitionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class RelationDefinitionRestApi
