// @superoffice/tsclient.webapi/src/RelationRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          RelationRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Relationship between a (contact/person) and another (contact/person), as described by the RelationDefinition.
 * Contact/Person relations
 *
 * `api/v1/Relation`
 */
export interface IRelationRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ContactRelationEntity >;
    Post(entity: carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity >;
    PutById(id:number, entity: carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity >;
    GetById(ContactRelationEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ContactRelationEntity >;
} // interface IRelationRestApi




/**
 * Relationship between a (contact/person) and another (contact/person), as described by the RelationDefinition.
 * Contact/Person relations
 *
 * `api/v1/Relation`
 */
export class RelationRestApi extends base.WebApiBase implements IRelationRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Relation/";
        super(options);
    }


    /**
     * `GET /api/v1/Relation/default`
     * Set default values into a new ContactRelationEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.ContactRelationEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactRelationEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ContactRelationEntity = converters.Create_ContactRelationEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Relation/`
     * Creates a new ContactRelationEntity 
     * @param newRelation - The ContactRelationEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New ContactRelationEntity
     */
    public async Post(newRelation: carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactRelationEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newRelation, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.ContactRelationEntity = converters.Create_ContactRelationEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Relation/123`
     * Updates the existing ContactRelationEntity
     * @param relation - The ContactRelationEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ContactRelationEntity
     */
    public async PutById(id:number, relation: carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactRelationEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), relation, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactRelationEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactRelationEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Relation/123`
     * Updates an existing ContactRelationEntity
     * @param relation - The partial ContactRelationEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated ContactRelationEntity
     */
    public async PatchById(id:number, relation: base.WebApiPatchOperation[] | carriers.ContactRelationEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.ContactRelationEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactRelationEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), relation, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactRelationEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactRelationEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Relation/123`
     * Get an existing ContactRelationEntity
     * @param relationId - The ContactRelationEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The ContactRelationEntity
     */
    public async GetById( relationId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.ContactRelationEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.ContactRelationEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( relationId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.ContactRelationEntity = null;
                if( response.status != 304 )
                    res = converters.Create_ContactRelationEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class RelationRestApi
