// @superoffice/tsclient.webapi/src/RoleRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          RoleRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * 
 * User administration
 *
 * `api/v1/Role`
 * 
 * **Online Restricted:** This agent is not available in Online by default. User management is not allowed for partner apps.
 */
export interface IRoleRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.RoleEntity >;
    Post(entity: carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity >;
    PutById(id:number, entity: carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity >;
    GetById(RoleEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.RoleEntity >;
} // interface IRoleRestApi




/**
 * 
 * User administration
 *
 * `api/v1/Role`
 */
export class RoleRestApi extends base.WebApiBase implements IRoleRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Role/";
        super(options);
    }


    /**
     * `GET /api/v1/Role/default`
     * Set default values into a new RoleEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.RoleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RoleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.RoleEntity = converters.Create_RoleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Role/`
     * Creates a new RoleEntity 
     * @param newRole - The RoleEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New RoleEntity
     */
    public async Post(newRole: carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RoleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newRole, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.RoleEntity = converters.Create_RoleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Role/123`
     * Updates the existing RoleEntity
     * @param role - The RoleEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated RoleEntity
     */
    public async PutById(id:number, role: carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RoleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), role, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RoleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RoleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Role/123`
     * Updates an existing RoleEntity
     * @param role - The partial RoleEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated RoleEntity
     */
    public async PatchById(id:number, role: base.WebApiPatchOperation[] | carriers.RoleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.RoleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RoleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), role, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RoleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RoleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Role/123`
     * Get an existing RoleEntity
     * @param roleId - The RoleEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The RoleEntity
     */
    public async GetById( roleId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.RoleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.RoleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( roleId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.RoleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_RoleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class RoleRestApi
