// @superoffice/tsclient.webapi/src/SaleStakeholderRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          SaleStakeholderRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * 
 * Sale data services
 *
 * `api/v1/SaleStakeholder`
 */
export interface ISaleStakeholderRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SaleStakeholder >;
    Post(entity: carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder >;
    PutById(id:number, entity: carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder >;
    DeleteById(SaleStakeholderId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(SaleStakeholderId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SaleStakeholder >;
} // interface ISaleStakeholderRestApi




/**
 * 
 * Sale data services
 *
 * `api/v1/SaleStakeholder`
 */
export class SaleStakeholderRestApi extends base.WebApiBase implements ISaleStakeholderRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/SaleStakeholder/";
        super(options);
    }


    /**
     * `GET /api/v1/SaleStakeholder/default`
     * Set default values into a new SaleStakeholder.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SaleStakeholder > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleStakeholder >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SaleStakeholder = converters.Create_SaleStakeholder_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/SaleStakeholder/`
     * Creates a new SaleStakeholder 
     * @param newSaleStakeholder - The SaleStakeholder to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New SaleStakeholder
     */
    public async Post(newSaleStakeholder: carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleStakeholder >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newSaleStakeholder, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SaleStakeholder = converters.Create_SaleStakeholder_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/SaleStakeholder/123`
     * Updates the existing SaleStakeholder
     * @param saleStakeholder - The SaleStakeholder to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SaleStakeholder
     */
    public async PutById(id:number, saleStakeholder: carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleStakeholder >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), saleStakeholder, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleStakeholder = null;
                if( response.status != 304 )
                    res = converters.Create_SaleStakeholder_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/SaleStakeholder/123`
     * Updates an existing SaleStakeholder
     * @param saleStakeholder - The partial SaleStakeholder containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SaleStakeholder
     */
    public async PatchById(id:number, saleStakeholder: base.WebApiPatchOperation[] | carriers.SaleStakeholder, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleStakeholder > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleStakeholder >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), saleStakeholder, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleStakeholder = null;
                if( response.status != 304 )
                    res = converters.Create_SaleStakeholder_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/SaleStakeholder/123`
     * Deletes the SaleStakeholder
     * @param saleStakeholderId - The id of the SaleStakeholder to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(saleStakeholderId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( saleStakeholderId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/SaleStakeholder/123`
     * Get an existing SaleStakeholder
     * @param saleStakeholderId - The SaleStakeholder to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The SaleStakeholder
     */
    public async GetById( saleStakeholderId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SaleStakeholder > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleStakeholder >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( saleStakeholderId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleStakeholder = null;
                if( response.status != 304 )
                    res = converters.Create_SaleStakeholder_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class SaleStakeholderRestApi
