// @superoffice/tsclient.webapi/src/SaleRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          SaleRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The Sale Entity contains the sale amount, currency, and sale members. Sales are linked to contacts, persons, and/or projects.
 * Sale data services
 *
 * `api/v1/Sale`
 */
export interface ISaleRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SaleEntity >;
    Post(entity: carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity >;
    PutById(id:number, entity: carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity >;
    DeleteById(SaleEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(SaleEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SaleEntity >;
} // interface ISaleRestApi




/**
 * The Sale Entity contains the sale amount, currency, and sale members. Sales are linked to contacts, persons, and/or projects.
 * Sale data services
 *
 * `api/v1/Sale`
 */
export class SaleRestApi extends base.WebApiBase implements ISaleRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Sale/";
        super(options);
    }


    /**
     * `GET /api/v1/Sale/default`
     * Set default values into a new SaleEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SaleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SaleEntity = converters.Create_SaleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Sale/`
     * Creates a new SaleEntity 
     * @param newSale - The SaleEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New SaleEntity
     */
    public async Post(newSale: carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newSale, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SaleEntity = converters.Create_SaleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Sale/123`
     * Updates the existing SaleEntity
     * @param sale - The SaleEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SaleEntity
     */
    public async PutById(id:number, sale: carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), sale, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SaleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Sale/123`
     * Updates an existing SaleEntity
     * @param sale - The partial SaleEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SaleEntity
     */
    public async PatchById(id:number, sale: base.WebApiPatchOperation[] | carriers.SaleEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SaleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), sale, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SaleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Sale/123`
     * Deletes the SaleEntity
     * @param saleId - The id of the SaleEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(saleId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( saleId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Sale/123`
     * Get an existing SaleEntity
     * @param saleId - The SaleEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The SaleEntity
     */
    public async GetById( saleId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SaleEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SaleEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( saleId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SaleEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SaleEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class SaleRestApi
