// @superoffice/tsclient.webapi/src/SelectionRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          SelectionRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * 
 * Selections, entities, members and tasks
 *
 * `api/v1/Selection`
 */
export interface ISelectionRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SelectionEntity >;
    Post(entity: carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity >;
    PutById(id:number, entity: carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity >;
    DeleteById(SelectionEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(SelectionEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SelectionEntity >;
} // interface ISelectionRestApi




/**
 * 
 * Selections, entities, members and tasks
 *
 * `api/v1/Selection`
 */
export class SelectionRestApi extends base.WebApiBase implements ISelectionRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Selection/";
        super(options);
    }


    /**
     * `GET /api/v1/Selection/default`
     * Set default values into a new SelectionEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SelectionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SelectionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SelectionEntity = converters.Create_SelectionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Selection/`
     * Creates a new SelectionEntity 
     * @param newSelection - The SelectionEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New SelectionEntity
     */
    public async Post(newSelection: carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SelectionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newSelection, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SelectionEntity = converters.Create_SelectionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Selection/123`
     * Updates the existing SelectionEntity
     * @param selection - The SelectionEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SelectionEntity
     */
    public async PutById(id:number, selection: carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SelectionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), selection, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SelectionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SelectionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Selection/123`
     * Updates an existing SelectionEntity
     * @param selection - The partial SelectionEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SelectionEntity
     */
    public async PatchById(id:number, selection: base.WebApiPatchOperation[] | carriers.SelectionEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SelectionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SelectionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), selection, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SelectionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SelectionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Selection/123`
     * Deletes the SelectionEntity
     * @param selectionId - The id of the SelectionEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(selectionId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( selectionId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Selection/123`
     * Get an existing SelectionEntity
     * @param selectionId - The SelectionEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The SelectionEntity
     */
    public async GetById( selectionId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SelectionEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SelectionEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( selectionId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SelectionEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SelectionEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class SelectionRestApi
