// @superoffice/tsclient.webapi/src/SystemEventRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          SystemEventRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Entity for system events
 * User interface configuration - XMLs and other elements such as inter-client URLs
 *
 * `api/v1/SystemEvent`
 */
export interface ISystemEventRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SystemEventEntity >;
    Post(entity: carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity >;
    PutById(id:number, entity: carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity >;
    DeleteById(SystemEventEntityId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(SystemEventEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SystemEventEntity >;
} // interface ISystemEventRestApi




/**
 * Entity for system events
 * User interface configuration - XMLs and other elements such as inter-client URLs
 *
 * `api/v1/SystemEvent`
 */
export class SystemEventRestApi extends base.WebApiBase implements ISystemEventRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/SystemEvent/";
        super(options);
    }


    /**
     * `GET /api/v1/SystemEvent/default`
     * Set default values into a new SystemEventEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.SystemEventEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SystemEventEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SystemEventEntity = converters.Create_SystemEventEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/SystemEvent/`
     * Creates a new SystemEventEntity 
     * @param newSystemEvent - The SystemEventEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New SystemEventEntity
     */
    public async Post(newSystemEvent: carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SystemEventEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newSystemEvent, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.SystemEventEntity = converters.Create_SystemEventEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/SystemEvent/123`
     * Updates the existing SystemEventEntity
     * @param systemEvent - The SystemEventEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SystemEventEntity
     */
    public async PutById(id:number, systemEvent: carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SystemEventEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), systemEvent, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SystemEventEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SystemEventEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/SystemEvent/123`
     * Updates an existing SystemEventEntity
     * @param systemEvent - The partial SystemEventEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated SystemEventEntity
     */
    public async PatchById(id:number, systemEvent: base.WebApiPatchOperation[] | carriers.SystemEventEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.SystemEventEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SystemEventEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), systemEvent, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SystemEventEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SystemEventEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/SystemEvent/123`
     * Deletes the SystemEventEntity
     * @param systemEventId - The id of the SystemEventEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(systemEventId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( systemEventId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/SystemEvent/123`
     * Get an existing SystemEventEntity
     * @param systemEventId - The SystemEventEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The SystemEventEntity
     */
    public async GetById( systemEventId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.SystemEventEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.SystemEventEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( systemEventId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.SystemEventEntity = null;
                if( response.status != 304 )
                    res = converters.Create_SystemEventEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class SystemEventRestApi
