// @superoffice/tsclient.webapi/src/TicketMessageRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          TicketMessageRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The ticket messages are e-mails, SMS or other messages attached to a ticket (aka request).
 * Agent used for Ticket functions
 *
 * `api/v1/TicketMessage`
 */
export interface ITicketMessageRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.TicketMessageEntity >;
    Post(entity: carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity >;
    PutById(id:number, entity: carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity >;
    GetById(TicketMessageEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.TicketMessageEntity >;
} // interface ITicketMessageRestApi




/**
 * The ticket messages are e-mails, SMS or other messages attached to a ticket (aka request).
 * Agent used for Ticket functions
 *
 * `api/v1/TicketMessage`
 */
export class TicketMessageRestApi extends base.WebApiBase implements ITicketMessageRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/TicketMessage/";
        super(options);
    }


    /**
     * `GET /api/v1/TicketMessage/default`
     * Set default values into a new TicketMessageEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.TicketMessageEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketMessageEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.TicketMessageEntity = converters.Create_TicketMessageEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/TicketMessage/`
     * Creates a new TicketMessageEntity 
     * @param newTicketMessage - The TicketMessageEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New TicketMessageEntity
     */
    public async Post(newTicketMessage: carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketMessageEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newTicketMessage, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.TicketMessageEntity = converters.Create_TicketMessageEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/TicketMessage/123`
     * Updates the existing TicketMessageEntity
     * @param ticketMessage - The TicketMessageEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketMessageEntity
     */
    public async PutById(id:number, ticketMessage: carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketMessageEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), ticketMessage, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketMessageEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketMessageEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/TicketMessage/123`
     * Updates an existing TicketMessageEntity
     * @param ticketMessage - The partial TicketMessageEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketMessageEntity
     */
    public async PatchById(id:number, ticketMessage: base.WebApiPatchOperation[] | carriers.TicketMessageEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketMessageEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketMessageEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), ticketMessage, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketMessageEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketMessageEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/TicketMessage/123`
     * Get an existing TicketMessageEntity
     * @param ticketMessageId - The TicketMessageEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The TicketMessageEntity
     */
    public async GetById( ticketMessageId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.TicketMessageEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketMessageEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( ticketMessageId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketMessageEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketMessageEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class TicketMessageRestApi
