// @superoffice/tsclient.webapi/src/TicketRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          TicketRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * The ticket entity is used for representing a service ticket
 * Agent used for Ticket functions
 *
 * `api/v1/Ticket`
 */
export interface ITicketRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.TicketEntity >;
    Post(entity: carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity >;
    PutById(id:number, entity: carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity >;
    GetById(TicketEntityId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.TicketEntity >;
} // interface ITicketRestApi




/**
 * The ticket entity is used for representing a service ticket
 * Agent used for Ticket functions
 *
 * `api/v1/Ticket`
 */
export class TicketRestApi extends base.WebApiBase implements ITicketRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Ticket/";
        super(options);
    }


    /**
     * `GET /api/v1/Ticket/default`
     * Set default values into a new TicketEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.TicketEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.TicketEntity = converters.Create_TicketEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Ticket/`
     * Creates a new TicketEntity 
     * @param newTicket - The TicketEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New TicketEntity
     */
    public async Post(newTicket: carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newTicket, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.TicketEntity = converters.Create_TicketEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Ticket/123`
     * Updates the existing TicketEntity
     * @param ticket - The TicketEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketEntity
     */
    public async PutById(id:number, ticket: carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), ticket, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Ticket/123`
     * Updates an existing TicketEntity
     * @param ticket - The partial TicketEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketEntity
     */
    public async PatchById(id:number, ticket: base.WebApiPatchOperation[] | carriers.TicketEntity, options?:base.WebApiPutRequestOptions) : Promise< carriers.TicketEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), ticket, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Ticket/123`
     * Get an existing TicketEntity
     * @param ticketId - The TicketEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The TicketEntity
     */
    public async GetById( ticketId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.TicketEntity > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.TicketEntity >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( ticketId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.TicketEntity = null;
                if( response.status != 304 )
                    res = converters.Create_TicketEntity_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class TicketRestApi
