// @superoffice/tsclient.webapi/src/WebhookRestApi.ts

/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          WebhookRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------

import { AxiosRequestConfig, AxiosError } from 'axios';
import * as moment from 'moment';
import * as enums from './Enums';
import * as base from './WebApiBase';
import * as carriers from './Carriers';
import * as converters from './Converters';
import { FromBase64ToString, ToBase64 } from './Base64';

/**
 * Webhook definitions. Webhooks broadcast events from NetServer to remote servers.
 * Webhook definitions - webhooks signal other systems about events inside NetServer
 *
 * `api/v1/Webhook`
 * 
 * **Online Restricted:** This agent is not available in Online by default. Access must be requested specifically when app is registered.
 */
export interface IWebhookRestApi {

    GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Webhook >;
    Post(entity: carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook >;
    PutById(id:number, entity: carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook >;
    PatchById(id:number, entity: base.WebApiPatchOperation[] | carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook >;
    DeleteById(WebhookId: number, options?:base.WebApiPutRequestOptions) : Promise<void>;
    GetById(WebhookId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.Webhook >;
} // interface IWebhookRestApi




/**
 * Webhook definitions. Webhooks broadcast events from NetServer to remote servers.
 * Webhook definitions - webhooks signal other systems about events inside NetServer
 *
 * `api/v1/Webhook`
 */
export class WebhookRestApi extends base.WebApiBase implements IWebhookRestApi {
    constructor(baseUrl?: string, languageCode?:string, cultureCode?: string);
    constructor(baseUrl?: string, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceManager:any);
    constructor(baseUrl: string, config:AxiosRequestConfig, resourceParsing?: base.ResourceParsing);
    constructor(baseUrl: string, config:AxiosRequestConfig, languageCode?:string, cultureCode?: string);
    constructor(options: base.WebApiOptions);

    constructor(baseUrlOrOptions: string|base.WebApiOptions|undefined,
                langOrconfigOrRM?: string|base.ResourceParsing|AxiosRequestConfig|any, 
                langOrResourceOrRM?:string|base.ResourceParsing|any, 
                cultureCode?: string, 
                ) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Webhook/";
        super(options);
    }


    /**
     * `GET /api/v1/Webhook/default`
     * Set default values into a new Webhook.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    public async GetDefault(options?:base.WebApiGetRequestOptions): Promise< carriers.Webhook > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Webhook >(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Webhook = converters.Create_Webhook_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `POST /api/v1/Webhook/`
     * Creates a new Webhook 
     * @param newWebhook - The Webhook to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New Webhook
     */
    public async Post(newWebhook: carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Webhook >(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newWebhook, config);
                base.WebApiBase.SetCompleted(options);
                const res : carriers.Webhook = converters.Create_Webhook_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


    /*
     * `PUT /api/v1/Webhook/123`
     * Updates the existing Webhook
     * @param webhook - The Webhook to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated Webhook
     */
    public async PutById(id:number, webhook: carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Webhook >(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), webhook, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Webhook = null;
                if( response.status != 304 )
                    res = converters.Create_Webhook_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }

    /*
     * `PATCH /api/v1/Webhook/123`
     * Updates an existing Webhook
     * @param webhook - The partial Webhook containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated Webhook
     */
    public async PatchById(id:number, webhook: base.WebApiPatchOperation[] | carriers.Webhook, options?:base.WebApiPutRequestOptions) : Promise< carriers.Webhook > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Webhook >(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), webhook, config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Webhook = null;
                if( response.status != 304 )
                    res = converters.Create_Webhook_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Webhook/123`
     * Deletes the Webhook
     * @param webhookId - The id of the Webhook to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    public async DeleteById(webhookId: number, options?:base.WebApiPutRequestOptions) : Promise<void> {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< void >(async (resolve, reject) => {
            try {
                const response = await this._axios.delete( webhookId.toString(), config );
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Webhook/123`
     * Get an existing Webhook
     * @param webhookId - The Webhook to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The Webhook
     */
    public async GetById( webhookId: number, options?:base.WebApiGetRequestOptions) :  Promise< carriers.Webhook > {
        const config : AxiosRequestConfig = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise< carriers.Webhook >(async (resolve, reject) => {
            try {
                const response = await this._axios.get( webhookId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res : carriers.Webhook = null;
                if( response.status != 304 )
                    res = converters.Create_Webhook_From( response.data );
                resolve(res);
            }
            catch( err ) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }


} // class WebhookRestApi
