// @superoffice/tsclient.webapi/dist/AppointmentRestApi.js
"use strict";
/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          AppointmentRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppointmentRestApi = void 0;
const tslib_1 = require("tslib");
const base = tslib_1.__importStar(require("./WebApiBase"));
const converters = tslib_1.__importStar(require("./Converters"));
/**
 * Appointments appear in the diary, and have links to a Contact/Person and possibly a Project or Sale. They have start and end time+date.
 * Collection of all services that works with Appointment data.
 *
 * `api/v1/Appointment`
 */
class AppointmentRestApi extends base.WebApiBase {
    constructor(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Appointment/";
        super(options);
    }
    /**
     * `GET /api/v1/Appointment/default`
     * Set default values into a new AppointmentEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    async GetDefault(options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res = converters.Create_AppointmentEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `POST /api/v1/Appointment/`
     * Creates a new AppointmentEntity
     * @param newAppointment - The AppointmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New AppointmentEntity
     */
    async Post(newAppointment, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newAppointment, config);
                base.WebApiBase.SetCompleted(options);
                const res = converters.Create_AppointmentEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `PUT /api/v1/Appointment/123`
     * Updates the existing AppointmentEntity
     * @param appointment - The AppointmentEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AppointmentEntity
     */
    async PutById(id, appointment, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), appointment, config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_AppointmentEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `PATCH /api/v1/Appointment/123`
     * Updates an existing AppointmentEntity
     * @param appointment - The partial AppointmentEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated AppointmentEntity
     */
    async PatchById(id, appointment, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), appointment, config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_AppointmentEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `DELETE /api/v1/Appointment/123`
     * Deletes the AppointmentEntity
     * @param appointmentId - The id of the AppointmentEntity to be deleted.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     */
    async DeleteById(appointmentId, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.delete(appointmentId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                const res = response.data;
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Appointment/123`
     * Get an existing AppointmentEntity
     * @param appointmentId - The AppointmentEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The AppointmentEntity
     */
    async GetById(appointmentId, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.get(appointmentId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_AppointmentEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
} // class AppointmentRestApi 
exports.AppointmentRestApi = AppointmentRestApi;
