// @superoffice/tsclient.webapi/dist/TicketRestApi.js
"use strict";
/* tslint:disable */
//----------------------------------------------------------------------------
// G E N E R A T E D   F I L E !                       D O   N O T   E D I T !
//----------------------------------------------------------------------------
//
// File:          TicketRestApi.ts
//
//----------------------------------------------------------------------------
//                                                Copyright (C) SuperOffice AS
//----------------------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.TicketRestApi = void 0;
const tslib_1 = require("tslib");
const base = tslib_1.__importStar(require("./WebApiBase"));
const converters = tslib_1.__importStar(require("./Converters"));
/**
 * The ticket entity is used for representing a service ticket
 * Agent used for Ticket functions
 *
 * `api/v1/Ticket`
 */
class TicketRestApi extends base.WebApiBase {
    constructor(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode) {
        const options = base.WebApiBase.FigureOutOptions(baseUrlOrOptions, langOrconfigOrRM, langOrResourceOrRM, cultureCode);
        options.basePath = "v1/Ticket/";
        super(options);
    }
    /**
     * `GET /api/v1/Ticket/default`
     * Set default values into a new TicketEntity.
     * NetServer calculates default values on the entity, which is required when creating/storing a new instance
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns A blank entity.
     */
    async GetDefault(options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.get("default", config);
                base.WebApiBase.SetCompleted(options);
                const res = converters.Create_TicketEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `POST /api/v1/Ticket/`
     * Creates a new TicketEntity
     * @param newTicket - The TicketEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns New TicketEntity
     */
    async Post(newTicket, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.post("", newTicket, config);
                base.WebApiBase.SetCompleted(options);
                const res = converters.Create_TicketEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `PUT /api/v1/Ticket/123`
     * Updates the existing TicketEntity
     * @param ticket - The TicketEntity to be saved.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketEntity
     */
    async PutById(id, ticket, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.put(id.toString(), ticket, config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_TicketEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `PATCH /api/v1/Ticket/123`
     * Updates an existing TicketEntity
     * @param ticket - The partial TicketEntity containing updated values, or an array of RFC6902 PATCH operations.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The updated TicketEntity
     */
    async PatchById(id, ticket, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.patch(id.toString(), ticket, config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_TicketEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
    /*
     * `GET /api/v1/Ticket/123`
     * Get an existing TicketEntity
     * @param ticketId - The TicketEntity to be get.
     * @param options - Override language/culture codes on this request. Provide cancellation and request status values.
     * @returns The TicketEntity
     */
    async GetById(ticketId, options) {
        const config = base.WebApiBase.GetAxiosRequestConfig(options, this._options);
        base.WebApiBase.SetPending(options);
        return new Promise(async (resolve, reject) => {
            try {
                const response = await this._axios.get(ticketId.toString(), config);
                base.WebApiBase.SetCompleted(options);
                let res = null;
                if (response.status != 304)
                    res = converters.Create_TicketEntity_From(response.data);
                resolve(res);
            }
            catch (err) {
                base.WebApiBase.HandleError(options, config, err);
                reject(err);
            }
        });
    }
} // class TicketRestApi 
exports.TicketRestApi = TicketRestApi;
